{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# The BioBB REST API\n",
    "\n",
    "The **[BioBB REST API](https://mmb.irbbarcelona.org/biobb-api)** allows the execution of the **[BioExcel Building Blocks](https://mmb.irbbarcelona.org/biobb/)** in a remote server.\n",
    "\n",
    "## Documentation\n",
    "\n",
    "For an extense documentation section, please go to the **[BioBB REST API website help](https://mmb.irbbarcelona.org/biobb-api/rest)**.\n",
    "\n",
    "## Settings\n",
    "\n",
    "### Auxiliar libraries used\n",
    "\n",
    "* [requests](https://pypi.org/project/requests/): Requests allows you to send *organic, grass-fed* HTTP/1.1 requests, without the need for manual labor.\n",
    "* [nb_conda_kernels](https://github.com/Anaconda-Platform/nb_conda_kernels): Enables a Jupyter Notebook or JupyterLab application in one conda environment to access kernels for Python, R, and other languages found in other environments.\n",
    "* [nglview](http://nglviewer.org/#nglview): Jupyter/IPython widget to interactively view molecular structures and trajectories in notebooks.\n",
    "* [ipywidgets](https://github.com/jupyter-widgets/ipywidgets): Interactive HTML widgets for Jupyter notebooks and the IPython kernel.\n",
    "* [plotly](https://plot.ly/python/offline/): Python interactive graphing library integrated in Jupyter notebooks.\n",
    "\n",
    "### Conda Installation and Launch\n",
    "\n",
    "```console\n",
    "git clone https://github.com/bioexcel/biobb_REST_API_documentation.git\n",
    "cd biobb_REST_API_documentation\n",
    "conda env create -f conda_env/environment.yml\n",
    "conda activate biobb_REST_API_documentation\n",
    "jupyter-nbextension enable --py --user widgetsnbextension\n",
    "jupyter-nbextension enable --py --user nglview\n",
    "jupyter-notebook biobb_REST_API_documentation/notebooks/biobb_REST_API_documentation.ipynb\n",
    "```\n",
    "\n",
    "***\n",
    "\n",
    "## Index\n",
    "\n",
    " * [Behaviour](#behaviour)\n",
    "     * [Tools information](#tools_info)\n",
    "         * [List of packages](#list_pckg)\n",
    "         * [List of tools](#list_tools)\n",
    "         * [Tool's properties](#tools_prop)\n",
    "     * [Launch tool](#launch_tool)\n",
    "     * [Retrieve status](#retrieve_status)\n",
    "     * [Retrieve data](#retrieve_data)\n",
    "     * [Sample files](#sample_files)\n",
    "         * [All sample files](#all_sample)\n",
    "         * [Package sample files](#pckg_sample)\n",
    "         * [Tool sample files](#tool_sample)\n",
    "         * [Single sample file](#sample)\n",
    " * [Examples](#examples)\n",
    "     * [Tools information](#tools_info_ex)\n",
    "         * [List of packages](#list_pckg_ex)\n",
    "         * [List of tools from a specific package](#list_tools_ex)\n",
    "         * [Tool's properties](#tools_prop_ex)\n",
    "     * [Launch tool](#launch_tool_ex)\n",
    "         * [Launch job with a YAML file config](#tool_yml_ex)\n",
    "         * [Launch job with a JSON file config](#tool_json_ex)\n",
    "         * [Launch job with a piython dictionary config](#tool_dict_ex)\n",
    "     * [Retrieve status](#retrieve_status_ex)\n",
    "     * [Retrieve data](#retrieve_data_ex)\n",
    " * [Practical cases](#practical_cases)\n",
    "     * [Example 1: download PDB file from RSCB database](#example1)\n",
    "     * [Example 2: extract heteroatom from a given structure](#example2)\n",
    "     * [Example 3: extract energy components from a given GROMACS energy file](#example3)\n",
    "\n",
    "***\n",
    "<img src=\"https://bioexcel.eu/wp-content/uploads/2019/04/Bioexcell_logo_1080px_transp.png\" alt=\"Bioexcel2 logo\"\n",
    "\ttitle=\"Bioexcel2 logo\" width=\"400\" />\n",
    "***\n",
    "\n",
    "<a id=\"behaviour\"></a>\n",
    "## Behaviour\n",
    "\n",
    "The **BioBB REST API** works as an asynchronous launcher of jobs, as these jobs can last from a few seconds to several minutes, there are some steps that must be performed for having the complete results of every tool.\n",
    "\n",
    "**BioExcel Building Blocks** are structured in **[packages and tools](http://mmb.irbbarcelona.org/biobb/availability/source)**. Every call to the **BioBB REST API** executes one single tool and returns the output file(s) related to this specific tool.\n",
    "\n",
    "<a id=\"tools_info\"></a>\n",
    "### Tools information\n",
    "\n",
    "<a id=\"list_pckg\"></a>\n",
    "#### List of packages\n",
    "\n",
    "In order to get a complete **list of available packages**, we must do a **GET** request to the following endpoint:\n",
    "\n",
    "`http://mmb.irbbarcelona.org/biobb-api/rest/v1/launch`\n",
    "\n",
    "This endpoint returns a **JSON HTTP response** with status `200`. More information in the [BioBB REST API Documentation section](http://mmb.irbbarcelona.org/biobb-api/rest/#/List%20of%20Services/getPckgList).\n",
    "\n",
    "<a id=\"list_tools\"></a>\n",
    "#### List of tools\n",
    "\n",
    "If there is need for a **list of tools for a single package**, we must do a **GET** request to the following endpoint:\n",
    "\n",
    "`http://mmb.irbbarcelona.org/biobb-api/rest/v1/launch/{package}`\n",
    "\n",
    "This endpoint returns a **JSON HTTP response** with status `200` or a `404` status if the package id is incorrect. More information in the [BioBB REST API Documentation section](http://mmb.irbbarcelona.org/biobb-api/rest/#/List%20of%20Services/getToolsList).\n",
    "\n",
    "<a id=\"tools_prop\"></a>\n",
    "#### Tool's properties\n",
    "\n",
    "If there is only need for the **information of a single tool**, we must do a **GET** request to the following endpoint:\n",
    "\n",
    "`http://mmb.irbbarcelona.org/biobb-api/rest/v1/launch/{package}/{tool}`\n",
    "\n",
    "This endpoint returns a **JSON HTTP response** with status `200` or a `404` status if the package id and / or the tool id are incorrect. The reason for failure should be detailed in the JSON response. More information in the [BioBB REST API Documentation section](http://mmb.irbbarcelona.org/biobb-api/rest/#/Launch%20Tool/getLaunchTool).\n",
    "\n",
    "<a id=\"launch_tool\"></a>\n",
    "### Launch tool\n",
    "\n",
    "For **launching a tool**, we must do a **POST** request to the following endpoint:\n",
    "\n",
    "`http://mmb.irbbarcelona.org/biobb-api/rest/v1/launch/{package}/{tool}`\n",
    "\n",
    "In the body of this POST request, **we must add the file(s) needed as input** (included the properties config file in **JSON** or **YAML** format) and the name for the output(s). The detailed list of inputs and outputs with its respectives properties can be found in the **GET** request of this same endpoint.\n",
    "\n",
    "This endpoint returns a **JSON HTTP response** with the following possible status:\n",
    "\n",
    "* `303`: **The job has been successfully launched** and the user must save the token provided and follow to the next endpoint (defined in the same JSON response)\n",
    "* `404`: **There was some error launching the tool.** The reason for failure should be detailed in the JSON response.\n",
    "* `500`: The job has been launched, but **some internal server error** has occurred during the execution.\n",
    "\n",
    "More information for a generic call in the [BioBB REST API Documentation section](http://mmb.irbbarcelona.org/biobb-api/rest/#/Launch%20Tool/postLaunchTool). The documentation for all the tools is available in the [BioBB REST API Tools Documentation section](http://mmb.irbbarcelona.org/biobb-api/tools-documentation?docExpansion=none). Interactive examples for all the tools are available in the [BioBB REST API Tools Execution section](http://mmb.irbbarcelona.org/biobb-api/tools-execution).\n",
    "\n",
    "<a id=\"retrieve_status\"></a>\n",
    "### Retrieve status\n",
    "\n",
    "If the previous endpoint returned a `303` status, we must do a **GET** request to the following endpoint providing the given token in the path:\n",
    "\n",
    "`http://mmb.irbbarcelona.org/biobb-api/rest/v1/retrieve/status/{token}`\n",
    "\n",
    "This endpoint checks the state of the job and returns a **JSON HTTP response** with the following possible status:\n",
    "\n",
    "* `200`: **The job has finished successfully** and in the JSON response we can found a list of output files generated by the job with its correspondent id for retrieving them on the next endpoint (defined in the same JSON message).\n",
    "* `202`: The job is **still running**.\n",
    "* `404`: **Token incorrect, job unexisting or expired.**\n",
    "* `500`: Some **internal server error** has occurred during the execution.\n",
    "\n",
    "More information in the [BioBB REST API Documentation section](http://mmb.irbbarcelona.org/biobb-api/rest/#/Retrieve/getRetrieveStatus).\n",
    "\n",
    "<a id=\"retrieve_data\"></a>\n",
    "### Retrieve data\n",
    "\n",
    "Once the previous endpoint returns a `200` status, the output file(s) are ready for its retrieval, so we must do a **GET** request to the following endpoint providing the given **file id** in the path:\n",
    "\n",
    "`http://mmb.irbbarcelona.org/biobb-api/rest/v1/retrieve/data/{id}`\n",
    "\n",
    "This endpoint returns the **requested file** with a `200` status or a `404` status if the provided id is incorrect, the file doesn't exist or it has expired. More information in the [BioBB REST API Documentation section](http://mmb.irbbarcelona.org/biobb-api/rest/#/Retrieve/getRetrieveData).\n",
    "\n",
    "Note that if we have executed a job that returns multiple output files, a call to this endpoint must be done **for each of the output files** generated by the job.\n",
    "\n",
    "<a id=\"sample_files\"></a>\n",
    "### Sample files\n",
    "\n",
    "The **BioBB REST API** provides sample files for most of the inputs and outputs of each tool. Files can be accessed thought the whole **BioBB REST API** hierarchical range.\n",
    "\n",
    "<a id=\"all_sample\"></a>\n",
    "#### All sample files\n",
    "\n",
    "In order to download **all the sample files**, we must do a **GET** request to the following endpoint:\n",
    "\n",
    "`http://mmb.irbbarcelona.org/biobb-api/rest/v1/sample`\n",
    "\n",
    "This endpoint returns the **requested file** with a `200` status. More information in the [BioBB REST API Documentation section](http://mmb.irbbarcelona.org/biobb-api/rest#/Sample%20Files/getSample).\n",
    "\n",
    "<a id=\"pckg_sample\"></a>\n",
    "#### Package sample files\n",
    "\n",
    "In order to download **all the sample files of a package**, we must do a **GET** request to the following endpoint:\n",
    "\n",
    "`http://mmb.irbbarcelona.org/biobb-api/rest/v1/sample/{package}`\n",
    "\n",
    "This endpoint returns the **requested file** with a `200` status or a `404` status if the package id is incorrect. More information in the [BioBB REST API Documentation section](http://mmb.irbbarcelona.org/biobb-api/rest#/Sample%20Files/getPackageSample).\n",
    "\n",
    "<a id=\"tool_sample\"></a>\n",
    "#### Tool sample files\n",
    "\n",
    "In order to download **all the sample files of a tool**, we must do a **GET** request to the following endpoint:\n",
    "\n",
    "`http://mmb.irbbarcelona.org/biobb-api/rest/v1/sample/{package}/{tool}`\n",
    "\n",
    "This endpoint returns the **requested file** with a `200` status or a `404` status if the package id and / or the tool id are incorrect. The reason for failure should be detailed in the JSON response. More information in the [BioBB REST API Documentation section](http://mmb.irbbarcelona.org/biobb-api/rest#/Sample%20Files/getToolSample).\n",
    "\n",
    "<a id=\"sample\"></a>\n",
    "#### Single sample file\n",
    "\n",
    "In order to download **a single sample file**, we must do a **GET** request to the following endpoint:\n",
    "\n",
    "`http://mmb.irbbarcelona.org/biobb-api/rest/v1/sample/{package}/{tool}/{id}`\n",
    "\n",
    "This endpoint returns the **requested file** with a `200` status or a `404` status if the package id and / or the tool id and / or the file id are incorrect. The reason for failure should be detailed in the JSON response. More information in the [BioBB REST API Documentation section](http://mmb.irbbarcelona.org/biobb-api/rest#/Sample%20Files/getSingleSample)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"examples\"></a>\n",
    "## Examples\n",
    "\n",
    "Below we will do **calls to all the previously defined endpoints** and define some **functions** for make easier the connection to the **BioBB REST API** through **Jupyter Notebook**.\n",
    "\n",
    "First off, we will import the Python requests and json library and set the root URI for the **BioBB REST API**."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "\n",
    "apiURL  = \"http://mmb.irbbarcelona.org/biobb-api/rest/v1/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"tools_info_ex\"></a>\n",
    "### Tools information"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Definition of simple GET / POST request functions and a class Response:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Class for returning response status and json content of a requested URL\n",
    "class Response:\n",
    "  def __init__(self, status, json):\n",
    "    self.status = status\n",
    "    self.json = json\n",
    "\n",
    "# Perform GET request\n",
    "def get_data(url):\n",
    "    r = requests.get(url)\n",
    "    return Response(r.status_code, json.loads(r.text))\n",
    "\n",
    "# Perform POST request\n",
    "def post_data(url, d, f):\n",
    "    r = requests.post(url, data = d, files = f)\n",
    "    return Response(r.status_code, json.loads(r.text))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"list_pckg_ex\"></a>\n",
    "#### List of packages\n",
    "\n",
    "For more information about this endpoint, please visit the [BioBB REST API Documentation section](http://mmb.irbbarcelona.org/biobb-api/rest#/List%20of%20Services/getPckgList).\n",
    "\n",
    "##### Endpoint"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**GET** `http://mmb.irbbarcelona.org/biobb-api/rest/v1/launch`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"packages\": [\n",
      "    {\n",
      "      \"id\": \"biobb_analysis\",\n",
      "      \"tools\": [\n",
      "        {\n",
      "          \"id\": \"gmx_cluster\",\n",
      "          \"description\": \"Creates cluster structures from a given GROMACS compatible trajectory\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the gmx_cluster tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_cluster.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.tpr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.brk$\",\n",
      "                \".*\\\\.ent$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS trajectory file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.cpt$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_index_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the GROMACS index file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output cluster file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_cluster.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.cpt$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"gmx_rms\",\n",
      "          \"description\": \"Performs a Root Mean Square deviation (RMSd) analysis from a given GROMACS compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the gmx_rms tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_rms.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.tpr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.brk$\",\n",
      "                \".*\\\\.ent$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS trajectory file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.cpt$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_index_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the GROMACS index file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_xvg_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the XVG output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_rms.xvg\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xvg$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"gmx_rgyr\",\n",
      "          \"description\": \"Computes the radius of gyration (Rgyr) of a molecule about the x-, y- and z-axes, as a function of time, from a given GROMACS compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the gmx_rgyr tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_rgyr.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.tpr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.brk$\",\n",
      "                \".*\\\\.ent$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS trajectory file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.cpt$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_index_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the GROMACS index file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_xvg_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the XVG output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_rgyr.xvg\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xvg$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"gmx_energy\",\n",
      "          \"description\": \"Extracts energy components from a given GROMACS energy file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the gmx_energy tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_energy.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_energy_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input EDR file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/energy.edr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.edr$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_xvg_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the XVG output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_energy.xvg\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xvg$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"gmx_image\",\n",
      "          \"description\": \"Corrects periodicity (image) from a given GROMACS compatible trajectory file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the gmx_image tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_image.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS trajectory file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.cpt$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS input topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.tpr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.brk$\",\n",
      "                \".*\\\\.ent$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_index_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the GROMACS index file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_image.xtc\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"gmx_trjconv_str\",\n",
      "          \"description\": \"Converts between GROMACS compatible structure file formats and/or extracts a selection of atoms.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the gmx_trjconv_str tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_trjconv_str.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.cpt$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS input topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.tpr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.brk$\",\n",
      "                \".*\\\\.ent$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_index_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the GROMACS index file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_str_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_trjconv.str.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"gmx_trjconv_str_ens\",\n",
      "          \"description\": \"Extracts an ensemble of frames containing a selection of atoms from GROMACS compatible trajectory files.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the gmx_trjconv_str_ens tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_trjconv_str_ens.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS trajectory file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.cpt$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS input topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.tpr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.brk$\",\n",
      "                \".*\\\\.ent$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_index_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the GROMACS index file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_str_ens_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_trjconv.str.ens.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"gmx_trjconv_trj\",\n",
      "          \"description\": \"Converts between GROMACS compatible trajectory file formats and/or extracts a selection of atoms.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the gmx_trjconv_trj tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_trjconv_trj.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS trajectory file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.cpt$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_index_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the GROMACS index file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_trjconv.trj.xtc\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_average\",\n",
      "          \"description\": \"Calculates a structure average of a given cpptraj compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_average tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_average.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed structure\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.average.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.rst7$\",\n",
      "                \".*\\\\.ncrst$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.sqm$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_bfactor\",\n",
      "          \"description\": \"Calculates the Bfactor fluctuations of a given cpptraj compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_bfactor tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_exp_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the experimental reference file (required if reference = experimental)\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/experimental.1e5t.pdb\",\n",
      "              \"formats\": null\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed analysis\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.bfactor.first.dat\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.dat$\",\n",
      "                \".*\\\\.agr$\",\n",
      "                \".*\\\\.xmgr$\",\n",
      "                \".*\\\\.gnu$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_rms\",\n",
      "          \"description\": \"Calculates the Root Mean Square deviation (RMSd) of a given cpptraj compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_rms tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_exp_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the experimental reference file (required if reference = experimental)\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/experimental.1e5t.pdb\",\n",
      "              \"formats\": null\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed analysis\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.rms.first.dat\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.dat$\",\n",
      "                \".*\\\\.agr$\",\n",
      "                \".*\\\\.xmgr$\",\n",
      "                \".*\\\\.gnu$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_rmsf\",\n",
      "          \"description\": \"Calculates the Root Mean Square fluctuations (RMSf) of a given cpptraj compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_rmsf tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_exp_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the experimental reference file (required if reference = experimental)\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/experimental.1e5t.pdb\",\n",
      "              \"formats\": null\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed analysis\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.rmsf.first.dat\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.dat$\",\n",
      "                \".*\\\\.agr$\",\n",
      "                \".*\\\\.xmgr$\",\n",
      "                \".*\\\\.gnu$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_rgyr\",\n",
      "          \"description\": \"Computes the radius of gyration (Rgyr) from a given cpptraj compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_rgyr tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_rgyr.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output analysis\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.rgyr.dat\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.dat$\",\n",
      "                \".*\\\\.agr$\",\n",
      "                \".*\\\\.xmgr$\",\n",
      "                \".*\\\\.gnu$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_dry\",\n",
      "          \"description\": \"Dehydrates a given cpptraj compatible trajectory stripping out solvent molecules and ions.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_dry tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_dry.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed trajectory\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.dry.netcdf\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.rst7$\",\n",
      "                \".*\\\\.ncrst$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.sqm$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_strip\",\n",
      "          \"description\": \"Strips a defined set of atoms (mask) from a given cpptraj compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_strip tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_strip.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed trajectory\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.strip.netcdf\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.rst7$\",\n",
      "                \".*\\\\.ncrst$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.sqm$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_snapshot\",\n",
      "          \"description\": \"Extracts a particular snapshot from a given cpptraj compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_snapshot tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_snapshot.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed structure\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.snapshot.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.rst7$\",\n",
      "                \".*\\\\.ncrst$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.sqm$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_slice\",\n",
      "          \"description\": \"Extracts a particular trajectory slice from a given cpptraj compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_slice tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_slice.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed trajectory\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.slice.netcdf\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.rst7$\",\n",
      "                \".*\\\\.ncrst$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.sqm$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_convert\",\n",
      "          \"description\": \"Converts between cpptraj compatible trajectory file formats and/or extracts a selection of atoms or frames.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_convert tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_convert.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed trajectory\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.convert.netcdf\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.rst7$\",\n",
      "                \".*\\\\.ncrst$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.sqm$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_mask\",\n",
      "          \"description\": \"Extracts a selection of atoms from a given cpptraj compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_mask tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_mask.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed trajectory\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.mask.netcdf\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.rst7$\",\n",
      "                \".*\\\\.ncrst$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.sqm$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_image\",\n",
      "          \"description\": \"Corrects periodicity (image) from a given cpptraj trajectory file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_image tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_image.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed trajectory\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.image.netcdf\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.rst7$\",\n",
      "                \".*\\\\.ncrst$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.sqm$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"biobb_chemistry\",\n",
      "      \"tools\": [\n",
      "        {\n",
      "          \"id\": \"acpype_params_ac\",\n",
      "          \"description\": \"Small molecule parameterization for AMBER MD package.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the acpype_params_ac tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_chemistry/master/biobb_chemistry/test/data/config/config_acpype_params_ac.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/acpype/acpype.params.mol2\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.mol2$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_frcmod\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the FRCMOD output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.ac.frcmod\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.frcmod$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_inpcrd\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the INPCRD output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.ac.inpcrd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.inpcrd$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_lib\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the LIB output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.ac.lib\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.lib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_prmtop\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the PRMTOP output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.ac.prmtop\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.prmtop$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"acpype_params_cns\",\n",
      "          \"description\": \"Small molecule parameterization for CNS/XPLOR MD package.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the acpype_params_cns tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_chemistry/master/biobb_chemistry/test/data/config/config_acpype_params_cns.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/acpype/acpype.params.mol2\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.mol2$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_par\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the PAR output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.cns.par\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.par$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_inp\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the INP output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.cns.inp\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.inp$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_top\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the TOP output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.cns.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"acpype_params_gmx\",\n",
      "          \"description\": \"Small molecule parameterization for GROMACS MD package.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the acpype_params_gmx tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_chemistry/master/biobb_chemistry/test/data/config/config_acpype_params_gmx.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/acpype/acpype.params.mol2\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.mol2$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_gro\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GRO output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.gmx.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_itp\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the ITP output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.gmx.itp\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.itp$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_top\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the TOP output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.gmx.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"acpype_params_gmx_opls\",\n",
      "          \"description\": \"Small molecule parameterization for OPLS/AA MD package.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the acpype_params_gmx_opls tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_chemistry/master/biobb_chemistry/test/data/config/config_acpype_params_gmx_opls.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/acpype/acpype.params.mol2\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.mol2$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_itp\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the ITP output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.gmx.opls.itp\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.itp$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_top\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the TOP output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.gmx.opls.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"babel_convert\",\n",
      "          \"description\": \"Small molecule format conversion.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the babel_convert tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_chemistry/master/biobb_chemistry/test/data/config/config_babel_convert.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/babel/babel.smi\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.abinit$\",\n",
      "                \".*\\\\.acesout$\",\n",
      "                \".*\\\\.acr$\",\n",
      "                \".*\\\\.adfout$\",\n",
      "                \".*\\\\.alc$\",\n",
      "                \".*\\\\.aoforce$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.axsf$\",\n",
      "                \".*\\\\.bgf$\",\n",
      "                \".*\\\\.box$\",\n",
      "                \".*\\\\.bs$\",\n",
      "                \".*\\\\.c09out$\",\n",
      "                \".*\\\\.c3d2$\",\n",
      "                \".*\\\\.caccrt$\",\n",
      "                \".*\\\\.can$\",\n",
      "                \".*\\\\.car$\",\n",
      "                \".*\\\\.castep$\",\n",
      "                \".*\\\\.ccc$\",\n",
      "                \".*\\\\.cdjson$\",\n",
      "                \".*\\\\.cdx$\",\n",
      "                \".*\\\\.cdxml$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.ck$\",\n",
      "                \".*\\\\.cml$\",\n",
      "                \".*\\\\.cmlr$\",\n",
      "                \".*\\\\.CONFIG$\",\n",
      "                \".*\\\\.CONTCAR$\",\n",
      "                \".*\\\\.CONTFF$\",\n",
      "                \".*\\\\.crk2d$\",\n",
      "                \".*\\\\.crk3d$\",\n",
      "                \".*\\\\.ct$\",\n",
      "                \".*\\\\.cub$\",\n",
      "                \".*\\\\.cube$\",\n",
      "                \".*\\\\.dallog$\",\n",
      "                \".*\\\\.dalmol$\",\n",
      "                \".*\\\\.dat$\",\n",
      "                \".*\\\\.dmol$\",\n",
      "                \".*\\\\.dx$\",\n",
      "                \".*\\\\.ent$\",\n",
      "                \".*\\\\.exyz$\",\n",
      "                \".*\\\\.fa$\",\n",
      "                \".*\\\\.fasta$\",\n",
      "                \".*\\\\.fch$\",\n",
      "                \".*\\\\.fchk$\",\n",
      "                \".*\\\\.fck$\",\n",
      "                \".*\\\\.feat$\",\n",
      "                \".*\\\\.fhiaims$\",\n",
      "                \".*\\\\.fract$\",\n",
      "                \".*\\\\.fs$\",\n",
      "                \".*\\\\.fsa$\",\n",
      "                \".*\\\\.g03$\",\n",
      "                \".*\\\\.g09$\",\n",
      "                \".*\\\\.g92$\",\n",
      "                \".*\\\\.g94$\",\n",
      "                \".*\\\\.g98$\",\n",
      "                \".*\\\\.gal$\",\n",
      "                \".*\\\\.gam$\",\n",
      "                \".*\\\\.gamess$\",\n",
      "                \".*\\\\.gamin$\",\n",
      "                \".*\\\\.gamout$\",\n",
      "                \".*\\\\.got$\",\n",
      "                \".*\\\\.gpr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.gukin$\",\n",
      "                \".*\\\\.gukout$\",\n",
      "                \".*\\\\.gzmat$\",\n",
      "                \".*\\\\.hin$\",\n",
      "                \".*\\\\.HISTORY$\",\n",
      "                \".*\\\\.inchi$\",\n",
      "                \".*\\\\.inp$\",\n",
      "                \".*\\\\.ins$\",\n",
      "                \".*\\\\.jin$\",\n",
      "                \".*\\\\.jout$\",\n",
      "                \".*\\\\.log$\",\n",
      "                \".*\\\\.lpmd$\",\n",
      "                \".*\\\\.mcdl$\",\n",
      "                \".*\\\\.mcif$\",\n",
      "                \".*\\\\.MDFF$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.ml2$\",\n",
      "                \".*\\\\.mmcif$\",\n",
      "                \".*\\\\.mmd$\",\n",
      "                \".*\\\\.mmod$\",\n",
      "                \".*\\\\.mol$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.mold$\",\n",
      "                \".*\\\\.molden$\",\n",
      "                \".*\\\\.molf$\",\n",
      "                \".*\\\\.moo$\",\n",
      "                \".*\\\\.mop$\",\n",
      "                \".*\\\\.mopcrt$\",\n",
      "                \".*\\\\.mopin$\",\n",
      "                \".*\\\\.mopout$\",\n",
      "                \".*\\\\.mpc$\",\n",
      "                \".*\\\\.mpo$\",\n",
      "                \".*\\\\.mpqc$\",\n",
      "                \".*\\\\.mrv$\",\n",
      "                \".*\\\\.msi$\",\n",
      "                \".*\\\\.nwo$\",\n",
      "                \".*\\\\.orca$\",\n",
      "                \".*\\\\.out$\",\n",
      "                \".*\\\\.outmol$\",\n",
      "                \".*\\\\.output$\",\n",
      "                \".*\\\\.pc$\",\n",
      "                \".*\\\\.pcjson$\",\n",
      "                \".*\\\\.pcm$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.pdbqt$\",\n",
      "                \".*\\\\.png$\",\n",
      "                \".*\\\\.pos$\",\n",
      "                \".*\\\\.POSCAR$\",\n",
      "                \".*\\\\.POSFF$\",\n",
      "                \".*\\\\.pqr$\",\n",
      "                \".*\\\\.pqs$\",\n",
      "                \".*\\\\.prep$\",\n",
      "                \".*\\\\.pwscf$\",\n",
      "                \".*\\\\.qcout$\",\n",
      "                \".*\\\\.res$\",\n",
      "                \".*\\\\.rsmi$\",\n",
      "                \".*\\\\.rxn$\",\n",
      "                \".*\\\\.sd$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.siesta$\",\n",
      "                \".*\\\\.smi$\",\n",
      "                \".*\\\\.smiles$\",\n",
      "                \".*\\\\.smy$\",\n",
      "                \".*\\\\.sy2$\",\n",
      "                \".*\\\\.t41$\",\n",
      "                \".*\\\\.tdd$\",\n",
      "                \".*\\\\.text$\",\n",
      "                \".*\\\\.therm$\",\n",
      "                \".*\\\\.tmol$\",\n",
      "                \".*\\\\.txt$\",\n",
      "                \".*\\\\.txyz$\",\n",
      "                \".*\\\\.unixyz$\",\n",
      "                \".*\\\\.VASP$\",\n",
      "                \".*\\\\.vmol$\",\n",
      "                \".*\\\\.xml$\",\n",
      "                \".*\\\\.xsf$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.xyz$\",\n",
      "                \".*\\\\.yob$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/babel/ref_babel.convert.mol2\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.acesin$\",\n",
      "                \".*\\\\.adf$\",\n",
      "                \".*\\\\.alc$\",\n",
      "                \".*\\\\.ascii$\",\n",
      "                \".*\\\\.bgf$\",\n",
      "                \".*\\\\.box$\",\n",
      "                \".*\\\\.bs$\",\n",
      "                \".*\\\\.c3d1$\",\n",
      "                \".*\\\\.c3d2$\",\n",
      "                \".*\\\\.cac$\",\n",
      "                \".*\\\\.caccrt$\",\n",
      "                \".*\\\\.cache$\",\n",
      "                \".*\\\\.cacint$\",\n",
      "                \".*\\\\.can$\",\n",
      "                \".*\\\\.cdjson$\",\n",
      "                \".*\\\\.cdxml$\",\n",
      "                \".*\\\\.cht$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.ck$\",\n",
      "                \".*\\\\.cml$\",\n",
      "                \".*\\\\.cmlr$\",\n",
      "                \".*\\\\.com$\",\n",
      "                \".*\\\\.confabreport$\",\n",
      "                \".*\\\\.CONFIG$\",\n",
      "                \".*\\\\.CONTCAR$\",\n",
      "                \".*\\\\.CONTFF$\",\n",
      "                \".*\\\\.copy$\",\n",
      "                \".*\\\\.crk2d$\",\n",
      "                \".*\\\\.crk3d$\",\n",
      "                \".*\\\\.csr$\",\n",
      "                \".*\\\\.cssr$\",\n",
      "                \".*\\\\.ct$\",\n",
      "                \".*\\\\.cub$\",\n",
      "                \".*\\\\.cube$\",\n",
      "                \".*\\\\.dalmol$\",\n",
      "                \".*\\\\.dmol$\",\n",
      "                \".*\\\\.dx$\",\n",
      "                \".*\\\\.ent$\",\n",
      "                \".*\\\\.exyz$\",\n",
      "                \".*\\\\.fa$\",\n",
      "                \".*\\\\.fasta$\",\n",
      "                \".*\\\\.feat$\",\n",
      "                \".*\\\\.fh$\",\n",
      "                \".*\\\\.fhiaims$\",\n",
      "                \".*\\\\.fix$\",\n",
      "                \".*\\\\.fps$\",\n",
      "                \".*\\\\.fpt$\",\n",
      "                \".*\\\\.fract$\",\n",
      "                \".*\\\\.fs$\",\n",
      "                \".*\\\\.fsa$\",\n",
      "                \".*\\\\.gamin$\",\n",
      "                \".*\\\\.gau$\",\n",
      "                \".*\\\\.gjc$\",\n",
      "                \".*\\\\.gjf$\",\n",
      "                \".*\\\\.gpr$\",\n",
      "                \".*\\\\.gr96$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.gukin$\",\n",
      "                \".*\\\\.gukout$\",\n",
      "                \".*\\\\.gzmat$\",\n",
      "                \".*\\\\.hin$\",\n",
      "                \".*\\\\.inchi$\",\n",
      "                \".*\\\\.inchikey$\",\n",
      "                \".*\\\\.inp$\",\n",
      "                \".*\\\\.jin$\",\n",
      "                \".*\\\\.k$\",\n",
      "                \".*\\\\.lmpdat$\",\n",
      "                \".*\\\\.lpmd$\",\n",
      "                \".*\\\\.mcdl$\",\n",
      "                \".*\\\\.mcif$\",\n",
      "                \".*\\\\.MDFF$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.ml2$\",\n",
      "                \".*\\\\.mmcif$\",\n",
      "                \".*\\\\.mmd$\",\n",
      "                \".*\\\\.mmod$\",\n",
      "                \".*\\\\.mna$\",\n",
      "                \".*\\\\.mol$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.mold$\",\n",
      "                \".*\\\\.molden$\",\n",
      "                \".*\\\\.molf$\",\n",
      "                \".*\\\\.molreport$\",\n",
      "                \".*\\\\.mop$\",\n",
      "                \".*\\\\.mopcrt$\",\n",
      "                \".*\\\\.mopin$\",\n",
      "                \".*\\\\.mp$\",\n",
      "                \".*\\\\.mpc$\",\n",
      "                \".*\\\\.mpd$\",\n",
      "                \".*\\\\.mpqcin$\",\n",
      "                \".*\\\\.mrv$\",\n",
      "                \".*\\\\.msms$\",\n",
      "                \".*\\\\.nul$\",\n",
      "                \".*\\\\.nw$\",\n",
      "                \".*\\\\.orcainp$\",\n",
      "                \".*\\\\.outmol$\",\n",
      "                \".*\\\\.paint$\",\n",
      "                \".*\\\\.pcjson$\",\n",
      "                \".*\\\\.pcm$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.pdbqt$\",\n",
      "                \".*\\\\.png$\",\n",
      "                \".*\\\\.pointcloud$\",\n",
      "                \".*\\\\.POSCAR$\",\n",
      "                \".*\\\\.POSFF$\",\n",
      "                \".*\\\\.pov$\",\n",
      "                \".*\\\\.pqr$\",\n",
      "                \".*\\\\.pqs$\",\n",
      "                \".*\\\\.qcin$\",\n",
      "                \".*\\\\.report$\",\n",
      "                \".*\\\\.rsmi$\",\n",
      "                \".*\\\\.rxn$\",\n",
      "                \".*\\\\.sd$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.smi$\",\n",
      "                \".*\\\\.smiles$\",\n",
      "                \".*\\\\.stl$\",\n",
      "                \".*\\\\.svg$\",\n",
      "                \".*\\\\.sy2$\",\n",
      "                \".*\\\\.tdd$\",\n",
      "                \".*\\\\.text$\",\n",
      "                \".*\\\\.therm$\",\n",
      "                \".*\\\\.tmol$\",\n",
      "                \".*\\\\.txt$\",\n",
      "                \".*\\\\.txyz$\",\n",
      "                \".*\\\\.unixyz$\",\n",
      "                \".*\\\\.VASP$\",\n",
      "                \".*\\\\.vmol$\",\n",
      "                \".*\\\\.xed$\",\n",
      "                \".*\\\\.xyz$\",\n",
      "                \".*\\\\.yob$\",\n",
      "                \".*\\\\.zin$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"babel_add_hydrogens\",\n",
      "          \"description\": \"Adds hydrogen atoms to small molecules.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the babel_add_hydrogens tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_chemistry/master/biobb_chemistry/test/data/config/config_babel_add_hydrogens.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/babel/babel.no.H.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.abinit$\",\n",
      "                \".*\\\\.acesout$\",\n",
      "                \".*\\\\.acr$\",\n",
      "                \".*\\\\.adfout$\",\n",
      "                \".*\\\\.alc$\",\n",
      "                \".*\\\\.aoforce$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.axsf$\",\n",
      "                \".*\\\\.bgf$\",\n",
      "                \".*\\\\.box$\",\n",
      "                \".*\\\\.bs$\",\n",
      "                \".*\\\\.c09out$\",\n",
      "                \".*\\\\.c3d2$\",\n",
      "                \".*\\\\.caccrt$\",\n",
      "                \".*\\\\.can$\",\n",
      "                \".*\\\\.car$\",\n",
      "                \".*\\\\.castep$\",\n",
      "                \".*\\\\.ccc$\",\n",
      "                \".*\\\\.cdjson$\",\n",
      "                \".*\\\\.cdx$\",\n",
      "                \".*\\\\.cdxml$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.ck$\",\n",
      "                \".*\\\\.cml$\",\n",
      "                \".*\\\\.cmlr$\",\n",
      "                \".*\\\\.CONFIG$\",\n",
      "                \".*\\\\.CONTCAR$\",\n",
      "                \".*\\\\.CONTFF$\",\n",
      "                \".*\\\\.crk2d$\",\n",
      "                \".*\\\\.crk3d$\",\n",
      "                \".*\\\\.ct$\",\n",
      "                \".*\\\\.cub$\",\n",
      "                \".*\\\\.cube$\",\n",
      "                \".*\\\\.dallog$\",\n",
      "                \".*\\\\.dalmol$\",\n",
      "                \".*\\\\.dat$\",\n",
      "                \".*\\\\.dmol$\",\n",
      "                \".*\\\\.dx$\",\n",
      "                \".*\\\\.ent$\",\n",
      "                \".*\\\\.exyz$\",\n",
      "                \".*\\\\.fa$\",\n",
      "                \".*\\\\.fasta$\",\n",
      "                \".*\\\\.fch$\",\n",
      "                \".*\\\\.fchk$\",\n",
      "                \".*\\\\.fck$\",\n",
      "                \".*\\\\.feat$\",\n",
      "                \".*\\\\.fhiaims$\",\n",
      "                \".*\\\\.fract$\",\n",
      "                \".*\\\\.fs$\",\n",
      "                \".*\\\\.fsa$\",\n",
      "                \".*\\\\.g03$\",\n",
      "                \".*\\\\.g09$\",\n",
      "                \".*\\\\.g92$\",\n",
      "                \".*\\\\.g94$\",\n",
      "                \".*\\\\.g98$\",\n",
      "                \".*\\\\.gal$\",\n",
      "                \".*\\\\.gam$\",\n",
      "                \".*\\\\.gamess$\",\n",
      "                \".*\\\\.gamin$\",\n",
      "                \".*\\\\.gamout$\",\n",
      "                \".*\\\\.got$\",\n",
      "                \".*\\\\.gpr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.gukin$\",\n",
      "                \".*\\\\.gukout$\",\n",
      "                \".*\\\\.gzmat$\",\n",
      "                \".*\\\\.hin$\",\n",
      "                \".*\\\\.HISTORY$\",\n",
      "                \".*\\\\.inchi$\",\n",
      "                \".*\\\\.inp$\",\n",
      "                \".*\\\\.ins$\",\n",
      "                \".*\\\\.jin$\",\n",
      "                \".*\\\\.jout$\",\n",
      "                \".*\\\\.log$\",\n",
      "                \".*\\\\.lpmd$\",\n",
      "                \".*\\\\.mcdl$\",\n",
      "                \".*\\\\.mcif$\",\n",
      "                \".*\\\\.MDFF$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.ml2$\",\n",
      "                \".*\\\\.mmcif$\",\n",
      "                \".*\\\\.mmd$\",\n",
      "                \".*\\\\.mmod$\",\n",
      "                \".*\\\\.mol$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.mold$\",\n",
      "                \".*\\\\.molden$\",\n",
      "                \".*\\\\.molf$\",\n",
      "                \".*\\\\.moo$\",\n",
      "                \".*\\\\.mop$\",\n",
      "                \".*\\\\.mopcrt$\",\n",
      "                \".*\\\\.mopin$\",\n",
      "                \".*\\\\.mopout$\",\n",
      "                \".*\\\\.mpc$\",\n",
      "                \".*\\\\.mpo$\",\n",
      "                \".*\\\\.mpqc$\",\n",
      "                \".*\\\\.mrv$\",\n",
      "                \".*\\\\.msi$\",\n",
      "                \".*\\\\.nwo$\",\n",
      "                \".*\\\\.orca$\",\n",
      "                \".*\\\\.out$\",\n",
      "                \".*\\\\.outmol$\",\n",
      "                \".*\\\\.output$\",\n",
      "                \".*\\\\.pc$\",\n",
      "                \".*\\\\.pcjson$\",\n",
      "                \".*\\\\.pcm$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.pdbqt$\",\n",
      "                \".*\\\\.png$\",\n",
      "                \".*\\\\.pos$\",\n",
      "                \".*\\\\.POSCAR$\",\n",
      "                \".*\\\\.POSFF$\",\n",
      "                \".*\\\\.pqr$\",\n",
      "                \".*\\\\.pqs$\",\n",
      "                \".*\\\\.prep$\",\n",
      "                \".*\\\\.pwscf$\",\n",
      "                \".*\\\\.qcout$\",\n",
      "                \".*\\\\.res$\",\n",
      "                \".*\\\\.rsmi$\",\n",
      "                \".*\\\\.rxn$\",\n",
      "                \".*\\\\.sd$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.siesta$\",\n",
      "                \".*\\\\.smi$\",\n",
      "                \".*\\\\.smiles$\",\n",
      "                \".*\\\\.smy$\",\n",
      "                \".*\\\\.sy2$\",\n",
      "                \".*\\\\.t41$\",\n",
      "                \".*\\\\.tdd$\",\n",
      "                \".*\\\\.text$\",\n",
      "                \".*\\\\.therm$\",\n",
      "                \".*\\\\.tmol$\",\n",
      "                \".*\\\\.txt$\",\n",
      "                \".*\\\\.txyz$\",\n",
      "                \".*\\\\.unixyz$\",\n",
      "                \".*\\\\.VASP$\",\n",
      "                \".*\\\\.vmol$\",\n",
      "                \".*\\\\.xml$\",\n",
      "                \".*\\\\.xsf$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.xyz$\",\n",
      "                \".*\\\\.yob$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/babel/ref_babel.hydrogens.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.acesin$\",\n",
      "                \".*\\\\.adf$\",\n",
      "                \".*\\\\.alc$\",\n",
      "                \".*\\\\.ascii$\",\n",
      "                \".*\\\\.bgf$\",\n",
      "                \".*\\\\.box$\",\n",
      "                \".*\\\\.bs$\",\n",
      "                \".*\\\\.c3d1$\",\n",
      "                \".*\\\\.c3d2$\",\n",
      "                \".*\\\\.cac$\",\n",
      "                \".*\\\\.caccrt$\",\n",
      "                \".*\\\\.cache$\",\n",
      "                \".*\\\\.cacint$\",\n",
      "                \".*\\\\.can$\",\n",
      "                \".*\\\\.cdjson$\",\n",
      "                \".*\\\\.cdxml$\",\n",
      "                \".*\\\\.cht$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.ck$\",\n",
      "                \".*\\\\.cml$\",\n",
      "                \".*\\\\.cmlr$\",\n",
      "                \".*\\\\.com$\",\n",
      "                \".*\\\\.confabreport$\",\n",
      "                \".*\\\\.CONFIG$\",\n",
      "                \".*\\\\.CONTCAR$\",\n",
      "                \".*\\\\.CONTFF$\",\n",
      "                \".*\\\\.copy$\",\n",
      "                \".*\\\\.crk2d$\",\n",
      "                \".*\\\\.crk3d$\",\n",
      "                \".*\\\\.csr$\",\n",
      "                \".*\\\\.cssr$\",\n",
      "                \".*\\\\.ct$\",\n",
      "                \".*\\\\.cub$\",\n",
      "                \".*\\\\.cube$\",\n",
      "                \".*\\\\.dalmol$\",\n",
      "                \".*\\\\.dmol$\",\n",
      "                \".*\\\\.dx$\",\n",
      "                \".*\\\\.ent$\",\n",
      "                \".*\\\\.exyz$\",\n",
      "                \".*\\\\.fa$\",\n",
      "                \".*\\\\.fasta$\",\n",
      "                \".*\\\\.feat$\",\n",
      "                \".*\\\\.fh$\",\n",
      "                \".*\\\\.fhiaims$\",\n",
      "                \".*\\\\.fix$\",\n",
      "                \".*\\\\.fps$\",\n",
      "                \".*\\\\.fpt$\",\n",
      "                \".*\\\\.fract$\",\n",
      "                \".*\\\\.fs$\",\n",
      "                \".*\\\\.fsa$\",\n",
      "                \".*\\\\.gamin$\",\n",
      "                \".*\\\\.gau$\",\n",
      "                \".*\\\\.gjc$\",\n",
      "                \".*\\\\.gjf$\",\n",
      "                \".*\\\\.gpr$\",\n",
      "                \".*\\\\.gr96$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.gukin$\",\n",
      "                \".*\\\\.gukout$\",\n",
      "                \".*\\\\.gzmat$\",\n",
      "                \".*\\\\.hin$\",\n",
      "                \".*\\\\.inchi$\",\n",
      "                \".*\\\\.inchikey$\",\n",
      "                \".*\\\\.inp$\",\n",
      "                \".*\\\\.jin$\",\n",
      "                \".*\\\\.k$\",\n",
      "                \".*\\\\.lmpdat$\",\n",
      "                \".*\\\\.lpmd$\",\n",
      "                \".*\\\\.mcdl$\",\n",
      "                \".*\\\\.mcif$\",\n",
      "                \".*\\\\.MDFF$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.ml2$\",\n",
      "                \".*\\\\.mmcif$\",\n",
      "                \".*\\\\.mmd$\",\n",
      "                \".*\\\\.mmod$\",\n",
      "                \".*\\\\.mna$\",\n",
      "                \".*\\\\.mol$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.mold$\",\n",
      "                \".*\\\\.molden$\",\n",
      "                \".*\\\\.molf$\",\n",
      "                \".*\\\\.molreport$\",\n",
      "                \".*\\\\.mop$\",\n",
      "                \".*\\\\.mopcrt$\",\n",
      "                \".*\\\\.mopin$\",\n",
      "                \".*\\\\.mp$\",\n",
      "                \".*\\\\.mpc$\",\n",
      "                \".*\\\\.mpd$\",\n",
      "                \".*\\\\.mpqcin$\",\n",
      "                \".*\\\\.mrv$\",\n",
      "                \".*\\\\.msms$\",\n",
      "                \".*\\\\.nul$\",\n",
      "                \".*\\\\.nw$\",\n",
      "                \".*\\\\.orcainp$\",\n",
      "                \".*\\\\.outmol$\",\n",
      "                \".*\\\\.paint$\",\n",
      "                \".*\\\\.pcjson$\",\n",
      "                \".*\\\\.pcm$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.pdbqt$\",\n",
      "                \".*\\\\.png$\",\n",
      "                \".*\\\\.pointcloud$\",\n",
      "                \".*\\\\.POSCAR$\",\n",
      "                \".*\\\\.POSFF$\",\n",
      "                \".*\\\\.pov$\",\n",
      "                \".*\\\\.pqr$\",\n",
      "                \".*\\\\.pqs$\",\n",
      "                \".*\\\\.qcin$\",\n",
      "                \".*\\\\.report$\",\n",
      "                \".*\\\\.rsmi$\",\n",
      "                \".*\\\\.rxn$\",\n",
      "                \".*\\\\.sd$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.smi$\",\n",
      "                \".*\\\\.smiles$\",\n",
      "                \".*\\\\.stl$\",\n",
      "                \".*\\\\.svg$\",\n",
      "                \".*\\\\.sy2$\",\n",
      "                \".*\\\\.tdd$\",\n",
      "                \".*\\\\.text$\",\n",
      "                \".*\\\\.therm$\",\n",
      "                \".*\\\\.tmol$\",\n",
      "                \".*\\\\.txt$\",\n",
      "                \".*\\\\.txyz$\",\n",
      "                \".*\\\\.unixyz$\",\n",
      "                \".*\\\\.VASP$\",\n",
      "                \".*\\\\.vmol$\",\n",
      "                \".*\\\\.xed$\",\n",
      "                \".*\\\\.xyz$\",\n",
      "                \".*\\\\.yob$\",\n",
      "                \".*\\\\.zin$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"babel_remove_hydrogens\",\n",
      "          \"description\": \"Removes hydrogen atoms to small molecules.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the babel_remove_hydrogens tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_chemistry/master/biobb_chemistry/test/data/config/config_babel_remove_hydrogens.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/babel/babel.H.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.abinit$\",\n",
      "                \".*\\\\.acesout$\",\n",
      "                \".*\\\\.acr$\",\n",
      "                \".*\\\\.adfout$\",\n",
      "                \".*\\\\.alc$\",\n",
      "                \".*\\\\.aoforce$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.axsf$\",\n",
      "                \".*\\\\.bgf$\",\n",
      "                \".*\\\\.box$\",\n",
      "                \".*\\\\.bs$\",\n",
      "                \".*\\\\.c09out$\",\n",
      "                \".*\\\\.c3d2$\",\n",
      "                \".*\\\\.caccrt$\",\n",
      "                \".*\\\\.can$\",\n",
      "                \".*\\\\.car$\",\n",
      "                \".*\\\\.castep$\",\n",
      "                \".*\\\\.ccc$\",\n",
      "                \".*\\\\.cdjson$\",\n",
      "                \".*\\\\.cdx$\",\n",
      "                \".*\\\\.cdxml$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.ck$\",\n",
      "                \".*\\\\.cml$\",\n",
      "                \".*\\\\.cmlr$\",\n",
      "                \".*\\\\.CONFIG$\",\n",
      "                \".*\\\\.CONTCAR$\",\n",
      "                \".*\\\\.CONTFF$\",\n",
      "                \".*\\\\.crk2d$\",\n",
      "                \".*\\\\.crk3d$\",\n",
      "                \".*\\\\.ct$\",\n",
      "                \".*\\\\.cub$\",\n",
      "                \".*\\\\.cube$\",\n",
      "                \".*\\\\.dallog$\",\n",
      "                \".*\\\\.dalmol$\",\n",
      "                \".*\\\\.dat$\",\n",
      "                \".*\\\\.dmol$\",\n",
      "                \".*\\\\.dx$\",\n",
      "                \".*\\\\.ent$\",\n",
      "                \".*\\\\.exyz$\",\n",
      "                \".*\\\\.fa$\",\n",
      "                \".*\\\\.fasta$\",\n",
      "                \".*\\\\.fch$\",\n",
      "                \".*\\\\.fchk$\",\n",
      "                \".*\\\\.fck$\",\n",
      "                \".*\\\\.feat$\",\n",
      "                \".*\\\\.fhiaims$\",\n",
      "                \".*\\\\.fract$\",\n",
      "                \".*\\\\.fs$\",\n",
      "                \".*\\\\.fsa$\",\n",
      "                \".*\\\\.g03$\",\n",
      "                \".*\\\\.g09$\",\n",
      "                \".*\\\\.g92$\",\n",
      "                \".*\\\\.g94$\",\n",
      "                \".*\\\\.g98$\",\n",
      "                \".*\\\\.gal$\",\n",
      "                \".*\\\\.gam$\",\n",
      "                \".*\\\\.gamess$\",\n",
      "                \".*\\\\.gamin$\",\n",
      "                \".*\\\\.gamout$\",\n",
      "                \".*\\\\.got$\",\n",
      "                \".*\\\\.gpr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.gukin$\",\n",
      "                \".*\\\\.gukout$\",\n",
      "                \".*\\\\.gzmat$\",\n",
      "                \".*\\\\.hin$\",\n",
      "                \".*\\\\.HISTORY$\",\n",
      "                \".*\\\\.inchi$\",\n",
      "                \".*\\\\.inp$\",\n",
      "                \".*\\\\.ins$\",\n",
      "                \".*\\\\.jin$\",\n",
      "                \".*\\\\.jout$\",\n",
      "                \".*\\\\.log$\",\n",
      "                \".*\\\\.lpmd$\",\n",
      "                \".*\\\\.mcdl$\",\n",
      "                \".*\\\\.mcif$\",\n",
      "                \".*\\\\.MDFF$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.ml2$\",\n",
      "                \".*\\\\.mmcif$\",\n",
      "                \".*\\\\.mmd$\",\n",
      "                \".*\\\\.mmod$\",\n",
      "                \".*\\\\.mol$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.mold$\",\n",
      "                \".*\\\\.molden$\",\n",
      "                \".*\\\\.molf$\",\n",
      "                \".*\\\\.moo$\",\n",
      "                \".*\\\\.mop$\",\n",
      "                \".*\\\\.mopcrt$\",\n",
      "                \".*\\\\.mopin$\",\n",
      "                \".*\\\\.mopout$\",\n",
      "                \".*\\\\.mpc$\",\n",
      "                \".*\\\\.mpo$\",\n",
      "                \".*\\\\.mpqc$\",\n",
      "                \".*\\\\.mrv$\",\n",
      "                \".*\\\\.msi$\",\n",
      "                \".*\\\\.nwo$\",\n",
      "                \".*\\\\.orca$\",\n",
      "                \".*\\\\.out$\",\n",
      "                \".*\\\\.outmol$\",\n",
      "                \".*\\\\.output$\",\n",
      "                \".*\\\\.pc$\",\n",
      "                \".*\\\\.pcjson$\",\n",
      "                \".*\\\\.pcm$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.pdbqt$\",\n",
      "                \".*\\\\.png$\",\n",
      "                \".*\\\\.pos$\",\n",
      "                \".*\\\\.POSCAR$\",\n",
      "                \".*\\\\.POSFF$\",\n",
      "                \".*\\\\.pqr$\",\n",
      "                \".*\\\\.pqs$\",\n",
      "                \".*\\\\.prep$\",\n",
      "                \".*\\\\.pwscf$\",\n",
      "                \".*\\\\.qcout$\",\n",
      "                \".*\\\\.res$\",\n",
      "                \".*\\\\.rsmi$\",\n",
      "                \".*\\\\.rxn$\",\n",
      "                \".*\\\\.sd$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.siesta$\",\n",
      "                \".*\\\\.smi$\",\n",
      "                \".*\\\\.smiles$\",\n",
      "                \".*\\\\.smy$\",\n",
      "                \".*\\\\.sy2$\",\n",
      "                \".*\\\\.t41$\",\n",
      "                \".*\\\\.tdd$\",\n",
      "                \".*\\\\.text$\",\n",
      "                \".*\\\\.therm$\",\n",
      "                \".*\\\\.tmol$\",\n",
      "                \".*\\\\.txt$\",\n",
      "                \".*\\\\.txyz$\",\n",
      "                \".*\\\\.unixyz$\",\n",
      "                \".*\\\\.VASP$\",\n",
      "                \".*\\\\.vmol$\",\n",
      "                \".*\\\\.xml$\",\n",
      "                \".*\\\\.xsf$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.xyz$\",\n",
      "                \".*\\\\.yob$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/babel/ref_babel.nohydrogens.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.acesin$\",\n",
      "                \".*\\\\.adf$\",\n",
      "                \".*\\\\.alc$\",\n",
      "                \".*\\\\.ascii$\",\n",
      "                \".*\\\\.bgf$\",\n",
      "                \".*\\\\.box$\",\n",
      "                \".*\\\\.bs$\",\n",
      "                \".*\\\\.c3d1$\",\n",
      "                \".*\\\\.c3d2$\",\n",
      "                \".*\\\\.cac$\",\n",
      "                \".*\\\\.caccrt$\",\n",
      "                \".*\\\\.cache$\",\n",
      "                \".*\\\\.cacint$\",\n",
      "                \".*\\\\.can$\",\n",
      "                \".*\\\\.cdjson$\",\n",
      "                \".*\\\\.cdxml$\",\n",
      "                \".*\\\\.cht$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.ck$\",\n",
      "                \".*\\\\.cml$\",\n",
      "                \".*\\\\.cmlr$\",\n",
      "                \".*\\\\.com$\",\n",
      "                \".*\\\\.confabreport$\",\n",
      "                \".*\\\\.CONFIG$\",\n",
      "                \".*\\\\.CONTCAR$\",\n",
      "                \".*\\\\.CONTFF$\",\n",
      "                \".*\\\\.copy$\",\n",
      "                \".*\\\\.crk2d$\",\n",
      "                \".*\\\\.crk3d$\",\n",
      "                \".*\\\\.csr$\",\n",
      "                \".*\\\\.cssr$\",\n",
      "                \".*\\\\.ct$\",\n",
      "                \".*\\\\.cub$\",\n",
      "                \".*\\\\.cube$\",\n",
      "                \".*\\\\.dalmol$\",\n",
      "                \".*\\\\.dmol$\",\n",
      "                \".*\\\\.dx$\",\n",
      "                \".*\\\\.ent$\",\n",
      "                \".*\\\\.exyz$\",\n",
      "                \".*\\\\.fa$\",\n",
      "                \".*\\\\.fasta$\",\n",
      "                \".*\\\\.feat$\",\n",
      "                \".*\\\\.fh$\",\n",
      "                \".*\\\\.fhiaims$\",\n",
      "                \".*\\\\.fix$\",\n",
      "                \".*\\\\.fps$\",\n",
      "                \".*\\\\.fpt$\",\n",
      "                \".*\\\\.fract$\",\n",
      "                \".*\\\\.fs$\",\n",
      "                \".*\\\\.fsa$\",\n",
      "                \".*\\\\.gamin$\",\n",
      "                \".*\\\\.gau$\",\n",
      "                \".*\\\\.gjc$\",\n",
      "                \".*\\\\.gjf$\",\n",
      "                \".*\\\\.gpr$\",\n",
      "                \".*\\\\.gr96$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.gukin$\",\n",
      "                \".*\\\\.gukout$\",\n",
      "                \".*\\\\.gzmat$\",\n",
      "                \".*\\\\.hin$\",\n",
      "                \".*\\\\.inchi$\",\n",
      "                \".*\\\\.inchikey$\",\n",
      "                \".*\\\\.inp$\",\n",
      "                \".*\\\\.jin$\",\n",
      "                \".*\\\\.k$\",\n",
      "                \".*\\\\.lmpdat$\",\n",
      "                \".*\\\\.lpmd$\",\n",
      "                \".*\\\\.mcdl$\",\n",
      "                \".*\\\\.mcif$\",\n",
      "                \".*\\\\.MDFF$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.ml2$\",\n",
      "                \".*\\\\.mmcif$\",\n",
      "                \".*\\\\.mmd$\",\n",
      "                \".*\\\\.mmod$\",\n",
      "                \".*\\\\.mna$\",\n",
      "                \".*\\\\.mol$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.mold$\",\n",
      "                \".*\\\\.molden$\",\n",
      "                \".*\\\\.molf$\",\n",
      "                \".*\\\\.molreport$\",\n",
      "                \".*\\\\.mop$\",\n",
      "                \".*\\\\.mopcrt$\",\n",
      "                \".*\\\\.mopin$\",\n",
      "                \".*\\\\.mp$\",\n",
      "                \".*\\\\.mpc$\",\n",
      "                \".*\\\\.mpd$\",\n",
      "                \".*\\\\.mpqcin$\",\n",
      "                \".*\\\\.mrv$\",\n",
      "                \".*\\\\.msms$\",\n",
      "                \".*\\\\.nul$\",\n",
      "                \".*\\\\.nw$\",\n",
      "                \".*\\\\.orcainp$\",\n",
      "                \".*\\\\.outmol$\",\n",
      "                \".*\\\\.paint$\",\n",
      "                \".*\\\\.pcjson$\",\n",
      "                \".*\\\\.pcm$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.pdbqt$\",\n",
      "                \".*\\\\.png$\",\n",
      "                \".*\\\\.pointcloud$\",\n",
      "                \".*\\\\.POSCAR$\",\n",
      "                \".*\\\\.POSFF$\",\n",
      "                \".*\\\\.pov$\",\n",
      "                \".*\\\\.pqr$\",\n",
      "                \".*\\\\.pqs$\",\n",
      "                \".*\\\\.qcin$\",\n",
      "                \".*\\\\.report$\",\n",
      "                \".*\\\\.rsmi$\",\n",
      "                \".*\\\\.rxn$\",\n",
      "                \".*\\\\.sd$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.smi$\",\n",
      "                \".*\\\\.smiles$\",\n",
      "                \".*\\\\.stl$\",\n",
      "                \".*\\\\.svg$\",\n",
      "                \".*\\\\.sy2$\",\n",
      "                \".*\\\\.tdd$\",\n",
      "                \".*\\\\.text$\",\n",
      "                \".*\\\\.therm$\",\n",
      "                \".*\\\\.tmol$\",\n",
      "                \".*\\\\.txt$\",\n",
      "                \".*\\\\.txyz$\",\n",
      "                \".*\\\\.unixyz$\",\n",
      "                \".*\\\\.VASP$\",\n",
      "                \".*\\\\.vmol$\",\n",
      "                \".*\\\\.xed$\",\n",
      "                \".*\\\\.xyz$\",\n",
      "                \".*\\\\.yob$\",\n",
      "                \".*\\\\.zin$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"babel_minimize\",\n",
      "          \"description\": \"Energetically minimize small molecules.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the babel_minimize tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_chemistry/master/biobb_chemistry/test/data/config/config_babel_minimize.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/babel/babel.minimize.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/babel/ref_babel.minimize.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"reduce_add_hydrogens\",\n",
      "          \"description\": \"Adds hydrogen atoms to small molecules.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the reduce_add_hydrogens tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_chemistry/master/biobb_chemistry/test/data/config/config_reduce_add_hydrogens.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/ambertools/reduce.no.H.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/ambertools/ref_reduce.add.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"reduce_remove_hydrogens\",\n",
      "          \"description\": \"Removes hydrogen atoms to small molecules.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the reduce_remove_hydrogens tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/ambertools/reduce.H.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/ambertools/ref_reduce.remove.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"biobb_io\",\n",
      "      \"tools\": [\n",
      "        {\n",
      "          \"id\": \"ligand\",\n",
      "          \"description\": \"Downloads a ligand file from the MMB REST API.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the ligand tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_io/master/biobb_io/test/data/config/config_ligand.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output PDB ligand file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_io/raw/master/biobb_io/test/reference/api/ligand_12d.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"pdb\",\n",
      "          \"description\": \"Downloads a PDB file from the RCSB or MMB REST APIs.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the pdb tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output PDB file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_io/raw/master/biobb_io/test/reference/api/pdb_1ubq.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"pdb_variants\",\n",
      "          \"description\": \"Creates a text file containing a list of all the variants mapped to a RSCB PDB code from the corresponding UNIPROT entries.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the pdb_variants tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_mutations_list_txt\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the TXT file containing an ASCII comma separated values of the mutations\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.txt$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"pdb_cluster_zip\",\n",
      "          \"description\": \"Creates a zip file containing all the PDB files in the given sequence similarity cluster percentage of the given RSCB PDB code.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the pdb_cluster_zip tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_pdb_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the ZIP or PDB file containing the output PDB files\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_io/raw/master/biobb_io/test/reference/api/reference_output_pdb_zip_path.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"biobb_md\",\n",
      "      \"tools\": [\n",
      "        {\n",
      "          \"id\": \"pdb2gmx\",\n",
      "          \"description\": \"Creates a compressed (ZIP) GROMACS topology (TOP and ITP files) from a given PDB file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the pdb2gmx tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_md/master/biobb_md/test/data/config/config_pdb2gmx.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input PDB file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/egfr.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output GRO file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_pdb2gmx.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the output TOP topology in zip format\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_pdb2gmx.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"editconf\",\n",
      "          \"description\": \"Creates a GROMACS structure file (GRO) adding the information of the solvent box to the input structure file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the editconf tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input GRO file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/editconf.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output GRO file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_editconf.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"genion\",\n",
      "          \"description\": \"Creates a new compressed GROMACS topology adding ions until reaching the desired concentration to the input compressed GROMACS topology. \",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the genion tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_tpr_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input portable run input TPR file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/genion.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.tpr$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure GRO file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_genion.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the input TOP topology in zip format\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/genion.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the output topology TOP and ITP files zipball\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_genion.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"genrestr\",\n",
      "          \"description\": \"Creates a new GROMACS compressed topology applying the indicated force restrains to the given input compressed topology.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the genrestr tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure PDB, GRO or TPR format\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/genrestr.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.tpr$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_ndx_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input GROMACS index file, NDX format\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/genrestr.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_itp_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the output ITP topology file with restrains\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_genrestr.itp\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.itp$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"grompp\",\n",
      "          \"description\": \"Creates a GROMACS portable binary run input file (TPR) applying the desired properties from the input compressed GROMACS topology.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the grompp tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_md/master/biobb_md/test/data/config/config_grompp.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input GROMACS structure GRO file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/grompp.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the input GROMACS topology TOP and ITP files in zip format\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/grompp.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_tpr_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output portable binary run file TPR\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_grompp.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.tpr$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_cpt_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the input GROMACS checkpoint file CPT\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.cpt$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_ndx_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the input GROMACS index files NDX\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"mdrun\",\n",
      "          \"description\": \"Performs molecular dynamics simulations from an input GROMACS TPR file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the mdrun tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_tpr_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the portable binary run input file TPR\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/mdrun.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.tpr$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_trr_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS uncompressed raw trajectory file TRR\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_mdrun.trr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.trr$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output GROMACS structure GRO file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_mdrun.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_edr_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output GROMACS portable energy file EDR\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_mdrun.edr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.edr$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_log_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output GROMACS trajectory log file LOG\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.log$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_xtc_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the GROMACS compressed trajectory file XTC\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpt_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the output GROMACS checkpoint file CPT\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.cpt$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_dhdl_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the output dhdl\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xvg$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"make_ndx\",\n",
      "          \"description\": \"Creates a GROMACS index file (NDX) from an input selection and an input GROMACS structure file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the make_ndx tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input GRO/PDB/TPR file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/make_ndx.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tpr$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_ndx_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output index NDX file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_make_ndx.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_ndx_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the input index NDX file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"select\",\n",
      "          \"description\": \"Creates a GROMACS index file (NDX) from an input selection and an input GROMACS structure file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the select tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_md/master/biobb_md/test/data/config/config_select.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input GRO/PDB/TPR file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/make_ndx.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tpr$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_ndx_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output index NDX file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_select.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_ndx_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the input index NDX file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"solvate\",\n",
      "          \"description\": \"Creates a new compressed GROMACS topology file adding solvent molecules to a given input compressed GROMACS topology file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the solvate tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_solute_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input GRO file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/solvate.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output GRO file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_solvate.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the input TOP topology in zip format\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/solvate.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the output topology in zip format\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_solvate.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"ndx2resttop\",\n",
      "          \"description\": \"Creates a new GROMACS compressed topology applying the force restrains to the input groups in the input index file to the given input compressed topology.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the ndx2resttop tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_md/master/biobb_md/test/data/config/config_ndx2resttop.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_ndx_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input NDX index file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs_extra/ndx2resttop.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the input TOP topology in zip format\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs_extra/ndx2resttop.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the output TOP topology in zip format\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs_extra/ref_ndx2resttop.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"append_ligand\",\n",
      "          \"description\": \"Takes a ligand ITP file and inserts it in a topology.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the append_ligand tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the input topology TOP and ITP files zipball\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs_extra/ndx2resttop.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_itp_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the ligand ITP file to be inserted in the topology\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs_extra/pep_ligand.itp\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.itp$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path/Name the output topology TOP and ITP files zipball\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs_extra/ref_appendligand.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"biobb_model\",\n",
      "      \"tools\": [\n",
      "        {\n",
      "          \"id\": \"fix_side_chain\",\n",
      "          \"description\": \"Reconstructs the missing side chains and heavy atoms of the given PDB file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the fix_side_chain tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input PDB file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_model/raw/master/biobb_model/test/data/model/2ki5.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output PDB file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_model/raw/master/biobb_model/test/reference/model/output_pdb_path.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"mutate\",\n",
      "          \"description\": \"Creates a new PDB file performing the mutations given in a list of amino acid mutations to the input PDB file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the mutate tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_model/master/biobb_model/test/data/config/config_mutate.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input PDB file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_model/raw/master/biobb_model/test/data/model/2ki5.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output PDB file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_model/raw/master/biobb_model/test/reference/model/output_mutated_pdb_path.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"biobb_pmx\",\n",
      "      \"tools\": [\n",
      "        {\n",
      "          \"id\": \"mutate\",\n",
      "          \"description\": \"pmx tool to insert mutated residues in structure files for free energy simulations\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the mutate tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_pmx/master/biobb_pmx/test/data/config/config_mutate.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_pmx/raw/master/biobb_pmx/test/data/pmx/frame99.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output structure file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_pmx/raw/master/biobb_pmx/test/reference/pmx/ref_output_structure.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_b_structure_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the mutated input structure file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"gentop\",\n",
      "          \"description\": \"pmx tool to generate hybrid GROMACS topologies: adding a B state to an .itp or .top file for a hybrid residue\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the gentop tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_pmx/master/biobb_pmx/test/data/config/config_gentop.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the input GROMACS topology TOP and ITP files in zip format\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_pmx/raw/master/biobb_pmx/test/data/pmx/topology.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the output TOP topology in zip format\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_pmx/raw/master/biobb_pmx/test/reference/pmx/ref_output_topology.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"analyse\",\n",
      "          \"description\": \"pmx tool to calculate free energies from fast growth thermodynamic integration simulations.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the analyse tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_pmx/master/biobb_pmx/test/data/config/config_analyse.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_a_xvg_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the zip file containing the dgdl\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_pmx/raw/master/biobb_pmx/test/data/pmx/xvg_A.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_b_xvg_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the zip file containing the dgdl\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_pmx/raw/master/biobb_pmx/test/data/pmx/xvg_B.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_result_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the TXT results file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_pmx/raw/master/biobb_pmx/test/reference/pmx/ref_result.txt\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.txt$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_work_plot_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the PNG plot results file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.png$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"biobb_structure_utils\",\n",
      "      \"tools\": [\n",
      "        {\n",
      "          \"id\": \"cat_pdb\",\n",
      "          \"description\": \"Class to concat two PDB structures in a single PDB file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cat_pdb tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure1\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input structure 1 file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/cat_protein.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure2\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input structure 2 file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/cat_ligand.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output protein file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/ref_cat_pdb.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"extract_atoms\",\n",
      "          \"description\": \"Class to extract atoms from a 3D structure.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the extract_atoms tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_structure_utils/master/biobb_structure_utils/test/data/config/config_extract_atoms.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input structure file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/2vgb.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output structure file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/OE2_atoms.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"extract_chain\",\n",
      "          \"description\": \"Class to extract a chain from a 3D structure.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the extract_chain tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_structure_utils/master/biobb_structure_utils/test/data/config/config_extract_chain.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input structure file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/extract_chain.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output structure file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/ref_extract_chain.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"extract_heteroatoms\",\n",
      "          \"description\": \"Class to extract hetero-atoms from a 3D structure.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the extract_heteroatoms tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_structure_utils/master/biobb_structure_utils/test/data/config/config_extract_heteroatoms.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input structure file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/extract_heteroatom.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_heteroatom_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output heteroatom file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/ref_extract_heteroatom.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"extract_model\",\n",
      "          \"description\": \"Class to extract a model from a 3D structure.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the extract_model tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_structure_utils/master/biobb_structure_utils/test/data/config/config_extract_model.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input structure file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/extract_model.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output structure file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/ref_extract_model.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"extract_protein\",\n",
      "          \"description\": \"Class to extract a protein from a 3D structure.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the extract_protein tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input structure file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/extract_protein.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_protein_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output protein file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/ref_extract_protein.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"remove_ligand\",\n",
      "          \"description\": \"Class to remove the selected ligand atoms from a 3D structure.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the remove_ligand tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_structure_utils/master/biobb_structure_utils/test/data/config/config_remove_ligand.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input structure file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/WT_aq4_md_1.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output structure file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/WT_apo_md_1.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"remove_pdb_water\",\n",
      "          \"description\": \"Class to remove water molecules from PDB 3D structures.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the remove_pdb_water tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input PDB file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/WT_aq4_md_WAT.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output PDB file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/WT_apo_no_wat.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"renumber_structure\",\n",
      "          \"description\": \"Class to renumber atomic indexes from a 3D structure.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the renumber_structure tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input structure file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/cl3.noH.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output structure file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/renum_cl3_noH.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_mapping_json_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output mapping json file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/cl3_output_mapping_json_path.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"sort_gro_residues\",\n",
      "          \"description\": \"Class to sort the selected residues from a GRO 3D structure.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the sort_gro_residues tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_structure_utils/master/biobb_structure_utils/test/data/config/config_sort_gro_residues.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input GRO file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/WT_aq4_md_1.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output sorted GRO file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/WT_aq4_md_sorted.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    }\n",
      "  ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "url = apiURL + 'launch'\n",
    "response = get_data(url)\n",
    "\n",
    "print(json.dumps(response.json, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"list_tools_ex\"></a>\n",
    "#### List of tools from a specific package\n",
    "\n",
    "For more information about this endpoint, please visit the [BioBB REST API Documentation section](http://mmb.irbbarcelona.org/biobb-api/rest#/List%20of%20Services/getToolsList).\n",
    "\n",
    "##### Endpoint"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**GET** `http://mmb.irbbarcelona.org/biobb-api/rest/v1/launch/{package}`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"id\": \"biobb_analysis\",\n",
      "  \"tools\": [\n",
      "    {\n",
      "      \"id\": \"gmx_cluster\",\n",
      "      \"description\": \"Creates cluster structures from a given GROMACS compatible trajectory\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the gmx_cluster tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_cluster.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_structure_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.tpr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.brk$\",\n",
      "            \".*\\\\.ent$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the GROMACS trajectory file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.cpt$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_index_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the GROMACS index file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.ndx$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_pdb_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output cluster file\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_cluster.pdb\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.cpt$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"gmx_rms\",\n",
      "      \"description\": \"Performs a Root Mean Square deviation (RMSd) analysis from a given GROMACS compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the gmx_rms tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_rms.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_structure_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.tpr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.brk$\",\n",
      "            \".*\\\\.ent$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the GROMACS trajectory file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.cpt$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_index_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the GROMACS index file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.ndx$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_xvg_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the XVG output file\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_rms.xvg\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xvg$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"gmx_rgyr\",\n",
      "      \"description\": \"Computes the radius of gyration (Rgyr) of a molecule about the x-, y- and z-axes, as a function of time, from a given GROMACS compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the gmx_rgyr tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_rgyr.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_structure_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.tpr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.brk$\",\n",
      "            \".*\\\\.ent$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the GROMACS trajectory file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.cpt$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_index_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the GROMACS index file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.ndx$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_xvg_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the XVG output file\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_rgyr.xvg\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xvg$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"gmx_energy\",\n",
      "      \"description\": \"Extracts energy components from a given GROMACS energy file.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the gmx_energy tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_energy.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_energy_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input EDR file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/energy.edr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.edr$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_xvg_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the XVG output file\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_energy.xvg\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xvg$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"gmx_image\",\n",
      "      \"description\": \"Corrects periodicity (image) from a given GROMACS compatible trajectory file.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the gmx_image tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_image.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the GROMACS trajectory file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.cpt$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the GROMACS input topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.tpr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.brk$\",\n",
      "            \".*\\\\.ent$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_index_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the GROMACS index file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.ndx$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output file\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_image.xtc\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"gmx_trjconv_str\",\n",
      "      \"description\": \"Converts between GROMACS compatible structure file formats and/or extracts a selection of atoms.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the gmx_trjconv_str tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_trjconv_str.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_structure_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.cpt$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the GROMACS input topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.tpr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.brk$\",\n",
      "            \".*\\\\.ent$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_index_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the GROMACS index file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.ndx$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_str_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output file\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_trjconv.str.pdb\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"gmx_trjconv_str_ens\",\n",
      "      \"description\": \"Extracts an ensemble of frames containing a selection of atoms from GROMACS compatible trajectory files.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the gmx_trjconv_str_ens tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_trjconv_str_ens.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the GROMACS trajectory file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.cpt$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the GROMACS input topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.tpr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.brk$\",\n",
      "            \".*\\\\.ent$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_index_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the GROMACS index file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.ndx$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_str_ens_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output file\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_trjconv.str.ens.zip\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"gmx_trjconv_trj\",\n",
      "      \"description\": \"Converts between GROMACS compatible trajectory file formats and/or extracts a selection of atoms.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the gmx_trjconv_trj tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_trjconv_trj.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the GROMACS trajectory file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.cpt$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_index_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the GROMACS index file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.ndx$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output file\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_trjconv.trj.xtc\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_average\",\n",
      "      \"description\": \"Calculates a structure average of a given cpptraj compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_average tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_average.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed structure\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.average.pdb\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.rst7$\",\n",
      "            \".*\\\\.ncrst$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.sqm$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_bfactor\",\n",
      "      \"description\": \"Calculates the Bfactor fluctuations of a given cpptraj compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_bfactor tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": null,\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_exp_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the experimental reference file (required if reference = experimental)\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/experimental.1e5t.pdb\",\n",
      "          \"formats\": null\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed analysis\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.bfactor.first.dat\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.dat$\",\n",
      "            \".*\\\\.agr$\",\n",
      "            \".*\\\\.xmgr$\",\n",
      "            \".*\\\\.gnu$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_rms\",\n",
      "      \"description\": \"Calculates the Root Mean Square deviation (RMSd) of a given cpptraj compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_rms tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": null,\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_exp_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the experimental reference file (required if reference = experimental)\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/experimental.1e5t.pdb\",\n",
      "          \"formats\": null\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed analysis\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.rms.first.dat\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.dat$\",\n",
      "            \".*\\\\.agr$\",\n",
      "            \".*\\\\.xmgr$\",\n",
      "            \".*\\\\.gnu$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_rmsf\",\n",
      "      \"description\": \"Calculates the Root Mean Square fluctuations (RMSf) of a given cpptraj compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_rmsf tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": null,\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_exp_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the experimental reference file (required if reference = experimental)\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/experimental.1e5t.pdb\",\n",
      "          \"formats\": null\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed analysis\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.rmsf.first.dat\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.dat$\",\n",
      "            \".*\\\\.agr$\",\n",
      "            \".*\\\\.xmgr$\",\n",
      "            \".*\\\\.gnu$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_rgyr\",\n",
      "      \"description\": \"Computes the radius of gyration (Rgyr) from a given cpptraj compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_rgyr tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_rgyr.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output analysis\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.rgyr.dat\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.dat$\",\n",
      "            \".*\\\\.agr$\",\n",
      "            \".*\\\\.xmgr$\",\n",
      "            \".*\\\\.gnu$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_dry\",\n",
      "      \"description\": \"Dehydrates a given cpptraj compatible trajectory stripping out solvent molecules and ions.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_dry tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_dry.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed trajectory\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.dry.netcdf\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.rst7$\",\n",
      "            \".*\\\\.ncrst$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.sqm$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_strip\",\n",
      "      \"description\": \"Strips a defined set of atoms (mask) from a given cpptraj compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_strip tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_strip.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed trajectory\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.strip.netcdf\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.rst7$\",\n",
      "            \".*\\\\.ncrst$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.sqm$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_snapshot\",\n",
      "      \"description\": \"Extracts a particular snapshot from a given cpptraj compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_snapshot tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_snapshot.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed structure\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.snapshot.pdb\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.rst7$\",\n",
      "            \".*\\\\.ncrst$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.sqm$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_slice\",\n",
      "      \"description\": \"Extracts a particular trajectory slice from a given cpptraj compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_slice tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_slice.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed trajectory\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.slice.netcdf\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.rst7$\",\n",
      "            \".*\\\\.ncrst$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.sqm$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_convert\",\n",
      "      \"description\": \"Converts between cpptraj compatible trajectory file formats and/or extracts a selection of atoms or frames.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_convert tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_convert.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed trajectory\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.convert.netcdf\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.rst7$\",\n",
      "            \".*\\\\.ncrst$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.sqm$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_mask\",\n",
      "      \"description\": \"Extracts a selection of atoms from a given cpptraj compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_mask tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_mask.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed trajectory\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.mask.netcdf\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.rst7$\",\n",
      "            \".*\\\\.ncrst$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.sqm$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_image\",\n",
      "      \"description\": \"Corrects periodicity (image) from a given cpptraj trajectory file.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_image tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_image.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed trajectory\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.image.netcdf\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.rst7$\",\n",
      "            \".*\\\\.ncrst$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.sqm$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    }\n",
      "  ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "package = 'biobb_analysis'\n",
    "url = apiURL + 'launch/' + package\n",
    "response = get_data(url)\n",
    "\n",
    "print(json.dumps(response.json, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"tools_prop_ex\"></a>\n",
    "#### Tool's properties\n",
    "\n",
    "For more information about this endpoint, please visit the [BioBB REST API Documentation section](http://mmb.irbbarcelona.org/biobb-api/rest#/Launch%20Tool/getLaunchTool).\n",
    "\n",
    "##### Endpoint"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**GET** `http://mmb.irbbarcelona.org/biobb-api/rest/v1/launch/{package}/{tool}`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Code"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"id\": \"cpptraj_average\",\n",
      "  \"description\": \"Calculates a structure average of a given cpptraj compatible trajectory.\",\n",
      "  \"arguments\": [\n",
      "    {\n",
      "      \"id\": \"config\",\n",
      "      \"required\": false,\n",
      "      \"description\": \"Configuration file for the cpptraj_average tool\",\n",
      "      \"filetype\": \"input\",\n",
      "      \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_average.json\",\n",
      "      \"formats\": [\n",
      "        \".*\\\\.json$\",\n",
      "        \".*\\\\.yml$\"\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"input_top_path\",\n",
      "      \"required\": true,\n",
      "      \"description\": \"Path to the input structure or topology file\",\n",
      "      \"filetype\": \"input\",\n",
      "      \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "      \"formats\": [\n",
      "        \".*\\\\.top$\",\n",
      "        \".*\\\\.pdb$\",\n",
      "        \".*\\\\.prmtop$\",\n",
      "        \".*\\\\.parmtop$\",\n",
      "        \".*\\\\.zip$\"\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"input_traj_path\",\n",
      "      \"required\": true,\n",
      "      \"description\": \"Path to the input trajectory to be processed\",\n",
      "      \"filetype\": \"input\",\n",
      "      \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "      \"formats\": [\n",
      "        \".*\\\\.crd$\",\n",
      "        \".*\\\\.cdf$\",\n",
      "        \".*\\\\.netcdf$\",\n",
      "        \".*\\\\.restart$\",\n",
      "        \".*\\\\.ncrestart$\",\n",
      "        \".*\\\\.restartnc$\",\n",
      "        \".*\\\\.dcd$\",\n",
      "        \".*\\\\.charmm$\",\n",
      "        \".*\\\\.cor$\",\n",
      "        \".*\\\\.pdb$\",\n",
      "        \".*\\\\.mol2$\",\n",
      "        \".*\\\\.trr$\",\n",
      "        \".*\\\\.gro$\",\n",
      "        \".*\\\\.binpos$\",\n",
      "        \".*\\\\.xtc$\",\n",
      "        \".*\\\\.cif$\",\n",
      "        \".*\\\\.arc$\",\n",
      "        \".*\\\\.sqm$\",\n",
      "        \".*\\\\.sdf$\",\n",
      "        \".*\\\\.conflib$\"\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"output_cpptraj_path\",\n",
      "      \"required\": true,\n",
      "      \"description\": \"Path to the output processed structure\",\n",
      "      \"filetype\": \"output\",\n",
      "      \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.average.pdb\",\n",
      "      \"formats\": [\n",
      "        \".*\\\\.crd$\",\n",
      "        \".*\\\\.netcdf$\",\n",
      "        \".*\\\\.rst7$\",\n",
      "        \".*\\\\.ncrst$\",\n",
      "        \".*\\\\.dcd$\",\n",
      "        \".*\\\\.pdb$\",\n",
      "        \".*\\\\.mol2$\",\n",
      "        \".*\\\\.binpos$\",\n",
      "        \".*\\\\.trr$\",\n",
      "        \".*\\\\.xtc$\",\n",
      "        \".*\\\\.sqm$\"\n",
      "      ]\n",
      "    }\n",
      "  ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "package = 'biobb_analysis'\n",
    "tool = 'cpptraj_average'\n",
    "url = apiURL + 'launch/' + package + '/' + tool\n",
    "response = get_data(url)\n",
    "\n",
    "print(json.dumps(response.json, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"launch_tool_ex\"></a>\n",
    "### Launch tool\n",
    "\n",
    "For more information about this endpoint, please visit the [BioBB REST API Documentation section](http://mmb.irbbarcelona.org/biobb-api/rest#/Launch%20Tool/postLaunchTool). The documentation for all the tools is available in the [BioBB REST API Tools Documentation section](http://mmb.irbbarcelona.org/biobb-api/tools-documentation?docExpansion=none). Interactive examples for all the tools are available in the [BioBB REST API Tools Execution section](http://mmb.irbbarcelona.org/biobb-api/tools-execution).\n",
    "\n",
    "Definition of functions needed for launch a job:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from io import BytesIO\n",
    "from pathlib import Path\n",
    "\n",
    "# Function used for encode python dictionary to JSON file\n",
    "def encode_config(data):\n",
    "    jsonData = json.dumps(data)\n",
    "    binaryData = jsonData.encode()\n",
    "    return BytesIO(binaryData)\n",
    "\n",
    "# Launch job\n",
    "def launch_job(url, **kwargs):\n",
    "    data = {}\n",
    "    files = {}\n",
    "    # Fill data (output paths) and files (input files) objects\n",
    "    for key, value in kwargs.items():\n",
    "        # Inputs / Outputs\n",
    "        if type(value) is str:\n",
    "            if key.startswith('input'):\n",
    "                files[key] = (value,  open(value, 'rb'))\n",
    "            elif key.startswith('output'):\n",
    "                data[key] = value\n",
    "            elif Path(value).is_file():\n",
    "                files[key] = (value,  open(value, 'rb'))\n",
    "        # Properties (in case properties are provided as a dictionary instead of a file)\n",
    "        if type(value) is dict:\n",
    "            files['config'] = ('prop.json', encode_config(value))\n",
    "    # Request URL with data and files\n",
    "    response = post_data(url, data, files)\n",
    "    # Print REST API response\n",
    "    print(json.dumps(response.json, indent=2))\n",
    "    # Save token if status == 303\n",
    "    if response.status == 303:\n",
    "        token = response.json['token']\n",
    "        return token"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hereafter we will launch a job on *biobb_analysis.cpptraj_average* tool with the provided *files/* in the files folder of this same repository. The response is a JSON with the status code, the state of the job, a message and a token for checking the job status.\n",
    "\n",
    "<a id=\"tool_yml_ex\"></a>\n",
    "#### Launch job with a YAML file config\n",
    "\n",
    "##### File config"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```yaml \n",
    "properties:\n",
    "    in_parameters:\n",
    "      start: 1\n",
    "      end: -1\n",
    "      step: 1\n",
    "      mask: c-alpha\n",
    "    out_parameters:\n",
    "      format: pdb\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Endpoint"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**POST** `http://mmb.irbbarcelona.org/biobb-api/rest/v1/launch/{package}/{tool}`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function below sends POST data and files to the *{package}/{tool}* endpoint. The config properties are sent as a YAML file.\n",
    "\n",
    "The response is a JSON with the status code, the state of the job, a message and a token that will be used for checking the job status in the next step. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"code\": 303,\n",
      "  \"state\": \"RUNNING\",\n",
      "  \"message\": \"The requested job has has been successfully launched, please go to /retrieve/status/{token} for checking job status.\",\n",
      "  \"token\": \"fe2805760eeeec0d5b8a34fbc40aa6c2a2d68c7ba1663cccb88659b1e149c898a414bbc04e37bb73efc725b7a29de2a93ffb55e6ef85cd6467f3d62a06ea5bfa\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Launch BioBB on REST API with YAML config file\n",
    "\n",
    "token = launch_job(url = apiURL + 'launch/biobb_analysis/cpptraj_average', \n",
    "                   config = 'files/config.yml',\n",
    "                   input_top_path = 'files/cpptraj.parm.top',\n",
    "                   input_traj_path = 'files/cpptraj.traj.dcd',\n",
    "                   output_cpptraj_path = 'output.cpptraj.average.pdb')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"tool_json_ex\"></a>\n",
    "#### Launch job with a JSON file config\n",
    "\n",
    "File config:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```json\n",
    "{\n",
    "\t\"in_parameters\": {\n",
    "\t\t\"start\": 1,\n",
    "\t\t\"end\": -1,\n",
    "\t\t\"step\": 1,\n",
    "\t\t\"mask\": \"c-alpha\"\n",
    "\t},\n",
    "\t\"out_parameters\": {\n",
    "\t\t\"format\": \"pdb\"\n",
    "\t}\n",
    "}\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Endpoint"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**POST** `http://mmb.irbbarcelona.org/biobb-api/rest/v1/launch/{package}/{tool}`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function below sends POST data and files to the *{package}/{tool}* endpoint. The config properties are sent as a JSON file.\n",
    "\n",
    "The response is a JSON with the status code, the state of the job, a message and a token that will be used for checking the job status in the next step. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"code\": 303,\n",
      "  \"state\": \"RUNNING\",\n",
      "  \"message\": \"The requested job has has been successfully launched, please go to /retrieve/status/{token} for checking job status.\",\n",
      "  \"token\": \"84ab5ef63d82ab3fa4f120532949905d83f6aff65f101cb1ed5fdd5f05acb00421ddc4560098f877f26a96972a8ea8521ab222a0bb78a5ffa9d213c0ab2618c9\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Launch BioBB on REST API with JSON config file\n",
    "\n",
    "token = launch_job(url = apiURL + 'launch/biobb_analysis/cpptraj_average', \n",
    "                   config = 'files/config.json',\n",
    "                   input_top_path = 'files/cpptraj.parm.top',\n",
    "                   input_traj_path = 'files/cpptraj.traj.dcd',\n",
    "                   output_cpptraj_path = 'output.cpptraj.average.pdb')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"tool_dict_ex\"></a>\n",
    "#### Launch job with a python dictionary config"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Endpoint"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**POST** `http://mmb.irbbarcelona.org/biobb-api/rest/v1/launch/{package}/{tool}`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function below sends POST data and files to the *{package}/{tool}* endpoint. The config properties are sent as a python dictionary embedded in the code.\n",
    "\n",
    "The response is a JSON with the status code, the state of the job, a message and a token that will be used for checking the job status in the next step. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"code\": 303,\n",
      "  \"state\": \"RUNNING\",\n",
      "  \"message\": \"The requested job has has been successfully launched, please go to /retrieve/status/{token} for checking job status.\",\n",
      "  \"token\": \"98013d74bef397d5498db3eb1008e5e136702d63903b6ea0cb5a2db44c4a4e0adbcd1ce9999915acd90c444f8749880c052185bbbfc747c1ebc7d67d6d2c84c8\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Launch BioBB on REST API with JSON config file\n",
    "\n",
    "prop = {\n",
    "    \"in_parameters\" : {\n",
    "      \"start\": 1,\n",
    "      \"end\": -1,\n",
    "      \"step\": 1,\n",
    "      \"mask\": \"c-alpha\"\n",
    "    },\n",
    "    \"out_parameters\" : {\n",
    "      \"format\": \"pdb\"\n",
    "    }\n",
    "}\n",
    "\n",
    "token = launch_job(url = apiURL + 'launch/biobb_analysis/cpptraj_average', \n",
    "                   config = prop,\n",
    "                   input_top_path = 'files/cpptraj.parm.top',\n",
    "                   input_traj_path = 'files/cpptraj.traj.dcd',\n",
    "                   output_cpptraj_path = 'output.cpptraj.average.pdb')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"retrieve_status_ex\"></a>\n",
    "### Retrieve status\n",
    "\n",
    "For more information about this endpoint, please visit the [BioBB REST API Documentation section](http://mmb.irbbarcelona.org/biobb-api/rest#/Retrieve/getRetrieveStatus).\n",
    "\n",
    "Definition of functions needed for retrieve the status of a job:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import datetime\n",
    "from time import sleep\n",
    "\n",
    "# Checks status until a provided \"ok\" status is returned by the response\n",
    "def check_status(url, ok, error):\n",
    "    counter = 0\n",
    "    while True:\n",
    "        if counter < 10: slp = 1\n",
    "        if counter >= 10 and counter < 60: slp = 10\n",
    "        if counter >= 60: slp = 60\n",
    "        counter = counter + slp\n",
    "        sleep(slp)\n",
    "        r = requests.get(url)\n",
    "        if r.status_code == ok or r.status_code == error:\n",
    "            return counter\n",
    "            break\n",
    "\n",
    "# Function that checks the status and parses the reponse JSON for saving th output files in a list\n",
    "def check_job(token):\n",
    "    # define retrieve status URL\n",
    "    url = apiURL + 'retrieve/status/' + token\n",
    "    # check status until job has finished\n",
    "    counter = check_status(url, 200, 500)\n",
    "    # Get content when status = 200\n",
    "    response = get_data(url)\n",
    "    # Save id for the generated output_files\n",
    "    if response.status == 200:\n",
    "        out_files = []\n",
    "        for outf in response.json['output_files']:\n",
    "            item = { 'id': outf['id'], 'name': outf['name'] }\n",
    "            out_files.append(item)\n",
    "\n",
    "    # Print REST API response\n",
    "    print(\"Total elapsed time: %s\" % str(datetime.timedelta(seconds=counter)))\n",
    "    print(\"REST API JSON response:\")\n",
    "    print(json.dumps(response.json, indent=4))\n",
    "    \n",
    "    if response.status == 200: \n",
    "        return out_files\n",
    "    else: return None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Endpoint"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**GET** `http://mmb.irbbarcelona.org/biobb-api/rest/v1/retrieve/status/{token}`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function below checks the status of a job and awaits until the response status is `200`. The response is a JSON with the status code, the state of the job, a message, a list with all the generated output files and the date of the expiration of these files. Additionally, the function also provides the elapsed time since the job has been launched until it has finished. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total elapsed time: 0:00:20\n",
      "REST API JSON response:\n",
      "{\n",
      "    \"code\": 200,\n",
      "    \"state\": \"FINISHED\",\n",
      "    \"message\": \"The requested job has finished successfully, please go to /retrieve/data/{id} for each output_files.\",\n",
      "    \"output_files\": [\n",
      "        {\n",
      "            \"id\": \"5e42837a40fe75.05757111\",\n",
      "            \"name\": \"output.cpptraj.average.pdb\",\n",
      "            \"size\": 77397,\n",
      "            \"mimetype\": \"text/plain\"\n",
      "        }\n",
      "    ],\n",
      "    \"expiration\": \"February 13, 2020 00:00 GMT+0000\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Check job status\n",
    "out_files = check_job(token)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"retrieve_data_ex\"></a>\n",
    "### Retrieve data\n",
    "\n",
    "For more information about this endpoint, please visit the [BioBB REST API Documentation section](http://mmb.irbbarcelona.org/biobb-api/rest#/Retrieve/getRetrieveData).\n",
    "\n",
    "Definition of functions needed for retrieve the output file(s) generated by a job:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Downloads to disk a file from a given URL\n",
    "def get_file(url, filename):\n",
    "    r = requests.get(url, allow_redirects=True)\n",
    "    file = open(filename,'wb') \n",
    "    file.write(r.content) \n",
    "    file.close()\n",
    "\n",
    "# Retrieves all the files provided in the out_files list\n",
    "def retrieve_data(out_files):\n",
    "    if not out_files:\n",
    "        return \"No files provided\"\n",
    "    for outf in out_files:\n",
    "        get_file(apiURL + 'retrieve/data/' + outf['id'], outf['name'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Endpoint"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**GET** `http://mmb.irbbarcelona.org/biobb-api/rest/v1/retrieve/data/{id}`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Code"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function below makes a single call to the *retrieve/data* endpoint for each output file got in the *retrieve/status* endpoint and save the generated file(s) to disk."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save generated file(s) to disk\n",
    "\n",
    "retrieve_data(out_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"practical_cases\"></a>\n",
    "## Practical cases"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we will execute some Bioexcel Building Blocks through the BioBB REST API and with the results we will do some interactions with other python libraries such as [plotly](https://plot.ly/python/offline/) or [nglview](http://nglviewer.org/#nglview)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"example1\"></a>\n",
    "### Example 1: download PDB file from RSCB database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Launch the *biobb_io.pdb* job that downloads a PDB file from the RSCB database:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"code\": 303,\n",
      "  \"state\": \"RUNNING\",\n",
      "  \"message\": \"The requested job has has been successfully launched, please go to /retrieve/status/{token} for checking job status.\",\n",
      "  \"token\": \"af60d733db949a71167f3aa6a7a793fc520b5a4176b57a770bed4798654a79be2a47b81e6a77de4eb285de84f9b768b119004f0bfbbe4be9e5ff1ffe31b81fd9\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Downloading desired PDB file\n",
    "\n",
    "# Create properties dict and inputs/outputs\n",
    "downloaded_pdb = '3EBP.pdb'\n",
    "prop = {\n",
    "    'pdb_code': '3EBP',\n",
    "    'filter': False\n",
    "}\n",
    "\n",
    "# Launch bb on REST API\n",
    "token = launch_job(url = apiURL + 'launch/biobb_io/pdb', \n",
    "                   config = prop,\n",
    "                   output_pdb_path = downloaded_pdb)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total elapsed time: 0:00:06\n",
      "REST API JSON response:\n",
      "{\n",
      "    \"code\": 200,\n",
      "    \"state\": \"FINISHED\",\n",
      "    \"message\": \"The requested job has finished successfully, please go to /retrieve/data/{id} for each output_files.\",\n",
      "    \"output_files\": [\n",
      "        {\n",
      "            \"id\": \"5e428389eeafa3.49051362\",\n",
      "            \"name\": \"3EBP.pdb\",\n",
      "            \"size\": 609120,\n",
      "            \"mimetype\": \"text/plain\"\n",
      "        }\n",
      "    ],\n",
      "    \"expiration\": \"February 13, 2020 00:00 GMT+0000\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Check job status\n",
    "out_files = check_job(token)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save generated file to disk\n",
    "retrieve_data(out_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Visualize downloaded PDB in NGLView:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "0576fdf66f5f44b79c09775320c4a210",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import nglview\n",
    "\n",
    "# Show protein\n",
    "view = nglview.show_file(downloaded_pdb)\n",
    "view.add_representation(repr_type='ball+stick', selection='het')\n",
    "view._remote_call('setSize', target='Widget', args=['','600px'])\n",
    "view"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "view.render_image()\n",
    "view.download_image(filename='ngl1.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src='ngl1.png'></img>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"example2\"></a>\n",
    "### Example 2: extract heteroatom from a given structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Launch the *biobb_structure_utils.extract_heteroatoms* job that extracts a heteroatom from a PDB file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"code\": 303,\n",
      "  \"state\": \"RUNNING\",\n",
      "  \"message\": \"The requested job has has been successfully launched, please go to /retrieve/status/{token} for checking job status.\",\n",
      "  \"token\": \"740c9ac30767ba996e445e4ed05c151ee903fed2234c0cc7ace6ec3ba4e1fa8bdcd5a3c6835c7f1038530eb81c4cc319674f235cf55b38863903181dce09a8d1\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Extracting heteroatom from a given structure\n",
    "\n",
    "# Create properties dict and inputs/outputs\n",
    "heteroatom = 'CPB.pdb'\n",
    "prop = {\n",
    "    'heteroatoms': [{\n",
    "        'name': 'CPB'\n",
    "    }]\n",
    "}\n",
    "\n",
    "# Launch bb on REST API\n",
    "token = launch_job(url = apiURL + 'launch/biobb_structure_utils/extract_heteroatoms', \n",
    "                   config = prop,\n",
    "                   input_structure_path = downloaded_pdb,\n",
    "                   output_heteroatom_path = heteroatom)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total elapsed time: 0:00:20\n",
      "REST API JSON response:\n",
      "{\n",
      "    \"code\": 200,\n",
      "    \"state\": \"FINISHED\",\n",
      "    \"message\": \"The requested job has finished successfully, please go to /retrieve/data/{id} for each output_files.\",\n",
      "    \"output_files\": [\n",
      "        {\n",
      "            \"id\": \"5e4283986555a0.86371712\",\n",
      "            \"name\": \"CPB.pdb\",\n",
      "            \"size\": 2268,\n",
      "            \"mimetype\": \"text/plain\"\n",
      "        }\n",
      "    ],\n",
      "    \"expiration\": \"February 13, 2020 00:00 GMT+0000\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Check job status\n",
    "out_files = check_job(token)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save generated file to disk\n",
    "retrieve_data(out_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Visualize generated extracted heteroatom in NGLView:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "145e77a5e08b474e90f501f5e30e6c74",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "NGLWidget()"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Show protein\n",
    "view = nglview.show_file(heteroatom)\n",
    "view.add_representation(repr_type='ball+stick', selection='het')\n",
    "view._remote_call('setSize', target='Widget', args=['','600px'])\n",
    "view"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "view.render_image()\n",
    "view.download_image(filename='ngl2.png')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src='ngl2.png'></img>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"example3\"></a>\n",
    "### Example 3: extract energy components from a given GROMACS energy file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"code\": 303,\n",
      "  \"state\": \"RUNNING\",\n",
      "  \"message\": \"The requested job has has been successfully launched, please go to /retrieve/status/{token} for checking job status.\",\n",
      "  \"token\": \"170e8e2645d179eaa40e2de652f3e6dec909ef1df4642526ba789ed21806bab917bb4ce7f9fb730dfe635155f52ac9f3869c4429afcc767bd190f01337a8a718\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# GMXEnergy: Getting system energy by time \n",
    "\n",
    "# Create prop dict and inputs/outputs\n",
    "output_min_ene_xvg ='file_min_ene.xvg'\n",
    "output_min_edr = 'files/1AKI_min.edr'\n",
    "prop = {\n",
    "    'terms':  [\"Potential\"]\n",
    "}\n",
    "\n",
    "# Launch bb on REST API\n",
    "token = launch_job(url = apiURL + 'launch/biobb_analysis/gmx_energy',\n",
    "                   config = prop,\n",
    "                   input_energy_path = output_min_edr,\n",
    "                   output_xvg_path = output_min_ene_xvg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total elapsed time: 0:00:08\n",
      "REST API JSON response:\n",
      "{\n",
      "    \"code\": 200,\n",
      "    \"state\": \"FINISHED\",\n",
      "    \"message\": \"The requested job has finished successfully, please go to /retrieve/data/{id} for each output_files.\",\n",
      "    \"output_files\": [\n",
      "        {\n",
      "            \"id\": \"5e4283a6c70143.38956052\",\n",
      "            \"name\": \"file_min_ene.xvg\",\n",
      "            \"size\": 54143,\n",
      "            \"mimetype\": \"text/plain\"\n",
      "        }\n",
      "    ],\n",
      "    \"expiration\": \"February 13, 2020 00:00 GMT+0000\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Check job status\n",
    "out_files = check_job(token)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save generated file to disk\n",
    "retrieve_data(out_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Visualize generated energy file in plotly:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "        <script type=\"text/javascript\">\n",
       "        window.PlotlyConfig = {MathJaxConfig: 'local'};\n",
       "        if (window.MathJax) {MathJax.Hub.Config({SVG: {font: \"STIX-Web\"}});}\n",
       "        if (typeof require !== 'undefined') {\n",
       "        require.undef(\"plotly\");\n",
       "        requirejs.config({\n",
       "            paths: {\n",
       "                'plotly': ['https://cdn.plot.ly/plotly-latest.min']\n",
       "            }\n",
       "        });\n",
       "        require(['plotly'], function(Plotly) {\n",
       "            window._Plotly = Plotly;\n",
       "        });\n",
       "        }\n",
       "        </script>\n",
       "        "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/vnd.plotly.v1+json": {
       "config": {
        "linkText": "Export to plot.ly",
        "plotlyServerURL": "https://plot.ly",
        "showLink": false
       },
       "data": [
        {
         "type": "scatter",
         "uid": "35cce87e-0e89-46fd-963c-0961ffc0cbd6",
         "x": [
          0,
          1,
          2,
          3,
          4,
          5,
          6,
          7,
          8,
          9,
          11,
          12,
          13,
          15,
          16,
          17,
          18,
          19,
          21,
          22,
          23,
          25,
          26,
          27,
          28,
          29,
          31,
          33,
          34,
          35,
          36,
          38,
          39,
          40,
          42,
          43,
          44,
          45,
          46,
          48,
          49,
          50,
          52,
          53,
          54,
          55,
          56,
          58,
          59,
          60,
          62,
          63,
          64,
          65,
          67,
          68,
          69,
          71,
          72,
          73,
          74,
          75,
          77,
          78,
          79,
          81,
          82,
          83,
          84,
          86,
          87,
          88,
          90,
          91,
          92,
          93,
          95,
          96,
          97,
          99,
          100,
          101,
          102,
          104,
          105,
          106,
          108,
          109,
          110,
          111,
          112,
          114,
          115,
          116,
          118,
          119,
          120,
          121,
          123,
          124,
          125,
          127,
          128,
          129,
          130,
          132,
          133,
          134,
          136,
          137,
          138,
          139,
          140,
          141,
          143,
          144,
          145,
          147,
          148,
          149,
          150,
          152,
          153,
          154,
          156,
          157,
          158,
          159,
          160,
          162,
          163,
          164,
          166,
          167,
          168,
          169,
          171,
          172,
          173,
          175,
          176,
          177,
          178,
          180,
          181,
          182,
          184,
          185,
          186,
          187,
          189,
          190,
          191,
          193,
          194,
          195,
          196,
          197,
          198,
          200,
          202,
          203,
          204,
          205,
          207,
          208,
          209,
          211,
          212,
          213,
          215,
          216,
          217,
          218,
          219,
          221,
          222,
          223,
          225,
          226,
          227,
          228,
          229,
          231,
          232,
          233,
          235,
          236,
          237,
          238,
          239,
          241,
          242,
          243,
          245,
          246,
          247,
          248,
          249,
          251,
          252,
          253,
          255,
          256,
          257,
          258,
          259,
          261,
          262,
          263,
          265,
          266,
          267,
          268,
          269,
          271,
          272,
          273,
          275,
          276,
          277,
          278,
          280,
          281,
          282,
          284,
          285,
          286,
          287,
          289,
          290,
          291,
          293,
          294,
          295,
          296,
          297,
          299,
          300,
          301,
          303,
          304,
          305,
          306,
          307,
          309,
          310,
          311,
          313,
          314,
          315,
          316,
          317,
          319,
          320,
          321,
          323,
          324,
          325,
          326,
          328,
          329,
          330,
          332,
          333,
          334,
          335,
          337,
          338,
          339,
          341,
          342,
          343,
          344,
          345,
          347,
          348,
          349,
          351,
          352,
          353,
          354,
          355,
          357,
          358,
          359,
          361,
          362,
          363,
          364,
          366,
          367,
          368,
          370,
          371,
          372,
          373,
          375,
          376,
          377,
          379,
          380,
          381,
          382,
          383,
          384,
          386,
          387,
          388,
          390,
          391,
          392,
          393,
          395,
          396,
          397,
          399,
          400,
          401,
          402,
          403,
          405,
          406,
          407,
          409,
          410,
          411,
          412,
          414,
          415,
          416,
          418,
          419,
          420,
          421,
          422,
          423,
          425,
          427,
          428,
          429,
          430,
          432,
          433,
          434,
          436,
          437,
          438,
          440,
          441,
          442,
          443,
          444,
          446,
          447,
          448,
          450,
          451,
          452,
          453,
          454,
          455,
          456,
          458,
          459,
          460,
          462,
          463,
          464,
          465,
          467,
          468,
          469,
          471,
          472,
          473,
          474,
          475,
          477,
          478,
          479,
          481,
          482,
          483,
          484,
          486,
          487,
          488,
          490,
          491,
          492,
          493,
          495,
          496,
          497,
          499,
          500,
          501,
          502,
          503,
          505,
          506,
          507,
          509,
          510,
          511,
          512,
          513,
          515,
          516,
          517,
          519,
          520,
          521,
          522,
          524,
          525,
          526,
          528,
          529,
          530,
          531,
          533,
          534,
          535,
          537,
          538,
          539,
          540,
          541,
          543,
          544,
          545,
          547,
          548,
          549,
          550,
          551,
          553,
          554,
          555,
          557,
          558,
          559,
          560,
          561,
          563,
          564,
          565,
          567,
          568,
          569,
          570,
          572,
          573,
          574,
          576,
          577,
          578,
          579,
          581,
          582,
          583,
          585,
          586,
          587,
          588,
          589,
          591,
          592,
          593,
          595,
          596,
          597,
          598,
          599,
          601,
          602,
          603,
          605,
          606,
          607,
          608,
          610,
          611,
          612,
          614,
          615,
          616,
          617,
          619,
          620,
          621,
          623,
          624,
          625,
          626,
          627,
          629,
          630,
          631,
          633,
          634,
          635,
          636,
          637,
          639,
          640,
          641,
          643,
          644,
          645,
          646,
          648,
          649,
          650,
          652,
          653,
          654,
          655,
          657,
          658,
          659,
          661,
          662,
          663,
          664,
          665,
          666,
          668,
          669,
          670,
          672,
          673,
          675,
          676,
          677,
          679,
          680,
          681,
          682,
          683,
          685,
          686,
          687,
          689,
          690,
          691,
          692,
          693,
          694,
          695,
          697,
          699,
          700,
          701,
          702,
          704,
          705,
          706,
          708,
          709,
          710,
          711,
          712,
          714,
          715,
          716,
          718,
          719,
          720,
          721,
          722,
          724,
          725,
          726,
          728,
          729,
          730,
          731,
          732,
          733,
          735,
          736,
          737,
          739,
          740,
          741,
          742,
          744,
          745,
          746,
          748,
          749,
          750,
          751,
          753,
          754,
          755,
          757,
          758,
          759,
          760,
          761,
          763,
          764,
          765,
          767,
          768,
          769,
          770,
          772,
          773,
          774,
          776,
          777,
          778,
          779,
          781,
          782,
          783,
          785,
          786,
          787,
          788,
          789,
          790,
          792,
          793,
          794,
          796,
          797,
          799,
          800,
          801,
          803,
          804,
          805,
          806,
          807,
          808,
          809,
          811,
          812,
          813,
          815,
          816,
          817,
          818,
          819,
          821,
          822,
          823,
          825,
          826,
          827,
          828,
          830,
          831,
          832,
          834,
          835,
          836,
          837,
          839,
          840,
          841,
          843,
          844,
          845,
          846,
          847,
          849,
          850,
          851,
          853,
          854,
          855,
          856,
          858,
          859,
          860,
          862,
          863,
          864,
          865,
          867,
          868,
          869,
          871,
          872,
          873,
          874,
          875,
          877,
          878,
          879,
          881,
          882,
          883,
          884,
          885,
          886,
          888,
          889,
          890,
          892,
          893,
          895,
          896,
          897,
          899,
          900,
          901,
          902,
          903,
          905,
          906,
          907,
          909,
          910,
          911,
          912,
          913,
          914,
          915,
          917,
          919,
          920,
          921,
          922,
          924,
          925,
          926,
          928,
          929,
          930,
          931,
          932,
          934,
          935,
          936,
          938,
          939,
          940,
          941,
          942,
          944,
          945,
          946,
          948,
          949,
          950,
          951,
          952,
          953,
          955,
          956,
          957,
          959,
          960,
          961,
          962,
          964,
          965,
          966,
          968,
          969,
          970,
          971,
          973,
          974,
          975,
          977,
          978,
          979,
          980,
          981,
          983,
          984,
          985,
          987,
          988,
          989,
          990,
          991,
          993,
          994,
          995,
          997,
          998,
          999,
          1000,
          1002,
          1003,
          1004,
          1006,
          1007,
          1008,
          1009,
          1011,
          1012,
          1013,
          1015,
          1016,
          1017,
          1018,
          1019,
          1021,
          1022,
          1023,
          1025,
          1026,
          1027,
          1028,
          1030,
          1031,
          1032,
          1034,
          1035,
          1036,
          1037,
          1039,
          1040,
          1041,
          1043,
          1044,
          1045,
          1046,
          1047,
          1048,
          1049,
          1051,
          1053,
          1054,
          1055,
          1056,
          1058,
          1059,
          1060,
          1062,
          1063,
          1064,
          1065,
          1066,
          1068,
          1069,
          1070,
          1072,
          1073,
          1074,
          1075,
          1076,
          1078,
          1079,
          1080,
          1082,
          1083,
          1084,
          1085,
          1086,
          1087,
          1089,
          1091,
          1092,
          1093,
          1094,
          1096,
          1097,
          1098,
          1100,
          1101,
          1102,
          1103,
          1104,
          1106,
          1107,
          1108,
          1110,
          1111,
          1112,
          1113,
          1114,
          1115,
          1116,
          1118,
          1120,
          1121,
          1122,
          1123,
          1125,
          1126,
          1127,
          1129,
          1130,
          1131,
          1133,
          1134,
          1135,
          1136,
          1137,
          1138,
          1139,
          1141,
          1142,
          1143,
          1145,
          1146,
          1147,
          1148,
          1150,
          1151,
          1152,
          1154,
          1155,
          1156,
          1157,
          1158,
          1159,
          1161,
          1162,
          1163,
          1165,
          1166,
          1168,
          1169,
          1170,
          1172,
          1173,
          1174,
          1175,
          1176,
          1178,
          1179,
          1180,
          1182,
          1183,
          1184,
          1185,
          1186,
          1188,
          1189,
          1190,
          1192,
          1193,
          1194,
          1195,
          1196,
          1198,
          1199,
          1200,
          1202,
          1203,
          1204,
          1205,
          1206,
          1207,
          1209,
          1211,
          1212,
          1213,
          1214,
          1216,
          1217,
          1218,
          1220,
          1221,
          1222,
          1224,
          1225,
          1226,
          1227,
          1228,
          1229,
          1230,
          1232,
          1233,
          1234,
          1236,
          1237,
          1238,
          1239,
          1241,
          1242,
          1243,
          1245,
          1246,
          1247,
          1248,
          1249,
          1251,
          1252,
          1253,
          1255,
          1256,
          1257,
          1258,
          1260,
          1261,
          1262,
          1264,
          1265,
          1266,
          1267,
          1269,
          1270,
          1271,
          1273,
          1274,
          1275,
          1276,
          1277,
          1279,
          1280,
          1281,
          1283,
          1284,
          1285,
          1286,
          1287,
          1288,
          1290,
          1291,
          1292,
          1294,
          1295,
          1297,
          1298,
          1299,
          1301,
          1302,
          1303,
          1304,
          1305,
          1307,
          1308,
          1309,
          1311,
          1312,
          1313,
          1314,
          1315,
          1317,
          1318,
          1319,
          1321,
          1322,
          1323,
          1324,
          1325,
          1327,
          1328,
          1329,
          1331,
          1332,
          1333,
          1334,
          1335,
          1336,
          1338,
          1339,
          1340,
          1342,
          1343,
          1345,
          1346,
          1347,
          1349,
          1350,
          1351,
          1352,
          1353,
          1355,
          1356,
          1357,
          1359,
          1360,
          1361,
          1362,
          1363,
          1365,
          1366,
          1367,
          1369,
          1370,
          1371,
          1372,
          1373,
          1374,
          1376,
          1377,
          1378,
          1380,
          1381,
          1383,
          1384,
          1385,
          1387,
          1388,
          1389,
          1390,
          1391,
          1393,
          1394,
          1395,
          1397,
          1398,
          1399,
          1400,
          1401,
          1403,
          1404,
          1405,
          1407,
          1408,
          1409,
          1410,
          1411,
          1413,
          1414,
          1415,
          1417,
          1418,
          1419,
          1420,
          1421,
          1423,
          1424,
          1425,
          1427,
          1428,
          1429,
          1430,
          1431,
          1432,
          1434,
          1435,
          1436,
          1438,
          1439,
          1441,
          1442,
          1443,
          1445,
          1446,
          1447,
          1448,
          1449,
          1451,
          1452,
          1453,
          1455,
          1456,
          1457,
          1458,
          1459,
          1461,
          1462,
          1463,
          1465,
          1466,
          1467,
          1468,
          1469,
          1470,
          1472,
          1473,
          1474,
          1476,
          1477,
          1479,
          1480,
          1481,
          1483,
          1484,
          1485,
          1486,
          1487,
          1489,
          1490,
          1491,
          1493,
          1494,
          1495,
          1496,
          1497,
          1499,
          1500,
          1501,
          1503,
          1504,
          1505,
          1506,
          1507,
          1509,
          1510,
          1511,
          1513,
          1514,
          1515,
          1516,
          1517,
          1518,
          1520,
          1521,
          1522,
          1524,
          1525,
          1527,
          1528,
          1529,
          1531,
          1532,
          1533,
          1534,
          1535,
          1537,
          1538,
          1539,
          1541,
          1542,
          1543,
          1544,
          1545,
          1547,
          1548,
          1549,
          1551,
          1552,
          1553,
          1554,
          1555,
          1557,
          1558,
          1559,
          1561,
          1562,
          1563,
          1564,
          1565,
          1566,
          1568,
          1569,
          1570,
          1572,
          1573,
          1575,
          1576,
          1577,
          1579,
          1580,
          1581,
          1582,
          1583,
          1585,
          1586,
          1587,
          1589,
          1590,
          1591,
          1592,
          1593,
          1595,
          1596,
          1597,
          1599,
          1600,
          1601,
          1602,
          1603,
          1605,
          1606,
          1607,
          1609,
          1610,
          1611,
          1612,
          1613,
          1614,
          1616,
          1617,
          1618,
          1620,
          1621,
          1623,
          1624,
          1625,
          1627,
          1628,
          1629,
          1630,
          1631,
          1633,
          1634,
          1635,
          1637,
          1638,
          1639,
          1640,
          1641,
          1643,
          1644,
          1645,
          1647,
          1648,
          1649,
          1650,
          1651,
          1653,
          1654,
          1655,
          1657,
          1658,
          1659,
          1660,
          1661,
          1662,
          1664,
          1666,
          1667,
          1668,
          1669,
          1671,
          1672,
          1673,
          1675,
          1676,
          1677,
          1679,
          1680,
          1681,
          1682,
          1683,
          1684,
          1685,
          1687,
          1688,
          1689,
          1691,
          1692,
          1693,
          1694,
          1696,
          1697,
          1698,
          1700,
          1701,
          1702,
          1703,
          1704,
          1706,
          1707,
          1708,
          1710,
          1711,
          1712,
          1713,
          1715,
          1716,
          1717,
          1719,
          1720,
          1721,
          1722,
          1724,
          1725,
          1726,
          1728,
          1729,
          1730,
          1731,
          1732,
          1733,
          1734,
          1736,
          1738,
          1739,
          1740,
          1741,
          1743,
          1744,
          1745,
          1747,
          1748,
          1749,
          1750,
          1751,
          1753,
          1754,
          1755,
          1757,
          1758,
          1759,
          1760,
          1761,
          1763,
          1764,
          1765,
          1767,
          1768,
          1769,
          1770,
          1771,
          1772,
          1774,
          1776,
          1777,
          1778,
          1779,
          1781,
          1782,
          1783,
          1785,
          1786,
          1787,
          1788,
          1789,
          1791,
          1792,
          1793,
          1795,
          1796,
          1797,
          1798,
          1799,
          1801,
          1802,
          1803,
          1805,
          1806,
          1807,
          1808,
          1809,
          1810,
          1812,
          1813,
          1814,
          1816,
          1817,
          1818,
          1819,
          1821,
          1822,
          1823,
          1825,
          1826,
          1827,
          1828,
          1830,
          1831,
          1832,
          1834,
          1835,
          1836,
          1837,
          1838,
          1840,
          1841,
          1842,
          1844,
          1845,
          1846,
          1847,
          1848,
          1850,
          1851,
          1852,
          1854,
          1855,
          1856,
          1857,
          1859,
          1860,
          1861,
          1863,
          1864,
          1865,
          1866,
          1868,
          1869,
          1870,
          1872,
          1873,
          1874,
          1875,
          1876,
          1878,
          1879,
          1880,
          1882,
          1883,
          1884,
          1885,
          1886,
          1887,
          1889,
          1890,
          1891,
          1893,
          1894,
          1896,
          1897,
          1898,
          1900,
          1901,
          1902,
          1903,
          1904,
          1906,
          1907,
          1908,
          1910,
          1911,
          1912,
          1913,
          1914,
          1915,
          1916,
          1918,
          1920,
          1921,
          1922,
          1923,
          1925,
          1926,
          1927,
          1929,
          1930,
          1931,
          1932,
          1933,
          1935,
          1936,
          1937,
          1939,
          1940,
          1941,
          1942,
          1943,
          1945,
          1946,
          1947,
          1949,
          1950,
          1951,
          1952,
          1953,
          1954,
          1956,
          1957,
          1958,
          1960,
          1961,
          1962,
          1963,
          1965,
          1966,
          1967,
          1969,
          1970,
          1971,
          1972,
          1974,
          1975,
          1976,
          1978,
          1979,
          1980,
          1981,
          1982,
          1984,
          1985,
          1986,
          1988,
          1989,
          1990,
          1991,
          1993,
          1994,
          1995,
          1997,
          1998,
          1999,
          2000,
          2002,
          2003,
          2004,
          2006,
          2007,
          2008,
          2009,
          2010,
          2011,
          2013,
          2014,
          2015,
          2017,
          2018,
          2020,
          2021,
          2022,
          2024,
          2025,
          2026,
          2027,
          2028,
          2030,
          2031,
          2032,
          2034,
          2035,
          2036,
          2037,
          2038,
          2039,
          2040,
          2042,
          2043,
          2044,
          2046,
          2047,
          2048,
          2049,
          2051,
          2052,
          2053,
          2055,
          2056,
          2057,
          2058,
          2060,
          2061,
          2062,
          2064,
          2065,
          2066,
          2067,
          2068,
          2070,
          2071,
          2072,
          2074,
          2075,
          2076,
          2077,
          2078,
          2080,
          2081,
          2082,
          2084,
          2085,
          2086,
          2087,
          2089,
          2090,
          2091,
          2093,
          2094,
          2095,
          2096,
          2097,
          2098,
          2100,
          2101,
          2102,
          2104,
          2105,
          2106,
          2107,
          2109,
          2110,
          2111,
          2113,
          2114,
          2115,
          2116,
          2118,
          2119,
          2120,
          2122,
          2123,
          2124,
          2125,
          2126,
          2128,
          2129,
          2130,
          2132,
          2133,
          2134,
          2135,
          2137,
          2138,
          2139,
          2141,
          2142,
          2143,
          2144,
          2145,
          2146,
          2148,
          2150,
          2151,
          2152,
          2153,
          2155,
          2156,
          2157,
          2159,
          2160,
          2161,
          2162,
          2163,
          2165,
          2166,
          2167,
          2169,
          2170,
          2171,
          2172,
          2173,
          2175,
          2176,
          2177,
          2179,
          2180,
          2181,
          2182,
          2183,
          2184,
          2186,
          2187,
          2188,
          2190,
          2191,
          2192,
          2193,
          2195,
          2196,
          2197,
          2199,
          2200,
          2201,
          2202,
          2204,
          2205,
          2206,
          2208,
          2209,
          2210,
          2211,
          2212,
          2213,
          2215,
          2216,
          2217,
          2219,
          2220,
          2222,
          2223,
          2224,
          2226,
          2227,
          2228,
          2229,
          2230,
          2231,
          2232,
          2234,
          2236,
          2237,
          2238,
          2239,
          2241,
          2242,
          2243,
          2245,
          2246,
          2247,
          2248,
          2249,
          2251,
          2252,
          2253,
          2255,
          2256,
          2257,
          2258,
          2259,
          2261,
          2262,
          2263,
          2265,
          2266,
          2267,
          2268,
          2269,
          2270,
          2272,
          2273,
          2274,
          2276,
          2277,
          2278,
          2279,
          2281,
          2282,
          2283,
          2285,
          2286,
          2287,
          2288,
          2290,
          2291,
          2292,
          2294,
          2295,
          2296,
          2297,
          2298,
          2299,
          2301,
          2302,
          2303,
          2305,
          2306,
          2308,
          2309,
          2310,
          2312,
          2313,
          2314,
          2315,
          2316,
          2318,
          2319,
          2320,
          2322,
          2323,
          2324,
          2325,
          2326,
          2327,
          2328,
          2330,
          2331,
          2332,
          2334,
          2335,
          2336,
          2337,
          2339,
          2340,
          2341,
          2343,
          2344,
          2345,
          2346,
          2348,
          2349,
          2350,
          2352,
          2353,
          2354,
          2355,
          2356,
          2358
         ],
         "y": [
          -327612.125,
          -347306.34375,
          -371487.59375,
          -400072.09375,
          -427773.65625,
          -452044.53125,
          -457291.3125,
          -461900.9375,
          -463964.40625,
          -467420.09375,
          -472346.78125,
          -473178.28125,
          -479740.1875,
          -482194.40625,
          -483604.125,
          -485770.09375,
          -486202.65625,
          -487951.03125,
          -491239.78125,
          -491569.21875,
          -496079.75,
          -497212.5625,
          -497915.5,
          -498762.71875,
          -498798.46875,
          -499045.03125,
          -503261.5625,
          -504590.28125,
          -505913.375,
          -506152.90625,
          -507573.75,
          -508783.5,
          -509315.84375,
          -510723.34375,
          -511568.875,
          -512161.84375,
          -513046.96875,
          -513196.65625,
          -514056.8125,
          -515344.4375,
          -515769,
          -517366,
          -517970.4375,
          -518392.78125,
          -518939.5625,
          -519113.40625,
          -519501.9375,
          -520976.40625,
          -521471.625,
          -523378.9375,
          -523770.625,
          -524163.09375,
          -524394.5625,
          -524645,
          -525844.3125,
          -526626.375,
          -528134.0625,
          -528465.9375,
          -528803.875,
          -529018.625,
          -529233.0625,
          -529242.25,
          -530579.875,
          -531357.625,
          -533034.6875,
          -533255.4375,
          -533526.75,
          -533612.9375,
          -533750.875,
          -534756.625,
          -534980.1875,
          -536368.8125,
          -536553.1875,
          -536811.1875,
          -536874.375,
          -537042.9375,
          -537863.8125,
          -538096.5,
          -539172.5,
          -539360.3125,
          -539594.5625,
          -539680.875,
          -539867.625,
          -540479.6875,
          -540767.5625,
          -541523.375,
          -541727,
          -541907.125,
          -542068.1875,
          -542175.0625,
          -542254.5625,
          -542921.125,
          -543344,
          -544164.375,
          -544323.1875,
          -544502.125,
          -544596.25,
          -544721.375,
          -545209.6875,
          -545588.875,
          -546230.375,
          -546383.0625,
          -546559.5625,
          -546649.875,
          -546800.5625,
          -547169.8125,
          -547469.75,
          -547933.0625,
          -548090.75,
          -548234.5625,
          -548365.4375,
          -548466.25,
          -548545.4375,
          -548559.375,
          -549138.6875,
          -549492.125,
          -550260.4375,
          -550351.6875,
          -550511.5,
          -550521.1875,
          -550657,
          -551055.5625,
          -551097.4375,
          -551613.375,
          -551739.4375,
          -551837.5,
          -551935,
          -551975.25,
          -552010.375,
          -552502.125,
          -552619.6875,
          -553261.875,
          -553348.5,
          -553464.625,
          -553493,
          -553570.0625,
          -553953.75,
          -554042.0625,
          -554555,
          -554648.125,
          -554751.625,
          -554800.875,
          -554869.5,
          -555197.625,
          -555378.875,
          -555810.75,
          -555902.8125,
          -556005.75,
          -556059.4375,
          -556136,
          -556407.5,
          -556606,
          -556965,
          -557057.625,
          -557153.75,
          -557216.5,
          -557288.75,
          -557297.3125,
          -557313.6875,
          -557746.875,
          -557914.875,
          -558077.875,
          -558093.625,
          -558289.8125,
          -558423.1875,
          -558476.375,
          -558630.9375,
          -558754.8125,
          -558838.6875,
          -558985.1875,
          -559090.875,
          -559186.0625,
          -559304.875,
          -559339.3125,
          -559465.4375,
          -559616.9375,
          -559699.25,
          -559890,
          -559985.875,
          -560059.4375,
          -560159,
          -560189.8125,
          -560289.25,
          -560469.0625,
          -560558.625,
          -560776.875,
          -560867.125,
          -560932.6875,
          -561023.875,
          -561048.125,
          -561133.75,
          -561330.4375,
          -561408.8125,
          -561650.6875,
          -561733.9375,
          -561794.1875,
          -561874,
          -561895.625,
          -561963.75,
          -562182.375,
          -562251.125,
          -562520.8125,
          -562597.125,
          -562654.1875,
          -562721.4375,
          -562741.75,
          -562789.5,
          -563033.1875,
          -563097.3125,
          -563400.1875,
          -563467.0625,
          -563523.0625,
          -563575.375,
          -563597.75,
          -563621.1875,
          -563894.8125,
          -563959.4375,
          -564299.75,
          -564356.6875,
          -564413.625,
          -564448.9375,
          -564474.25,
          -564699.375,
          -564888,
          -565190.75,
          -565242.0625,
          -565323.3125,
          -565336.75,
          -565417.875,
          -565566.4375,
          -565614.1875,
          -565797.0625,
          -565867.5,
          -565914.9375,
          -565984.875,
          -565999.125,
          -566061.625,
          -566231.1875,
          -566280.5,
          -566492.6875,
          -566555.625,
          -566602.625,
          -566659.25,
          -566676.9375,
          -566716.9375,
          -566912.3125,
          -566971.5,
          -567216.875,
          -567271.4375,
          -567320.4375,
          -567361.875,
          -567385.5,
          -567400,
          -567624.125,
          -567708.8125,
          -567990.25,
          -568034.75,
          -568087.5625,
          -568111.875,
          -568142.4375,
          -568319.3125,
          -568429.3125,
          -568665.6875,
          -568710.6875,
          -568771.5625,
          -568791.5,
          -568847.375,
          -568977.5,
          -569041.375,
          -569204.5,
          -569259.9375,
          -569304.5,
          -569353.9375,
          -569377.75,
          -569413.875,
          -569565.0625,
          -569651.875,
          -569843.4375,
          -569891.75,
          -569938.0625,
          -569974,
          -570004.25,
          -570017.25,
          -570193.375,
          -570331.0625,
          -570553.5625,
          -570594.25,
          -570644.0625,
          -570664.875,
          -570702.1875,
          -570835.875,
          -570920.4375,
          -571097.5,
          -571141.3125,
          -571189.75,
          -571218.0625,
          -571257.5625,
          -571367.625,
          -571469.5625,
          -571608.875,
          -571655.25,
          -571700,
          -571738.375,
          -571771.125,
          -571792.75,
          -571801.5,
          -571975.0625,
          -572077.0625,
          -572307.1875,
          -572338.5625,
          -572389.9375,
          -572396.5,
          -572441.3125,
          -572565.1875,
          -572592,
          -572751.8125,
          -572795.1875,
          -572830.125,
          -572864.6875,
          -572882,
          -572898.75,
          -573051.9375,
          -573120.9375,
          -573320.75,
          -573353.8125,
          -573396.9375,
          -573412.125,
          -573444.4375,
          -573564.3125,
          -573620.8125,
          -573778.0625,
          -573816.4375,
          -573854.9375,
          -573881,
          -573908.75,
          -573912.8125,
          -573915.8125,
          -574106.875,
          -574182.1875,
          -574253.25,
          -574259.8125,
          -574346.25,
          -574402.8125,
          -574426.6875,
          -574491.6875,
          -574545.5625,
          -574584.9375,
          -574649.125,
          -574693.8125,
          -574737.25,
          -574787.1875,
          -574807.125,
          -574859.4375,
          -574924.1875,
          -574971.4375,
          -575053.0625,
          -575095,
          -575131,
          -575172.4375,
          -575194.8125,
          -575231.5625,
          -575232.0625,
          -575253,
          -575399.5625,
          -575400.6875,
          -575596.9375,
          -575624.25,
          -575659.75,
          -575668.9375,
          -575693.375,
          -575814.1875,
          -575844.6875,
          -576000.5,
          -576033.125,
          -576064.25,
          -576086.9375,
          -576103.0625,
          -576106.8125,
          -576252.8125,
          -576313.625,
          -576500.75,
          -576526.25,
          -576560.8125,
          -576570.75,
          -576593.0625,
          -576706,
          -576748.25,
          -576901.625,
          -576929.4375,
          -576966.5625,
          -576977.9375,
          -577009.875,
          -577098.875,
          -577136.9375,
          -577249.625,
          -577284.0625,
          -577311,
          -577341.5625,
          -577353.8125,
          -577374.9375,
          -577479.625,
          -577520.9375,
          -577653.625,
          -577684.4375,
          -577711.625,
          -577735.125,
          -577748.6875,
          -577758.125,
          -577880.4375,
          -577925.1875,
          -578079.75,
          -578106.125,
          -578134.1875,
          -578150.375,
          -578164.5,
          -578266,
          -578347.0625,
          -578487.6875,
          -578512.625,
          -578551.375,
          -578557.1875,
          -578596.4375,
          -578667.375,
          -578688.0625,
          -578776.375,
          -578810.6875,
          -578832.8125,
          -578867.5625,
          -578872,
          -578904.625,
          -578987.6875,
          -579003.125,
          -579106.8125,
          -579138.8125,
          -579159.4375,
          -579189.1875,
          -579193.375,
          -579216.6875,
          -579314.125,
          -579325,
          -579448.25,
          -579476.375,
          -579497.125,
          -579520.4375,
          -579524.9375,
          -579537.125,
          -579651.75,
          -579660.0625,
          -579806,
          -579830.125,
          -579851.875,
          -579866.8125,
          -579874.25,
          -579975.0625,
          -580055.25,
          -580194.3125,
          -580213.875,
          -580250.1875,
          -580251.0625,
          -580287.875,
          -580355.5,
          -580365.3125,
          -580450.5,
          -580481.3125,
          -580499.5,
          -580530.4375,
          -580531.5,
          -580559.0625,
          -580640.25,
          -580647.125,
          -580750.6875,
          -580778.375,
          -580796.5625,
          -580820.0625,
          -580824.3125,
          -580839.25,
          -580937.0625,
          -580946.125,
          -581072.4375,
          -581094.75,
          -581115.75,
          -581129.6875,
          -581139,
          -581226.1875,
          -581300.375,
          -581419.1875,
          -581438.375,
          -581469.6875,
          -581474.375,
          -581504.6875,
          -581566.125,
          -581584.125,
          -581661.8125,
          -581688.5625,
          -581707.875,
          -581732.875,
          -581739.75,
          -581759.5625,
          -581834,
          -581857.8125,
          -581953.25,
          -581976.5625,
          -581997.3125,
          -582015.0625,
          -582026.125,
          -582032.6875,
          -582122.875,
          -582166.6875,
          -582283.625,
          -582302.3125,
          -582326.375,
          -582334.875,
          -582351.6875,
          -582422.8125,
          -582456.4375,
          -582550.875,
          -582571.6875,
          -582595.0625,
          -582607.3125,
          -582626.75,
          -582685.3125,
          -582726.9375,
          -582802.4375,
          -582825.3125,
          -582846.75,
          -582864.6875,
          -582879.5,
          -582888.8125,
          -582889.375,
          -582986.375,
          -583028.375,
          -583158.375,
          -583171.4375,
          -583199.375,
          -583237.3125,
          -583269.4375,
          -583316.875,
          -583340.625,
          -583362.25,
          -583386.125,
          -583398.875,
          -583421.1875,
          -583465.875,
          -583506.9375,
          -583564.125,
          -583587.0625,
          -583607.1875,
          -583627.375,
          -583641.625,
          -583656.9375,
          -583659,
          -583662.625,
          -583760.25,
          -583797.8125,
          -583834.375,
          -583839.25,
          -583881.9375,
          -583912.875,
          -583926.5625,
          -583963.25,
          -583990.0625,
          -584010.375,
          -584041.25,
          -584044.3125,
          -584077.875,
          -584117.3125,
          -584130.3125,
          -584180.25,
          -584204.375,
          -584220.8125,
          -584244.9375,
          -584251.5,
          -584272.5625,
          -584323.5625,
          -584345.9375,
          -584412.8125,
          -584433.125,
          -584451.375,
          -584467.3125,
          -584480,
          -584486.625,
          -584488.5,
          -584575.25,
          -584594.6875,
          -584708.125,
          -584721.5625,
          -584741.1875,
          -584744.75,
          -584757.9375,
          -584825.5,
          -584838.0625,
          -584928.1875,
          -584944.5,
          -584963.75,
          -584970.9375,
          -584985.5625,
          -585041.8125,
          -585069.125,
          -585141.8125,
          -585159.6875,
          -585177.25,
          -585190.8125,
          -585201.6875,
          -585206.4375,
          -585274.1875,
          -585325,
          -585413,
          -585428.625,
          -585447.875,
          -585454.75,
          -585469.625,
          -585521.875,
          -585548.6875,
          -585618.5625,
          -585635.875,
          -585654.0625,
          -585664.9375,
          -585679.8125,
          -585724.5,
          -585764.8125,
          -585822.25,
          -585840.375,
          -585857.6875,
          -585872.625,
          -585884.9375,
          -585893.25,
          -585895.3125,
          -585967.25,
          -586006.0625,
          -586103.375,
          -586114.9375,
          -586137,
          -586165.8125,
          -586193.3125,
          -586228.8125,
          -586247.75,
          -586265.375,
          -586284.8125,
          -586295.5,
          -586314.375,
          -586314.4375,
          -586328.875,
          -586388.75,
          -586390.375,
          -586467.75,
          -586483.875,
          -586496.4375,
          -586508.5,
          -586513.1875,
          -586516.1875,
          -586589.5625,
          -586601.6875,
          -586697.25,
          -586709.125,
          -586725.125,
          -586729.125,
          -586739.375,
          -586797.8125,
          -586813.25,
          -586892.1875,
          -586905,
          -586922.6875,
          -586928.25,
          -586942.375,
          -586989.375,
          -587008.3125,
          -587069.1875,
          -587085.125,
          -587099.375,
          -587111.8125,
          -587119.9375,
          -587126.0625,
          -587183.5,
          -587215.1875,
          -587288.8125,
          -587302.375,
          -587318.5625,
          -587325.875,
          -587337.4375,
          -587382.875,
          -587414.6875,
          -587475.875,
          -587490.1875,
          -587507.3125,
          -587515.625,
          -587530.4375,
          -587567.4375,
          -587594.9375,
          -587642.1875,
          -587658.25,
          -587672.75,
          -587686.5,
          -587696.25,
          -587705.6875,
          -587750.1875,
          -587789.5,
          -587845.9375,
          -587860.5,
          -587875.625,
          -587886.1875,
          -587897.875,
          -587901,
          -587903,
          -587970.6875,
          -587980.875,
          -588072.8125,
          -588081.3125,
          -588099.6875,
          -588127.4375,
          -588150.5625,
          -588184.8125,
          -588201.375,
          -588215.1875,
          -588231.5625,
          -588240.25,
          -588254.875,
          -588287.25,
          -588315.6875,
          -588357.125,
          -588372.375,
          -588386.5,
          -588399.9375,
          -588409.6875,
          -588420.125,
          -588420.4375,
          -588422.8125,
          -588492.875,
          -588517.75,
          -588543,
          -588545.5625,
          -588575.6875,
          -588597.875,
          -588606.6875,
          -588633,
          -588652.1875,
          -588665.5,
          -588687.25,
          -588688.6875,
          -588711.25,
          -588740.5625,
          -588748.9375,
          -588786.1875,
          -588802.5,
          -588814.3125,
          -588829.9375,
          -588834.9375,
          -588848.25,
          -588886.25,
          -588905.0625,
          -588955,
          -588968.6875,
          -588982.8125,
          -588992.4375,
          -589003.4375,
          -589004.8125,
          -589009.375,
          -589069.8125,
          -589071,
          -589150.1875,
          -589160.75,
          -589172.6875,
          -589177,
          -589183.5625,
          -589234,
          -589252.625,
          -589321.5,
          -589331.5625,
          -589347.4375,
          -589350.3125,
          -589364.125,
          -589402.625,
          -589412.4375,
          -589461.4375,
          -589475.625,
          -589486.8125,
          -589498.4375,
          -589503.4375,
          -589510.625,
          -589557.75,
          -589573.75,
          -589634.125,
          -589645.8125,
          -589658.375,
          -589665.75,
          -589672.8125,
          -589673.3125,
          -589729.3125,
          -589761.4375,
          -589833.6875,
          -589843.25,
          -589857.25,
          -589860.125,
          -589870.75,
          -589913.125,
          -589924,
          -589980.9375,
          -589992.25,
          -590005.625,
          -590012,
          -590022.125,
          -590057.75,
          -590080.625,
          -590126.5,
          -590139.375,
          -590151.3125,
          -590161.6875,
          -590169.5,
          -590174.3125,
          -590217.375,
          -590251.875,
          -590306.625,
          -590318.125,
          -590331.1875,
          -590337.6875,
          -590347.8125,
          -590381.8125,
          -590410.5625,
          -590456.8125,
          -590468.75,
          -590482.4375,
          -590489.9375,
          -590502.25,
          -590530.1875,
          -590554.4375,
          -590589.5625,
          -590602.875,
          -590614.6875,
          -590626.4375,
          -590634.8125,
          -590643.5,
          -590643.625,
          -590646.1875,
          -590705.75,
          -590726.5,
          -590748.1875,
          -590750.8125,
          -590775.9375,
          -590794.9375,
          -590803,
          -590825.8125,
          -590841.8125,
          -590853.3125,
          -590871.3125,
          -590873.625,
          -590892.125,
          -590917.25,
          -590925.5,
          -590957.6875,
          -590971.75,
          -590982.375,
          -590995.25,
          -591000.375,
          -591010.875,
          -591043.5,
          -591062.8125,
          -591106.5,
          -591118.1875,
          -591130.6875,
          -591138.5,
          -591149,
          -591149.4375,
          -591154.8125,
          -591205.3125,
          -591225.375,
          -591244.375,
          -591249.3125,
          -591271.3125,
          -591289.3125,
          -591298.5,
          -591320.9375,
          -591335.125,
          -591346.9375,
          -591363.1875,
          -591367.4375,
          -591384.0625,
          -591407.125,
          -591419.9375,
          -591449.4375,
          -591462.3125,
          -591473.25,
          -591485.4375,
          -591492.25,
          -591501.4375,
          -591502,
          -591504.5625,
          -591559.625,
          -591580.75,
          -591601.9375,
          -591604.5,
          -591630.4375,
          -591647.0625,
          -591654.9375,
          -591672.875,
          -591689.1875,
          -591702.6875,
          -591723.1875,
          -591736.375,
          -591749,
          -591763,
          -591770.0625,
          -591783.75,
          -591784.25,
          -591795.1875,
          -591832.375,
          -591836.125,
          -591885.75,
          -591896.1875,
          -591906.5625,
          -591913.25,
          -591920.5,
          -591953.75,
          -591979.25,
          -592022.375,
          -592033.125,
          -592044,
          -592051.9375,
          -592059.8125,
          -592062.125,
          -592062.625,
          -592115.125,
          -592123.6875,
          -592195.9375,
          -592201.4375,
          -592215.9375,
          -592236.75,
          -592251.5625,
          -592277.3125,
          -592289.6875,
          -592299.75,
          -592312.375,
          -592318.0625,
          -592330,
          -592354.1875,
          -592371.8125,
          -592402.25,
          -592414.3125,
          -592424.0625,
          -592435,
          -592441.125,
          -592449.625,
          -592478.6875,
          -592500.4375,
          -592536.75,
          -592547.8125,
          -592558,
          -592567.3125,
          -592574,
          -592579.4375,
          -592612.9375,
          -592643.9375,
          -592687.125,
          -592697.125,
          -592708.0625,
          -592714.375,
          -592722.8125,
          -592723.125,
          -592724.4375,
          -592775.5625,
          -592796,
          -592815.4375,
          -592817.5625,
          -592842.1875,
          -592856.375,
          -592863.625,
          -592880.125,
          -592894.8125,
          -592907.5,
          -592926.625,
          -592938.3125,
          -592950,
          -592962.75,
          -592969.3125,
          -592981.75,
          -592982.4375,
          -592991.25,
          -593025.75,
          -593030.625,
          -593076.75,
          -593086,
          -593096.0625,
          -593101.375,
          -593108.5,
          -593138.4375,
          -593158.375,
          -593197.1875,
          -593206.9375,
          -593216.8125,
          -593224.6875,
          -593230.875,
          -593233.5,
          -593269.6875,
          -593299.5,
          -593345.6875,
          -593354.1875,
          -593365,
          -593369.8125,
          -593377.875,
          -593405.6875,
          -593425.625,
          -593463.5,
          -593472.9375,
          -593483.75,
          -593489.4375,
          -593499.4375,
          -593522.1875,
          -593541.375,
          -593570.4375,
          -593581.0625,
          -593590.3125,
          -593599.9375,
          -593606,
          -593612.6875,
          -593639.625,
          -593665.0625,
          -593699.125,
          -593709.3125,
          -593718.625,
          -593726.125,
          -593733.375,
          -593736.6875,
          -593737.6875,
          -593780.0625,
          -593794.9375,
          -593853.5625,
          -593858.9375,
          -593871.75,
          -593888.625,
          -593902.1875,
          -593922.75,
          -593933.75,
          -593943.0625,
          -593954.1875,
          -593959.4375,
          -593970.6875,
          -593990.4375,
          -594005.625,
          -594030.0625,
          -594040.4375,
          -594049.5625,
          -594059.3125,
          -594064.875,
          -594073.875,
          -594096.5,
          -594115.3125,
          -594143.9375,
          -594154.3125,
          -594163.25,
          -594171.9375,
          -594177.5625,
          -594183.5,
          -594210.1875,
          -594235.25,
          -594269.625,
          -594278.625,
          -594288.0625,
          -594295,
          -594301.625,
          -594304.25,
          -594305.125,
          -594346.9375,
          -594356.1875,
          -594413.25,
          -594418.375,
          -594430.5,
          -594446.75,
          -594459.875,
          -594480.25,
          -594490.625,
          -594499.5,
          -594510.3125,
          -594514.9375,
          -594525.0625,
          -594544.625,
          -594559.125,
          -594583.5625,
          -594593.625,
          -594602.25,
          -594611.5,
          -594616.375,
          -594624.5,
          -594646.8125,
          -594665.75,
          -594694.75,
          -594704.125,
          -594712.625,
          -594720.75,
          -594726.25,
          -594730.875,
          -594730.9375,
          -594768.75,
          -594795.875,
          -594848,
          -594853.25,
          -594865.6875,
          -594880,
          -594891.75,
          -594908.8125,
          -594918.6875,
          -594927.8125,
          -594938.8125,
          -594943.125,
          -594954.625,
          -594970.625,
          -594983.3125,
          -595003.4375,
          -595013.6875,
          -595021.625,
          -595031.6875,
          -595036.4375,
          -595045.625,
          -595064.625,
          -595079.125,
          -595103.25,
          -595112.3125,
          -595120.5625,
          -595129.3125,
          -595134.375,
          -595141.375,
          -595163.9375,
          -595182.125,
          -595210.5625,
          -595219.375,
          -595227.375,
          -595234.5625,
          -595240.375,
          -595244.6875,
          -595271,
          -595296.375,
          -595330.25,
          -595338.4375,
          -595347.125,
          -595352.3125,
          -595359,
          -595359.6875,
          -595361.0625,
          -595400.9375,
          -595401.125,
          -595456,
          -595460.25,
          -595471,
          -595487.1875,
          -595499.125,
          -595519.75,
          -595528.875,
          -595536.6875,
          -595546.0625,
          -595550.6875,
          -595559.1875,
          -595578.125,
          -595592.5625,
          -595616.625,
          -595625.625,
          -595633.1875,
          -595641.5,
          -595646.1875,
          -595652.125,
          -595674.75,
          -595693.5,
          -595722.1875,
          -595730.375,
          -595738.25,
          -595744.75,
          -595750.5,
          -595753.8125,
          -595754.125,
          -595789.8125,
          -595804.125,
          -595853.9375,
          -595858.4375,
          -595869.5625,
          -595883.25,
          -595894.125,
          -595911.125,
          -595920.125,
          -595928.3125,
          -595938,
          -595942.25,
          -595951.9375,
          -595968.125,
          -595980,
          -596000.125,
          -596009,
          -596016.6875,
          -596025.4375,
          -596029.625,
          -596037.25,
          -596056.125,
          -596070.75,
          -596094.4375,
          -596103.25,
          -596110.5625,
          -596117.75,
          -596122.5625,
          -596128.0625,
          -596150.25,
          -596169.4375,
          -596197.875,
          -596206,
          -596213.875,
          -596219.125,
          -596225.25,
          -596227.125,
          -596227.6875,
          -596262.8125,
          -596272.3125,
          -596321.0625,
          -596325.125,
          -596335.6875,
          -596349.5625,
          -596359.9375,
          -596376.8125,
          -596385,
          -596393,
          -596402.0625,
          -596406.3125,
          -596414.875,
          -596431.125,
          -596442.5625,
          -596463.0625,
          -596471.0625,
          -596478.625,
          -596486.625,
          -596490.9375,
          -596497.6875,
          -596516.4375,
          -596530.875,
          -596555.1875,
          -596562.6875,
          -596569.8125,
          -596576.4375,
          -596581.25,
          -596585.8125,
          -596608.1875,
          -596627.3125,
          -596655.875,
          -596663.1875,
          -596670.9375,
          -596676,
          -596681.375,
          -596683.0625,
          -596683.9375,
          -596718.4375,
          -596723.75,
          -596771.0625,
          -596775.4375,
          -596785.1875,
          -596798.75,
          -596808.625,
          -596825.8125,
          -596834.0625,
          -596841.0625,
          -596849.4375,
          -596853.4375,
          -596861.875,
          -596877.75,
          -596889,
          -596909.25,
          -596917.25,
          -596924.25,
          -596931.3125,
          -596935.5,
          -596941.4375,
          -596960.1875,
          -596974.9375,
          -596998.75,
          -597006.25,
          -597013.1875,
          -597019.625,
          -597023.8125,
          -597027.8125,
          -597050.1875,
          -597070.3125,
          -597098.25,
          -597104.75,
          -597112.1875,
          -597117.125,
          -597122.1875,
          -597123.625,
          -597123.9375,
          -597157.8125,
          -597160.1875,
          -597206.25,
          -597209.9375,
          -597219.3125,
          -597232.875,
          -597242.875,
          -597259.625,
          -597267.5625,
          -597274.25,
          -597282.125,
          -597285.9375,
          -597293.0625,
          -597309.25,
          -597320.5,
          -597340.5625,
          -597348.5625,
          -597354.8125,
          -597362,
          -597366,
          -597371.25,
          -597390,
          -597404.625,
          -597428.6875,
          -597435.875,
          -597442.6875,
          -597448.4375,
          -597452.75,
          -597455.9375,
          -597478.1875,
          -597499.25,
          -597527.6875,
          -597533.8125,
          -597541.25,
          -597545.5,
          -597550.6875,
          -597550.875,
          -597551.9375,
          -597585.375,
          -597598.25,
          -597610.625,
          -597612.125,
          -597628.125,
          -597637.5625,
          -597642.0625,
          -597652.6875,
          -597662.875,
          -597670.625,
          -597683.5625,
          -597691.4375,
          -597698.6875,
          -597706.8125,
          -597711.3125,
          -597719.25,
          -597719.6875,
          -597725.6875,
          -597748.375,
          -597751.4375,
          -597781.875,
          -597787.8125,
          -597794.1875,
          -597797.6875,
          -597802.375,
          -597822.3125,
          -597835.0625,
          -597860.3125,
          -597866.9375,
          -597873.3125,
          -597878.375,
          -597882.8125,
          -597884.8125,
          -597908.3125,
          -597927.75,
          -597958,
          -597963.75,
          -597970.8125,
          -597974.25,
          -597979.3125,
          -597997.6875,
          -598010.8125,
          -598035.8125,
          -598041.8125,
          -598049.125,
          -598053.0625,
          -598059.5625,
          -598074.5625,
          -598087.4375,
          -598106.4375,
          -598113.8125,
          -598119.8125,
          -598126.4375,
          -598130.375,
          -598135,
          -598135.375,
          -598135.8125,
          -598168.375,
          -598179.625,
          -598191.75,
          -598192.4375,
          -598206.25,
          -598216.625,
          -598220.4375,
          -598233.0625,
          -598241.6875,
          -598248.125,
          -598257.8125,
          -598258.25,
          -598268.75,
          -598282.125,
          -598286.25,
          -598303.5,
          -598311.0625,
          -598316.25,
          -598323.8125,
          -598326.3125,
          -598332.125,
          -598349.625,
          -598359.125,
          -598382.8125,
          -598389,
          -598395.6875,
          -598400,
          -598405.4375,
          -598406.0625,
          -598408.6875,
          -598436.375,
          -598447.1875,
          -598457.6875,
          -598460.0625,
          -598472.1875,
          -598481.8125,
          -598487.25,
          -598498.5,
          -598506.375,
          -598512.875,
          -598522.1875,
          -598524.375,
          -598533.1875,
          -598545.5625,
          -598552.375,
          -598568.375,
          -598575.375,
          -598581.375,
          -598588.375,
          -598591.8125,
          -598597.0625,
          -598613.125,
          -598628.25,
          -598649.625,
          -598656.25,
          -598663.1875,
          -598667.3125,
          -598673.625,
          -598674.0625,
          -598678.5625,
          -598702.25,
          -598702.6875,
          -598734.0625,
          -598739.4375,
          -598744.5625,
          -598748.125,
          -598750.5,
          -598772.1875,
          -598792.125,
          -598821.5625,
          -598827.0625,
          -598835.125,
          -598836.625,
          -598844.375,
          -598859.625,
          -598863.9375,
          -598883.375,
          -598890.125,
          -598895.1875,
          -598902.0625,
          -598903.4375,
          -598909.25,
          -598927.5,
          -598934.25,
          -598957.5,
          -598963.625,
          -598969.0625,
          -598974.0625,
          -598976.9375,
          -598979.5625,
          -599001.625,
          -599012.5,
          -599041.3125,
          -599046.9375,
          -599052.4375,
          -599055.4375,
          -599059.875,
          -599077.9375,
          -599090.875,
          -599114.875,
          -599120.875,
          -599127.5625,
          -599131,
          -599137.1875,
          -599151.1875,
          -599161.6875,
          -599180.1875,
          -599186.4375,
          -599192.25,
          -599197.8125,
          -599201.5625,
          -599205.625,
          -599222.5,
          -599237.5,
          -599259.5,
          -599265.5625,
          -599271.8125,
          -599275.8125,
          -599280.0625,
          -599282.0625,
          -599282.625,
          -599309.6875,
          -599315.375,
          -599351.75,
          -599355.25,
          -599362.625,
          -599373.4375,
          -599382.4375,
          -599395.875,
          -599402.8125,
          -599408.25,
          -599414.875,
          -599418.8125,
          -599425,
          -599437.9375,
          -599448.625,
          -599464.625,
          -599471.3125,
          -599476.8125,
          -599482.8125,
          -599486.625,
          -599491.125,
          -599491.8125,
          -599492.625,
          -599520.25,
          -599530.75,
          -599541.1875,
          -599542.5,
          -599554.5625,
          -599563.5,
          -599567.375,
          -599578.25,
          -599585.6875,
          -599591.4375,
          -599600.125,
          -599601.25,
          -599610.4375,
          -599621.75,
          -599626.1875,
          -599640.8125,
          -599647.625,
          -599652.4375,
          -599659.3125,
          -599661.3125,
          -599667.1875,
          -599682.125,
          -599690.25,
          -599710.0625,
          -599716.0625,
          -599721.5,
          -599726.0625,
          -599730.1875,
          -599731.5625,
          -599733.625,
          -599757.875,
          -599761,
          -599792.75,
          -599797,
          -599802.4375,
          -599804.1875,
          -599807.6875,
          -599827.375,
          -599834.4375,
          -599861.125,
          -599866,
          -599872.0625,
          -599874,
          -599879.25,
          -599894.5,
          -599901.625,
          -599922.125,
          -599927.8125,
          -599933,
          -599937.5625,
          -599940.1875,
          -599943.0625,
          -599961.8125,
          -599974,
          -599998.8125,
          -600003.6875,
          -600009.5625,
          -600012.5,
          -600016.6875,
          -600032.0625,
          -600043.5625,
          -600064.625,
          -600069.8125,
          -600075.5625,
          -600078.875,
          -600084.25,
          -600096.8125,
          -600108.375,
          -600124.6875,
          -600130.375,
          -600135.4375,
          -600140.75,
          -600144.25,
          -600147.875,
          -600148.4375,
          -600169.1875,
          -600188.875,
          -600217.9375,
          -600221.3125,
          -600229.0625,
          -600237.1875,
          -600245.375,
          -600255.1875,
          -600261.375,
          -600267.1875,
          -600273.5625,
          -600277.125,
          -600283.8125,
          -600293,
          -600302.6875,
          -600315,
          -600320.5,
          -600326.5,
          -600332.0625,
          -600335.4375,
          -600340.75,
          -600341.5625,
          -600344.9375,
          -600364.9375,
          -600366.9375,
          -600392.9375,
          -600398,
          -600402.5,
          -600405.1875,
          -600407.6875,
          -600425.375,
          -600438.8125,
          -600462.8125,
          -600467.375,
          -600473.6875,
          -600475.1875,
          -600481.5625,
          -600494,
          -600499.8125,
          -600516.6875,
          -600522.375,
          -600526.3125,
          -600531.6875,
          -600533.9375,
          -600537.75,
          -600553.875,
          -600562.625,
          -600583.5,
          -600588.375,
          -600593.3125,
          -600597.125,
          -600600.1875,
          -600601,
          -600620.1875,
          -600638.6875,
          -600664.0625,
          -600668.125,
          -600674.125,
          -600675.6875,
          -600680.4375,
          -600695,
          -600700.625,
          -600719.6875,
          -600724.6875,
          -600729.6875,
          -600732.8125,
          -600737,
          -600737.375,
          -600738.5625,
          -600762,
          -600762.6875,
          -600793.875,
          -600797.125,
          -600802.25,
          -600803.125,
          -600806.3125,
          -600825,
          -600827.75,
          -600853.25,
          -600857.0625,
          -600862.375,
          -600864.125,
          -600867.75,
          -600883.75,
          -600889.5,
          -600909.4375,
          -600914.375,
          -600919,
          -600922.4375,
          -600925.3125,
          -600926.25,
          -600945.4375,
          -600958.25,
          -600982.625,
          -600986.8125,
          -600992.3125,
          -600994.1875,
          -600998.625,
          -601012.9375,
          -601018.6875,
          -601037.875,
          -601042.5,
          -601047.5,
          -601050.6875,
          -601054.5625,
          -601054.6875,
          -601055.5625,
          -601078.875,
          -601087.625,
          -601095.9375,
          -601097.1875,
          -601107.3125,
          -601114.9375,
          -601118,
          -601127.0625,
          -601133.125,
          -601138.0625,
          -601145.8125,
          -601146.3125,
          -601154.625,
          -601164.0625,
          -601167.375,
          -601179.1875,
          -601184.9375,
          -601188.9375,
          -601195.0625,
          -601196.5,
          -601201.6875,
          -601213.9375,
          -601219.375,
          -601235.5625,
          -601240.625,
          -601245.0625,
          -601248.9375,
          -601252.5625,
          -601254,
          -601254.625,
          -601275.75,
          -601281.75,
          -601309.5,
          -601312.75,
          -601317.9375,
          -601318.4375,
          -601322.5,
          -601338.3125,
          -601341.5,
          -601362.5625,
          -601367.1875,
          -601371.25,
          -601374,
          -601377.3125,
          -601390.9375,
          -601400.75,
          -601418.625,
          -601423.125,
          -601427.9375,
          -601431.4375,
          -601435,
          -601435.5625,
          -601436.0625,
          -601457.9375,
          -601459.6875,
          -601489.8125,
          -601492.5,
          -601498.125,
          -601507.125,
          -601514.4375,
          -601525.625,
          -601530.875,
          -601535.1875,
          -601540.5,
          -601543.5625,
          -601547.25,
          -601547.375,
          -601549.5,
          -601569.4375,
          -601576.6875,
          -601584.5,
          -601586.375,
          -601595.0625,
          -601601.625,
          -601605.6875,
          -601613.75,
          -601619.625,
          -601624.3125,
          -601630.875,
          -601632.3125,
          -601639.0625,
          -601648.125,
          -601652.8125,
          -601663.875,
          -601669.125,
          -601673.25,
          -601678.75,
          -601680.75,
          -601685.1875,
          -601696,
          -601704.8125,
          -601719.1875,
          -601724.1875,
          -601729.0625,
          -601732.1875,
          -601736.25,
          -601737.6875,
          -601739.25,
          -601756.9375,
          -601761.5,
          -601784.875,
          -601788.5,
          -601793,
          -601794.875,
          -601797.5,
          -601811.625,
          -601818.9375,
          -601837.9375,
          -601842.125,
          -601847.1875,
          -601849.3125,
          -601853.3125,
          -601864.9375,
          -601872.9375,
          -601888,
          -601892.3125,
          -601896.6875,
          -601900.5625,
          -601903.25,
          -601905.125,
          -601905.625,
          -601925,
          -601933.1875,
          -601959.875,
          -601962.5,
          -601968.125,
          -601975.6875,
          -601982.3125,
          -601991.75,
          -601996.8125,
          -602001.1875,
          -602006.125,
          -602009,
          -602013.5625,
          -602022.5625,
          -602031.3125,
          -602043.0625,
          -602047.6875,
          -602051.6875,
          -602056.375,
          -602059.25,
          -602062.5625,
          -602063.25,
          -602064.25,
          -602084.1875,
          -602085.25,
          -602111.25,
          -602114.5,
          -602118.8125,
          -602119.625,
          -602122.25,
          -602137.875,
          -602140.6875,
          -602161.5625,
          -602165.0625,
          -602169.5,
          -602171,
          -602174.5,
          -602187.75,
          -602193.625,
          -602210.5625,
          -602214.5625,
          -602218.875,
          -602221.6875,
          -602224.3125,
          -602225.3125,
          -602241.5
         ]
        }
       ],
       "layout": {
        "title": {
         "text": "Energy Minimization"
        },
        "xaxis": {
         "title": {
          "text": "Energy Minimization Step"
         }
        },
        "yaxis": {
         "title": {
          "text": "Potential Energy KJ/mol-1"
         }
        }
       }
      },
      "text/html": [
       "<div>\n",
       "        \n",
       "        \n",
       "            <div id=\"ad609f6e-83e5-4eb3-bccf-8446009b8f33\" class=\"plotly-graph-div\" style=\"height:525px; width:100%;\"></div>\n",
       "            <script type=\"text/javascript\">\n",
       "                require([\"plotly\"], function(Plotly) {\n",
       "                    window.PLOTLYENV=window.PLOTLYENV || {};\n",
       "                    window.PLOTLYENV.BASE_URL='https://plot.ly';\n",
       "                    \n",
       "                if (document.getElementById(\"ad609f6e-83e5-4eb3-bccf-8446009b8f33\")) {\n",
       "                    Plotly.newPlot(\n",
       "                        'ad609f6e-83e5-4eb3-bccf-8446009b8f33',\n",
       "                        [{\"type\": \"scatter\", \"uid\": \"35cce87e-0e89-46fd-963c-0961ffc0cbd6\", \"x\": [0.0, 1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 11.0, 12.0, 13.0, 15.0, 16.0, 17.0, 18.0, 19.0, 21.0, 22.0, 23.0, 25.0, 26.0, 27.0, 28.0, 29.0, 31.0, 33.0, 34.0, 35.0, 36.0, 38.0, 39.0, 40.0, 42.0, 43.0, 44.0, 45.0, 46.0, 48.0, 49.0, 50.0, 52.0, 53.0, 54.0, 55.0, 56.0, 58.0, 59.0, 60.0, 62.0, 63.0, 64.0, 65.0, 67.0, 68.0, 69.0, 71.0, 72.0, 73.0, 74.0, 75.0, 77.0, 78.0, 79.0, 81.0, 82.0, 83.0, 84.0, 86.0, 87.0, 88.0, 90.0, 91.0, 92.0, 93.0, 95.0, 96.0, 97.0, 99.0, 100.0, 101.0, 102.0, 104.0, 105.0, 106.0, 108.0, 109.0, 110.0, 111.0, 112.0, 114.0, 115.0, 116.0, 118.0, 119.0, 120.0, 121.0, 123.0, 124.0, 125.0, 127.0, 128.0, 129.0, 130.0, 132.0, 133.0, 134.0, 136.0, 137.0, 138.0, 139.0, 140.0, 141.0, 143.0, 144.0, 145.0, 147.0, 148.0, 149.0, 150.0, 152.0, 153.0, 154.0, 156.0, 157.0, 158.0, 159.0, 160.0, 162.0, 163.0, 164.0, 166.0, 167.0, 168.0, 169.0, 171.0, 172.0, 173.0, 175.0, 176.0, 177.0, 178.0, 180.0, 181.0, 182.0, 184.0, 185.0, 186.0, 187.0, 189.0, 190.0, 191.0, 193.0, 194.0, 195.0, 196.0, 197.0, 198.0, 200.0, 202.0, 203.0, 204.0, 205.0, 207.0, 208.0, 209.0, 211.0, 212.0, 213.0, 215.0, 216.0, 217.0, 218.0, 219.0, 221.0, 222.0, 223.0, 225.0, 226.0, 227.0, 228.0, 229.0, 231.0, 232.0, 233.0, 235.0, 236.0, 237.0, 238.0, 239.0, 241.0, 242.0, 243.0, 245.0, 246.0, 247.0, 248.0, 249.0, 251.0, 252.0, 253.0, 255.0, 256.0, 257.0, 258.0, 259.0, 261.0, 262.0, 263.0, 265.0, 266.0, 267.0, 268.0, 269.0, 271.0, 272.0, 273.0, 275.0, 276.0, 277.0, 278.0, 280.0, 281.0, 282.0, 284.0, 285.0, 286.0, 287.0, 289.0, 290.0, 291.0, 293.0, 294.0, 295.0, 296.0, 297.0, 299.0, 300.0, 301.0, 303.0, 304.0, 305.0, 306.0, 307.0, 309.0, 310.0, 311.0, 313.0, 314.0, 315.0, 316.0, 317.0, 319.0, 320.0, 321.0, 323.0, 324.0, 325.0, 326.0, 328.0, 329.0, 330.0, 332.0, 333.0, 334.0, 335.0, 337.0, 338.0, 339.0, 341.0, 342.0, 343.0, 344.0, 345.0, 347.0, 348.0, 349.0, 351.0, 352.0, 353.0, 354.0, 355.0, 357.0, 358.0, 359.0, 361.0, 362.0, 363.0, 364.0, 366.0, 367.0, 368.0, 370.0, 371.0, 372.0, 373.0, 375.0, 376.0, 377.0, 379.0, 380.0, 381.0, 382.0, 383.0, 384.0, 386.0, 387.0, 388.0, 390.0, 391.0, 392.0, 393.0, 395.0, 396.0, 397.0, 399.0, 400.0, 401.0, 402.0, 403.0, 405.0, 406.0, 407.0, 409.0, 410.0, 411.0, 412.0, 414.0, 415.0, 416.0, 418.0, 419.0, 420.0, 421.0, 422.0, 423.0, 425.0, 427.0, 428.0, 429.0, 430.0, 432.0, 433.0, 434.0, 436.0, 437.0, 438.0, 440.0, 441.0, 442.0, 443.0, 444.0, 446.0, 447.0, 448.0, 450.0, 451.0, 452.0, 453.0, 454.0, 455.0, 456.0, 458.0, 459.0, 460.0, 462.0, 463.0, 464.0, 465.0, 467.0, 468.0, 469.0, 471.0, 472.0, 473.0, 474.0, 475.0, 477.0, 478.0, 479.0, 481.0, 482.0, 483.0, 484.0, 486.0, 487.0, 488.0, 490.0, 491.0, 492.0, 493.0, 495.0, 496.0, 497.0, 499.0, 500.0, 501.0, 502.0, 503.0, 505.0, 506.0, 507.0, 509.0, 510.0, 511.0, 512.0, 513.0, 515.0, 516.0, 517.0, 519.0, 520.0, 521.0, 522.0, 524.0, 525.0, 526.0, 528.0, 529.0, 530.0, 531.0, 533.0, 534.0, 535.0, 537.0, 538.0, 539.0, 540.0, 541.0, 543.0, 544.0, 545.0, 547.0, 548.0, 549.0, 550.0, 551.0, 553.0, 554.0, 555.0, 557.0, 558.0, 559.0, 560.0, 561.0, 563.0, 564.0, 565.0, 567.0, 568.0, 569.0, 570.0, 572.0, 573.0, 574.0, 576.0, 577.0, 578.0, 579.0, 581.0, 582.0, 583.0, 585.0, 586.0, 587.0, 588.0, 589.0, 591.0, 592.0, 593.0, 595.0, 596.0, 597.0, 598.0, 599.0, 601.0, 602.0, 603.0, 605.0, 606.0, 607.0, 608.0, 610.0, 611.0, 612.0, 614.0, 615.0, 616.0, 617.0, 619.0, 620.0, 621.0, 623.0, 624.0, 625.0, 626.0, 627.0, 629.0, 630.0, 631.0, 633.0, 634.0, 635.0, 636.0, 637.0, 639.0, 640.0, 641.0, 643.0, 644.0, 645.0, 646.0, 648.0, 649.0, 650.0, 652.0, 653.0, 654.0, 655.0, 657.0, 658.0, 659.0, 661.0, 662.0, 663.0, 664.0, 665.0, 666.0, 668.0, 669.0, 670.0, 672.0, 673.0, 675.0, 676.0, 677.0, 679.0, 680.0, 681.0, 682.0, 683.0, 685.0, 686.0, 687.0, 689.0, 690.0, 691.0, 692.0, 693.0, 694.0, 695.0, 697.0, 699.0, 700.0, 701.0, 702.0, 704.0, 705.0, 706.0, 708.0, 709.0, 710.0, 711.0, 712.0, 714.0, 715.0, 716.0, 718.0, 719.0, 720.0, 721.0, 722.0, 724.0, 725.0, 726.0, 728.0, 729.0, 730.0, 731.0, 732.0, 733.0, 735.0, 736.0, 737.0, 739.0, 740.0, 741.0, 742.0, 744.0, 745.0, 746.0, 748.0, 749.0, 750.0, 751.0, 753.0, 754.0, 755.0, 757.0, 758.0, 759.0, 760.0, 761.0, 763.0, 764.0, 765.0, 767.0, 768.0, 769.0, 770.0, 772.0, 773.0, 774.0, 776.0, 777.0, 778.0, 779.0, 781.0, 782.0, 783.0, 785.0, 786.0, 787.0, 788.0, 789.0, 790.0, 792.0, 793.0, 794.0, 796.0, 797.0, 799.0, 800.0, 801.0, 803.0, 804.0, 805.0, 806.0, 807.0, 808.0, 809.0, 811.0, 812.0, 813.0, 815.0, 816.0, 817.0, 818.0, 819.0, 821.0, 822.0, 823.0, 825.0, 826.0, 827.0, 828.0, 830.0, 831.0, 832.0, 834.0, 835.0, 836.0, 837.0, 839.0, 840.0, 841.0, 843.0, 844.0, 845.0, 846.0, 847.0, 849.0, 850.0, 851.0, 853.0, 854.0, 855.0, 856.0, 858.0, 859.0, 860.0, 862.0, 863.0, 864.0, 865.0, 867.0, 868.0, 869.0, 871.0, 872.0, 873.0, 874.0, 875.0, 877.0, 878.0, 879.0, 881.0, 882.0, 883.0, 884.0, 885.0, 886.0, 888.0, 889.0, 890.0, 892.0, 893.0, 895.0, 896.0, 897.0, 899.0, 900.0, 901.0, 902.0, 903.0, 905.0, 906.0, 907.0, 909.0, 910.0, 911.0, 912.0, 913.0, 914.0, 915.0, 917.0, 919.0, 920.0, 921.0, 922.0, 924.0, 925.0, 926.0, 928.0, 929.0, 930.0, 931.0, 932.0, 934.0, 935.0, 936.0, 938.0, 939.0, 940.0, 941.0, 942.0, 944.0, 945.0, 946.0, 948.0, 949.0, 950.0, 951.0, 952.0, 953.0, 955.0, 956.0, 957.0, 959.0, 960.0, 961.0, 962.0, 964.0, 965.0, 966.0, 968.0, 969.0, 970.0, 971.0, 973.0, 974.0, 975.0, 977.0, 978.0, 979.0, 980.0, 981.0, 983.0, 984.0, 985.0, 987.0, 988.0, 989.0, 990.0, 991.0, 993.0, 994.0, 995.0, 997.0, 998.0, 999.0, 1000.0, 1002.0, 1003.0, 1004.0, 1006.0, 1007.0, 1008.0, 1009.0, 1011.0, 1012.0, 1013.0, 1015.0, 1016.0, 1017.0, 1018.0, 1019.0, 1021.0, 1022.0, 1023.0, 1025.0, 1026.0, 1027.0, 1028.0, 1030.0, 1031.0, 1032.0, 1034.0, 1035.0, 1036.0, 1037.0, 1039.0, 1040.0, 1041.0, 1043.0, 1044.0, 1045.0, 1046.0, 1047.0, 1048.0, 1049.0, 1051.0, 1053.0, 1054.0, 1055.0, 1056.0, 1058.0, 1059.0, 1060.0, 1062.0, 1063.0, 1064.0, 1065.0, 1066.0, 1068.0, 1069.0, 1070.0, 1072.0, 1073.0, 1074.0, 1075.0, 1076.0, 1078.0, 1079.0, 1080.0, 1082.0, 1083.0, 1084.0, 1085.0, 1086.0, 1087.0, 1089.0, 1091.0, 1092.0, 1093.0, 1094.0, 1096.0, 1097.0, 1098.0, 1100.0, 1101.0, 1102.0, 1103.0, 1104.0, 1106.0, 1107.0, 1108.0, 1110.0, 1111.0, 1112.0, 1113.0, 1114.0, 1115.0, 1116.0, 1118.0, 1120.0, 1121.0, 1122.0, 1123.0, 1125.0, 1126.0, 1127.0, 1129.0, 1130.0, 1131.0, 1133.0, 1134.0, 1135.0, 1136.0, 1137.0, 1138.0, 1139.0, 1141.0, 1142.0, 1143.0, 1145.0, 1146.0, 1147.0, 1148.0, 1150.0, 1151.0, 1152.0, 1154.0, 1155.0, 1156.0, 1157.0, 1158.0, 1159.0, 1161.0, 1162.0, 1163.0, 1165.0, 1166.0, 1168.0, 1169.0, 1170.0, 1172.0, 1173.0, 1174.0, 1175.0, 1176.0, 1178.0, 1179.0, 1180.0, 1182.0, 1183.0, 1184.0, 1185.0, 1186.0, 1188.0, 1189.0, 1190.0, 1192.0, 1193.0, 1194.0, 1195.0, 1196.0, 1198.0, 1199.0, 1200.0, 1202.0, 1203.0, 1204.0, 1205.0, 1206.0, 1207.0, 1209.0, 1211.0, 1212.0, 1213.0, 1214.0, 1216.0, 1217.0, 1218.0, 1220.0, 1221.0, 1222.0, 1224.0, 1225.0, 1226.0, 1227.0, 1228.0, 1229.0, 1230.0, 1232.0, 1233.0, 1234.0, 1236.0, 1237.0, 1238.0, 1239.0, 1241.0, 1242.0, 1243.0, 1245.0, 1246.0, 1247.0, 1248.0, 1249.0, 1251.0, 1252.0, 1253.0, 1255.0, 1256.0, 1257.0, 1258.0, 1260.0, 1261.0, 1262.0, 1264.0, 1265.0, 1266.0, 1267.0, 1269.0, 1270.0, 1271.0, 1273.0, 1274.0, 1275.0, 1276.0, 1277.0, 1279.0, 1280.0, 1281.0, 1283.0, 1284.0, 1285.0, 1286.0, 1287.0, 1288.0, 1290.0, 1291.0, 1292.0, 1294.0, 1295.0, 1297.0, 1298.0, 1299.0, 1301.0, 1302.0, 1303.0, 1304.0, 1305.0, 1307.0, 1308.0, 1309.0, 1311.0, 1312.0, 1313.0, 1314.0, 1315.0, 1317.0, 1318.0, 1319.0, 1321.0, 1322.0, 1323.0, 1324.0, 1325.0, 1327.0, 1328.0, 1329.0, 1331.0, 1332.0, 1333.0, 1334.0, 1335.0, 1336.0, 1338.0, 1339.0, 1340.0, 1342.0, 1343.0, 1345.0, 1346.0, 1347.0, 1349.0, 1350.0, 1351.0, 1352.0, 1353.0, 1355.0, 1356.0, 1357.0, 1359.0, 1360.0, 1361.0, 1362.0, 1363.0, 1365.0, 1366.0, 1367.0, 1369.0, 1370.0, 1371.0, 1372.0, 1373.0, 1374.0, 1376.0, 1377.0, 1378.0, 1380.0, 1381.0, 1383.0, 1384.0, 1385.0, 1387.0, 1388.0, 1389.0, 1390.0, 1391.0, 1393.0, 1394.0, 1395.0, 1397.0, 1398.0, 1399.0, 1400.0, 1401.0, 1403.0, 1404.0, 1405.0, 1407.0, 1408.0, 1409.0, 1410.0, 1411.0, 1413.0, 1414.0, 1415.0, 1417.0, 1418.0, 1419.0, 1420.0, 1421.0, 1423.0, 1424.0, 1425.0, 1427.0, 1428.0, 1429.0, 1430.0, 1431.0, 1432.0, 1434.0, 1435.0, 1436.0, 1438.0, 1439.0, 1441.0, 1442.0, 1443.0, 1445.0, 1446.0, 1447.0, 1448.0, 1449.0, 1451.0, 1452.0, 1453.0, 1455.0, 1456.0, 1457.0, 1458.0, 1459.0, 1461.0, 1462.0, 1463.0, 1465.0, 1466.0, 1467.0, 1468.0, 1469.0, 1470.0, 1472.0, 1473.0, 1474.0, 1476.0, 1477.0, 1479.0, 1480.0, 1481.0, 1483.0, 1484.0, 1485.0, 1486.0, 1487.0, 1489.0, 1490.0, 1491.0, 1493.0, 1494.0, 1495.0, 1496.0, 1497.0, 1499.0, 1500.0, 1501.0, 1503.0, 1504.0, 1505.0, 1506.0, 1507.0, 1509.0, 1510.0, 1511.0, 1513.0, 1514.0, 1515.0, 1516.0, 1517.0, 1518.0, 1520.0, 1521.0, 1522.0, 1524.0, 1525.0, 1527.0, 1528.0, 1529.0, 1531.0, 1532.0, 1533.0, 1534.0, 1535.0, 1537.0, 1538.0, 1539.0, 1541.0, 1542.0, 1543.0, 1544.0, 1545.0, 1547.0, 1548.0, 1549.0, 1551.0, 1552.0, 1553.0, 1554.0, 1555.0, 1557.0, 1558.0, 1559.0, 1561.0, 1562.0, 1563.0, 1564.0, 1565.0, 1566.0, 1568.0, 1569.0, 1570.0, 1572.0, 1573.0, 1575.0, 1576.0, 1577.0, 1579.0, 1580.0, 1581.0, 1582.0, 1583.0, 1585.0, 1586.0, 1587.0, 1589.0, 1590.0, 1591.0, 1592.0, 1593.0, 1595.0, 1596.0, 1597.0, 1599.0, 1600.0, 1601.0, 1602.0, 1603.0, 1605.0, 1606.0, 1607.0, 1609.0, 1610.0, 1611.0, 1612.0, 1613.0, 1614.0, 1616.0, 1617.0, 1618.0, 1620.0, 1621.0, 1623.0, 1624.0, 1625.0, 1627.0, 1628.0, 1629.0, 1630.0, 1631.0, 1633.0, 1634.0, 1635.0, 1637.0, 1638.0, 1639.0, 1640.0, 1641.0, 1643.0, 1644.0, 1645.0, 1647.0, 1648.0, 1649.0, 1650.0, 1651.0, 1653.0, 1654.0, 1655.0, 1657.0, 1658.0, 1659.0, 1660.0, 1661.0, 1662.0, 1664.0, 1666.0, 1667.0, 1668.0, 1669.0, 1671.0, 1672.0, 1673.0, 1675.0, 1676.0, 1677.0, 1679.0, 1680.0, 1681.0, 1682.0, 1683.0, 1684.0, 1685.0, 1687.0, 1688.0, 1689.0, 1691.0, 1692.0, 1693.0, 1694.0, 1696.0, 1697.0, 1698.0, 1700.0, 1701.0, 1702.0, 1703.0, 1704.0, 1706.0, 1707.0, 1708.0, 1710.0, 1711.0, 1712.0, 1713.0, 1715.0, 1716.0, 1717.0, 1719.0, 1720.0, 1721.0, 1722.0, 1724.0, 1725.0, 1726.0, 1728.0, 1729.0, 1730.0, 1731.0, 1732.0, 1733.0, 1734.0, 1736.0, 1738.0, 1739.0, 1740.0, 1741.0, 1743.0, 1744.0, 1745.0, 1747.0, 1748.0, 1749.0, 1750.0, 1751.0, 1753.0, 1754.0, 1755.0, 1757.0, 1758.0, 1759.0, 1760.0, 1761.0, 1763.0, 1764.0, 1765.0, 1767.0, 1768.0, 1769.0, 1770.0, 1771.0, 1772.0, 1774.0, 1776.0, 1777.0, 1778.0, 1779.0, 1781.0, 1782.0, 1783.0, 1785.0, 1786.0, 1787.0, 1788.0, 1789.0, 1791.0, 1792.0, 1793.0, 1795.0, 1796.0, 1797.0, 1798.0, 1799.0, 1801.0, 1802.0, 1803.0, 1805.0, 1806.0, 1807.0, 1808.0, 1809.0, 1810.0, 1812.0, 1813.0, 1814.0, 1816.0, 1817.0, 1818.0, 1819.0, 1821.0, 1822.0, 1823.0, 1825.0, 1826.0, 1827.0, 1828.0, 1830.0, 1831.0, 1832.0, 1834.0, 1835.0, 1836.0, 1837.0, 1838.0, 1840.0, 1841.0, 1842.0, 1844.0, 1845.0, 1846.0, 1847.0, 1848.0, 1850.0, 1851.0, 1852.0, 1854.0, 1855.0, 1856.0, 1857.0, 1859.0, 1860.0, 1861.0, 1863.0, 1864.0, 1865.0, 1866.0, 1868.0, 1869.0, 1870.0, 1872.0, 1873.0, 1874.0, 1875.0, 1876.0, 1878.0, 1879.0, 1880.0, 1882.0, 1883.0, 1884.0, 1885.0, 1886.0, 1887.0, 1889.0, 1890.0, 1891.0, 1893.0, 1894.0, 1896.0, 1897.0, 1898.0, 1900.0, 1901.0, 1902.0, 1903.0, 1904.0, 1906.0, 1907.0, 1908.0, 1910.0, 1911.0, 1912.0, 1913.0, 1914.0, 1915.0, 1916.0, 1918.0, 1920.0, 1921.0, 1922.0, 1923.0, 1925.0, 1926.0, 1927.0, 1929.0, 1930.0, 1931.0, 1932.0, 1933.0, 1935.0, 1936.0, 1937.0, 1939.0, 1940.0, 1941.0, 1942.0, 1943.0, 1945.0, 1946.0, 1947.0, 1949.0, 1950.0, 1951.0, 1952.0, 1953.0, 1954.0, 1956.0, 1957.0, 1958.0, 1960.0, 1961.0, 1962.0, 1963.0, 1965.0, 1966.0, 1967.0, 1969.0, 1970.0, 1971.0, 1972.0, 1974.0, 1975.0, 1976.0, 1978.0, 1979.0, 1980.0, 1981.0, 1982.0, 1984.0, 1985.0, 1986.0, 1988.0, 1989.0, 1990.0, 1991.0, 1993.0, 1994.0, 1995.0, 1997.0, 1998.0, 1999.0, 2000.0, 2002.0, 2003.0, 2004.0, 2006.0, 2007.0, 2008.0, 2009.0, 2010.0, 2011.0, 2013.0, 2014.0, 2015.0, 2017.0, 2018.0, 2020.0, 2021.0, 2022.0, 2024.0, 2025.0, 2026.0, 2027.0, 2028.0, 2030.0, 2031.0, 2032.0, 2034.0, 2035.0, 2036.0, 2037.0, 2038.0, 2039.0, 2040.0, 2042.0, 2043.0, 2044.0, 2046.0, 2047.0, 2048.0, 2049.0, 2051.0, 2052.0, 2053.0, 2055.0, 2056.0, 2057.0, 2058.0, 2060.0, 2061.0, 2062.0, 2064.0, 2065.0, 2066.0, 2067.0, 2068.0, 2070.0, 2071.0, 2072.0, 2074.0, 2075.0, 2076.0, 2077.0, 2078.0, 2080.0, 2081.0, 2082.0, 2084.0, 2085.0, 2086.0, 2087.0, 2089.0, 2090.0, 2091.0, 2093.0, 2094.0, 2095.0, 2096.0, 2097.0, 2098.0, 2100.0, 2101.0, 2102.0, 2104.0, 2105.0, 2106.0, 2107.0, 2109.0, 2110.0, 2111.0, 2113.0, 2114.0, 2115.0, 2116.0, 2118.0, 2119.0, 2120.0, 2122.0, 2123.0, 2124.0, 2125.0, 2126.0, 2128.0, 2129.0, 2130.0, 2132.0, 2133.0, 2134.0, 2135.0, 2137.0, 2138.0, 2139.0, 2141.0, 2142.0, 2143.0, 2144.0, 2145.0, 2146.0, 2148.0, 2150.0, 2151.0, 2152.0, 2153.0, 2155.0, 2156.0, 2157.0, 2159.0, 2160.0, 2161.0, 2162.0, 2163.0, 2165.0, 2166.0, 2167.0, 2169.0, 2170.0, 2171.0, 2172.0, 2173.0, 2175.0, 2176.0, 2177.0, 2179.0, 2180.0, 2181.0, 2182.0, 2183.0, 2184.0, 2186.0, 2187.0, 2188.0, 2190.0, 2191.0, 2192.0, 2193.0, 2195.0, 2196.0, 2197.0, 2199.0, 2200.0, 2201.0, 2202.0, 2204.0, 2205.0, 2206.0, 2208.0, 2209.0, 2210.0, 2211.0, 2212.0, 2213.0, 2215.0, 2216.0, 2217.0, 2219.0, 2220.0, 2222.0, 2223.0, 2224.0, 2226.0, 2227.0, 2228.0, 2229.0, 2230.0, 2231.0, 2232.0, 2234.0, 2236.0, 2237.0, 2238.0, 2239.0, 2241.0, 2242.0, 2243.0, 2245.0, 2246.0, 2247.0, 2248.0, 2249.0, 2251.0, 2252.0, 2253.0, 2255.0, 2256.0, 2257.0, 2258.0, 2259.0, 2261.0, 2262.0, 2263.0, 2265.0, 2266.0, 2267.0, 2268.0, 2269.0, 2270.0, 2272.0, 2273.0, 2274.0, 2276.0, 2277.0, 2278.0, 2279.0, 2281.0, 2282.0, 2283.0, 2285.0, 2286.0, 2287.0, 2288.0, 2290.0, 2291.0, 2292.0, 2294.0, 2295.0, 2296.0, 2297.0, 2298.0, 2299.0, 2301.0, 2302.0, 2303.0, 2305.0, 2306.0, 2308.0, 2309.0, 2310.0, 2312.0, 2313.0, 2314.0, 2315.0, 2316.0, 2318.0, 2319.0, 2320.0, 2322.0, 2323.0, 2324.0, 2325.0, 2326.0, 2327.0, 2328.0, 2330.0, 2331.0, 2332.0, 2334.0, 2335.0, 2336.0, 2337.0, 2339.0, 2340.0, 2341.0, 2343.0, 2344.0, 2345.0, 2346.0, 2348.0, 2349.0, 2350.0, 2352.0, 2353.0, 2354.0, 2355.0, 2356.0, 2358.0], \"y\": [-327612.125, -347306.34375, -371487.59375, -400072.09375, -427773.65625, -452044.53125, -457291.3125, -461900.9375, -463964.40625, -467420.09375, -472346.78125, -473178.28125, -479740.1875, -482194.40625, -483604.125, -485770.09375, -486202.65625, -487951.03125, -491239.78125, -491569.21875, -496079.75, -497212.5625, -497915.5, -498762.71875, -498798.46875, -499045.03125, -503261.5625, -504590.28125, -505913.375, -506152.90625, -507573.75, -508783.5, -509315.84375, -510723.34375, -511568.875, -512161.84375, -513046.96875, -513196.65625, -514056.8125, -515344.4375, -515769.0, -517366.0, -517970.4375, -518392.78125, -518939.5625, -519113.40625, -519501.9375, -520976.40625, -521471.625, -523378.9375, -523770.625, -524163.09375, -524394.5625, -524645.0, -525844.3125, -526626.375, -528134.0625, -528465.9375, -528803.875, -529018.625, -529233.0625, -529242.25, -530579.875, -531357.625, -533034.6875, -533255.4375, -533526.75, -533612.9375, -533750.875, -534756.625, -534980.1875, -536368.8125, -536553.1875, -536811.1875, -536874.375, -537042.9375, -537863.8125, -538096.5, -539172.5, -539360.3125, -539594.5625, -539680.875, -539867.625, -540479.6875, -540767.5625, -541523.375, -541727.0, -541907.125, -542068.1875, -542175.0625, -542254.5625, -542921.125, -543344.0, -544164.375, -544323.1875, -544502.125, -544596.25, -544721.375, -545209.6875, -545588.875, -546230.375, -546383.0625, -546559.5625, -546649.875, -546800.5625, -547169.8125, -547469.75, -547933.0625, -548090.75, -548234.5625, -548365.4375, -548466.25, -548545.4375, -548559.375, -549138.6875, -549492.125, -550260.4375, -550351.6875, -550511.5, -550521.1875, -550657.0, -551055.5625, -551097.4375, -551613.375, -551739.4375, -551837.5, -551935.0, -551975.25, -552010.375, -552502.125, -552619.6875, -553261.875, -553348.5, -553464.625, -553493.0, -553570.0625, -553953.75, -554042.0625, -554555.0, -554648.125, -554751.625, -554800.875, -554869.5, -555197.625, -555378.875, -555810.75, -555902.8125, -556005.75, -556059.4375, -556136.0, -556407.5, -556606.0, -556965.0, -557057.625, -557153.75, -557216.5, -557288.75, -557297.3125, -557313.6875, -557746.875, -557914.875, -558077.875, -558093.625, -558289.8125, -558423.1875, -558476.375, -558630.9375, -558754.8125, -558838.6875, -558985.1875, -559090.875, -559186.0625, -559304.875, -559339.3125, -559465.4375, -559616.9375, -559699.25, -559890.0, -559985.875, -560059.4375, -560159.0, -560189.8125, -560289.25, -560469.0625, -560558.625, -560776.875, -560867.125, -560932.6875, -561023.875, -561048.125, -561133.75, -561330.4375, -561408.8125, -561650.6875, -561733.9375, -561794.1875, -561874.0, -561895.625, -561963.75, -562182.375, -562251.125, -562520.8125, -562597.125, -562654.1875, -562721.4375, -562741.75, -562789.5, -563033.1875, -563097.3125, -563400.1875, -563467.0625, -563523.0625, -563575.375, -563597.75, -563621.1875, -563894.8125, -563959.4375, -564299.75, -564356.6875, -564413.625, -564448.9375, -564474.25, -564699.375, -564888.0, -565190.75, -565242.0625, -565323.3125, -565336.75, -565417.875, -565566.4375, -565614.1875, -565797.0625, -565867.5, -565914.9375, -565984.875, -565999.125, -566061.625, -566231.1875, -566280.5, -566492.6875, -566555.625, -566602.625, -566659.25, -566676.9375, -566716.9375, -566912.3125, -566971.5, -567216.875, -567271.4375, -567320.4375, -567361.875, -567385.5, -567400.0, -567624.125, -567708.8125, -567990.25, -568034.75, -568087.5625, -568111.875, -568142.4375, -568319.3125, -568429.3125, -568665.6875, -568710.6875, -568771.5625, -568791.5, -568847.375, -568977.5, -569041.375, -569204.5, -569259.9375, -569304.5, -569353.9375, -569377.75, -569413.875, -569565.0625, -569651.875, -569843.4375, -569891.75, -569938.0625, -569974.0, -570004.25, -570017.25, -570193.375, -570331.0625, -570553.5625, -570594.25, -570644.0625, -570664.875, -570702.1875, -570835.875, -570920.4375, -571097.5, -571141.3125, -571189.75, -571218.0625, -571257.5625, -571367.625, -571469.5625, -571608.875, -571655.25, -571700.0, -571738.375, -571771.125, -571792.75, -571801.5, -571975.0625, -572077.0625, -572307.1875, -572338.5625, -572389.9375, -572396.5, -572441.3125, -572565.1875, -572592.0, -572751.8125, -572795.1875, -572830.125, -572864.6875, -572882.0, -572898.75, -573051.9375, -573120.9375, -573320.75, -573353.8125, -573396.9375, -573412.125, -573444.4375, -573564.3125, -573620.8125, -573778.0625, -573816.4375, -573854.9375, -573881.0, -573908.75, -573912.8125, -573915.8125, -574106.875, -574182.1875, -574253.25, -574259.8125, -574346.25, -574402.8125, -574426.6875, -574491.6875, -574545.5625, -574584.9375, -574649.125, -574693.8125, -574737.25, -574787.1875, -574807.125, -574859.4375, -574924.1875, -574971.4375, -575053.0625, -575095.0, -575131.0, -575172.4375, -575194.8125, -575231.5625, -575232.0625, -575253.0, -575399.5625, -575400.6875, -575596.9375, -575624.25, -575659.75, -575668.9375, -575693.375, -575814.1875, -575844.6875, -576000.5, -576033.125, -576064.25, -576086.9375, -576103.0625, -576106.8125, -576252.8125, -576313.625, -576500.75, -576526.25, -576560.8125, -576570.75, -576593.0625, -576706.0, -576748.25, -576901.625, -576929.4375, -576966.5625, -576977.9375, -577009.875, -577098.875, -577136.9375, -577249.625, -577284.0625, -577311.0, -577341.5625, -577353.8125, -577374.9375, -577479.625, -577520.9375, -577653.625, -577684.4375, -577711.625, -577735.125, -577748.6875, -577758.125, -577880.4375, -577925.1875, -578079.75, -578106.125, -578134.1875, -578150.375, -578164.5, -578266.0, -578347.0625, -578487.6875, -578512.625, -578551.375, -578557.1875, -578596.4375, -578667.375, -578688.0625, -578776.375, -578810.6875, -578832.8125, -578867.5625, -578872.0, -578904.625, -578987.6875, -579003.125, -579106.8125, -579138.8125, -579159.4375, -579189.1875, -579193.375, -579216.6875, -579314.125, -579325.0, -579448.25, -579476.375, -579497.125, -579520.4375, -579524.9375, -579537.125, -579651.75, -579660.0625, -579806.0, -579830.125, -579851.875, -579866.8125, -579874.25, -579975.0625, -580055.25, -580194.3125, -580213.875, -580250.1875, -580251.0625, -580287.875, -580355.5, -580365.3125, -580450.5, -580481.3125, -580499.5, -580530.4375, -580531.5, -580559.0625, -580640.25, -580647.125, -580750.6875, -580778.375, -580796.5625, -580820.0625, -580824.3125, -580839.25, -580937.0625, -580946.125, -581072.4375, -581094.75, -581115.75, -581129.6875, -581139.0, -581226.1875, -581300.375, -581419.1875, -581438.375, -581469.6875, -581474.375, -581504.6875, -581566.125, -581584.125, -581661.8125, -581688.5625, -581707.875, -581732.875, -581739.75, -581759.5625, -581834.0, -581857.8125, -581953.25, -581976.5625, -581997.3125, -582015.0625, -582026.125, -582032.6875, -582122.875, -582166.6875, -582283.625, -582302.3125, -582326.375, -582334.875, -582351.6875, -582422.8125, -582456.4375, -582550.875, -582571.6875, -582595.0625, -582607.3125, -582626.75, -582685.3125, -582726.9375, -582802.4375, -582825.3125, -582846.75, -582864.6875, -582879.5, -582888.8125, -582889.375, -582986.375, -583028.375, -583158.375, -583171.4375, -583199.375, -583237.3125, -583269.4375, -583316.875, -583340.625, -583362.25, -583386.125, -583398.875, -583421.1875, -583465.875, -583506.9375, -583564.125, -583587.0625, -583607.1875, -583627.375, -583641.625, -583656.9375, -583659.0, -583662.625, -583760.25, -583797.8125, -583834.375, -583839.25, -583881.9375, -583912.875, -583926.5625, -583963.25, -583990.0625, -584010.375, -584041.25, -584044.3125, -584077.875, -584117.3125, -584130.3125, -584180.25, -584204.375, -584220.8125, -584244.9375, -584251.5, -584272.5625, -584323.5625, -584345.9375, -584412.8125, -584433.125, -584451.375, -584467.3125, -584480.0, -584486.625, -584488.5, -584575.25, -584594.6875, -584708.125, -584721.5625, -584741.1875, -584744.75, -584757.9375, -584825.5, -584838.0625, -584928.1875, -584944.5, -584963.75, -584970.9375, -584985.5625, -585041.8125, -585069.125, -585141.8125, -585159.6875, -585177.25, -585190.8125, -585201.6875, -585206.4375, -585274.1875, -585325.0, -585413.0, -585428.625, -585447.875, -585454.75, -585469.625, -585521.875, -585548.6875, -585618.5625, -585635.875, -585654.0625, -585664.9375, -585679.8125, -585724.5, -585764.8125, -585822.25, -585840.375, -585857.6875, -585872.625, -585884.9375, -585893.25, -585895.3125, -585967.25, -586006.0625, -586103.375, -586114.9375, -586137.0, -586165.8125, -586193.3125, -586228.8125, -586247.75, -586265.375, -586284.8125, -586295.5, -586314.375, -586314.4375, -586328.875, -586388.75, -586390.375, -586467.75, -586483.875, -586496.4375, -586508.5, -586513.1875, -586516.1875, -586589.5625, -586601.6875, -586697.25, -586709.125, -586725.125, -586729.125, -586739.375, -586797.8125, -586813.25, -586892.1875, -586905.0, -586922.6875, -586928.25, -586942.375, -586989.375, -587008.3125, -587069.1875, -587085.125, -587099.375, -587111.8125, -587119.9375, -587126.0625, -587183.5, -587215.1875, -587288.8125, -587302.375, -587318.5625, -587325.875, -587337.4375, -587382.875, -587414.6875, -587475.875, -587490.1875, -587507.3125, -587515.625, -587530.4375, -587567.4375, -587594.9375, -587642.1875, -587658.25, -587672.75, -587686.5, -587696.25, -587705.6875, -587750.1875, -587789.5, -587845.9375, -587860.5, -587875.625, -587886.1875, -587897.875, -587901.0, -587903.0, -587970.6875, -587980.875, -588072.8125, -588081.3125, -588099.6875, -588127.4375, -588150.5625, -588184.8125, -588201.375, -588215.1875, -588231.5625, -588240.25, -588254.875, -588287.25, -588315.6875, -588357.125, -588372.375, -588386.5, -588399.9375, -588409.6875, -588420.125, -588420.4375, -588422.8125, -588492.875, -588517.75, -588543.0, -588545.5625, -588575.6875, -588597.875, -588606.6875, -588633.0, -588652.1875, -588665.5, -588687.25, -588688.6875, -588711.25, -588740.5625, -588748.9375, -588786.1875, -588802.5, -588814.3125, -588829.9375, -588834.9375, -588848.25, -588886.25, -588905.0625, -588955.0, -588968.6875, -588982.8125, -588992.4375, -589003.4375, -589004.8125, -589009.375, -589069.8125, -589071.0, -589150.1875, -589160.75, -589172.6875, -589177.0, -589183.5625, -589234.0, -589252.625, -589321.5, -589331.5625, -589347.4375, -589350.3125, -589364.125, -589402.625, -589412.4375, -589461.4375, -589475.625, -589486.8125, -589498.4375, -589503.4375, -589510.625, -589557.75, -589573.75, -589634.125, -589645.8125, -589658.375, -589665.75, -589672.8125, -589673.3125, -589729.3125, -589761.4375, -589833.6875, -589843.25, -589857.25, -589860.125, -589870.75, -589913.125, -589924.0, -589980.9375, -589992.25, -590005.625, -590012.0, -590022.125, -590057.75, -590080.625, -590126.5, -590139.375, -590151.3125, -590161.6875, -590169.5, -590174.3125, -590217.375, -590251.875, -590306.625, -590318.125, -590331.1875, -590337.6875, -590347.8125, -590381.8125, -590410.5625, -590456.8125, -590468.75, -590482.4375, -590489.9375, -590502.25, -590530.1875, -590554.4375, -590589.5625, -590602.875, -590614.6875, -590626.4375, -590634.8125, -590643.5, -590643.625, -590646.1875, -590705.75, -590726.5, -590748.1875, -590750.8125, -590775.9375, -590794.9375, -590803.0, -590825.8125, -590841.8125, -590853.3125, -590871.3125, -590873.625, -590892.125, -590917.25, -590925.5, -590957.6875, -590971.75, -590982.375, -590995.25, -591000.375, -591010.875, -591043.5, -591062.8125, -591106.5, -591118.1875, -591130.6875, -591138.5, -591149.0, -591149.4375, -591154.8125, -591205.3125, -591225.375, -591244.375, -591249.3125, -591271.3125, -591289.3125, -591298.5, -591320.9375, -591335.125, -591346.9375, -591363.1875, -591367.4375, -591384.0625, -591407.125, -591419.9375, -591449.4375, -591462.3125, -591473.25, -591485.4375, -591492.25, -591501.4375, -591502.0, -591504.5625, -591559.625, -591580.75, -591601.9375, -591604.5, -591630.4375, -591647.0625, -591654.9375, -591672.875, -591689.1875, -591702.6875, -591723.1875, -591736.375, -591749.0, -591763.0, -591770.0625, -591783.75, -591784.25, -591795.1875, -591832.375, -591836.125, -591885.75, -591896.1875, -591906.5625, -591913.25, -591920.5, -591953.75, -591979.25, -592022.375, -592033.125, -592044.0, -592051.9375, -592059.8125, -592062.125, -592062.625, -592115.125, -592123.6875, -592195.9375, -592201.4375, -592215.9375, -592236.75, -592251.5625, -592277.3125, -592289.6875, -592299.75, -592312.375, -592318.0625, -592330.0, -592354.1875, -592371.8125, -592402.25, -592414.3125, -592424.0625, -592435.0, -592441.125, -592449.625, -592478.6875, -592500.4375, -592536.75, -592547.8125, -592558.0, -592567.3125, -592574.0, -592579.4375, -592612.9375, -592643.9375, -592687.125, -592697.125, -592708.0625, -592714.375, -592722.8125, -592723.125, -592724.4375, -592775.5625, -592796.0, -592815.4375, -592817.5625, -592842.1875, -592856.375, -592863.625, -592880.125, -592894.8125, -592907.5, -592926.625, -592938.3125, -592950.0, -592962.75, -592969.3125, -592981.75, -592982.4375, -592991.25, -593025.75, -593030.625, -593076.75, -593086.0, -593096.0625, -593101.375, -593108.5, -593138.4375, -593158.375, -593197.1875, -593206.9375, -593216.8125, -593224.6875, -593230.875, -593233.5, -593269.6875, -593299.5, -593345.6875, -593354.1875, -593365.0, -593369.8125, -593377.875, -593405.6875, -593425.625, -593463.5, -593472.9375, -593483.75, -593489.4375, -593499.4375, -593522.1875, -593541.375, -593570.4375, -593581.0625, -593590.3125, -593599.9375, -593606.0, -593612.6875, -593639.625, -593665.0625, -593699.125, -593709.3125, -593718.625, -593726.125, -593733.375, -593736.6875, -593737.6875, -593780.0625, -593794.9375, -593853.5625, -593858.9375, -593871.75, -593888.625, -593902.1875, -593922.75, -593933.75, -593943.0625, -593954.1875, -593959.4375, -593970.6875, -593990.4375, -594005.625, -594030.0625, -594040.4375, -594049.5625, -594059.3125, -594064.875, -594073.875, -594096.5, -594115.3125, -594143.9375, -594154.3125, -594163.25, -594171.9375, -594177.5625, -594183.5, -594210.1875, -594235.25, -594269.625, -594278.625, -594288.0625, -594295.0, -594301.625, -594304.25, -594305.125, -594346.9375, -594356.1875, -594413.25, -594418.375, -594430.5, -594446.75, -594459.875, -594480.25, -594490.625, -594499.5, -594510.3125, -594514.9375, -594525.0625, -594544.625, -594559.125, -594583.5625, -594593.625, -594602.25, -594611.5, -594616.375, -594624.5, -594646.8125, -594665.75, -594694.75, -594704.125, -594712.625, -594720.75, -594726.25, -594730.875, -594730.9375, -594768.75, -594795.875, -594848.0, -594853.25, -594865.6875, -594880.0, -594891.75, -594908.8125, -594918.6875, -594927.8125, -594938.8125, -594943.125, -594954.625, -594970.625, -594983.3125, -595003.4375, -595013.6875, -595021.625, -595031.6875, -595036.4375, -595045.625, -595064.625, -595079.125, -595103.25, -595112.3125, -595120.5625, -595129.3125, -595134.375, -595141.375, -595163.9375, -595182.125, -595210.5625, -595219.375, -595227.375, -595234.5625, -595240.375, -595244.6875, -595271.0, -595296.375, -595330.25, -595338.4375, -595347.125, -595352.3125, -595359.0, -595359.6875, -595361.0625, -595400.9375, -595401.125, -595456.0, -595460.25, -595471.0, -595487.1875, -595499.125, -595519.75, -595528.875, -595536.6875, -595546.0625, -595550.6875, -595559.1875, -595578.125, -595592.5625, -595616.625, -595625.625, -595633.1875, -595641.5, -595646.1875, -595652.125, -595674.75, -595693.5, -595722.1875, -595730.375, -595738.25, -595744.75, -595750.5, -595753.8125, -595754.125, -595789.8125, -595804.125, -595853.9375, -595858.4375, -595869.5625, -595883.25, -595894.125, -595911.125, -595920.125, -595928.3125, -595938.0, -595942.25, -595951.9375, -595968.125, -595980.0, -596000.125, -596009.0, -596016.6875, -596025.4375, -596029.625, -596037.25, -596056.125, -596070.75, -596094.4375, -596103.25, -596110.5625, -596117.75, -596122.5625, -596128.0625, -596150.25, -596169.4375, -596197.875, -596206.0, -596213.875, -596219.125, -596225.25, -596227.125, -596227.6875, -596262.8125, -596272.3125, -596321.0625, -596325.125, -596335.6875, -596349.5625, -596359.9375, -596376.8125, -596385.0, -596393.0, -596402.0625, -596406.3125, -596414.875, -596431.125, -596442.5625, -596463.0625, -596471.0625, -596478.625, -596486.625, -596490.9375, -596497.6875, -596516.4375, -596530.875, -596555.1875, -596562.6875, -596569.8125, -596576.4375, -596581.25, -596585.8125, -596608.1875, -596627.3125, -596655.875, -596663.1875, -596670.9375, -596676.0, -596681.375, -596683.0625, -596683.9375, -596718.4375, -596723.75, -596771.0625, -596775.4375, -596785.1875, -596798.75, -596808.625, -596825.8125, -596834.0625, -596841.0625, -596849.4375, -596853.4375, -596861.875, -596877.75, -596889.0, -596909.25, -596917.25, -596924.25, -596931.3125, -596935.5, -596941.4375, -596960.1875, -596974.9375, -596998.75, -597006.25, -597013.1875, -597019.625, -597023.8125, -597027.8125, -597050.1875, -597070.3125, -597098.25, -597104.75, -597112.1875, -597117.125, -597122.1875, -597123.625, -597123.9375, -597157.8125, -597160.1875, -597206.25, -597209.9375, -597219.3125, -597232.875, -597242.875, -597259.625, -597267.5625, -597274.25, -597282.125, -597285.9375, -597293.0625, -597309.25, -597320.5, -597340.5625, -597348.5625, -597354.8125, -597362.0, -597366.0, -597371.25, -597390.0, -597404.625, -597428.6875, -597435.875, -597442.6875, -597448.4375, -597452.75, -597455.9375, -597478.1875, -597499.25, -597527.6875, -597533.8125, -597541.25, -597545.5, -597550.6875, -597550.875, -597551.9375, -597585.375, -597598.25, -597610.625, -597612.125, -597628.125, -597637.5625, -597642.0625, -597652.6875, -597662.875, -597670.625, -597683.5625, -597691.4375, -597698.6875, -597706.8125, -597711.3125, -597719.25, -597719.6875, -597725.6875, -597748.375, -597751.4375, -597781.875, -597787.8125, -597794.1875, -597797.6875, -597802.375, -597822.3125, -597835.0625, -597860.3125, -597866.9375, -597873.3125, -597878.375, -597882.8125, -597884.8125, -597908.3125, -597927.75, -597958.0, -597963.75, -597970.8125, -597974.25, -597979.3125, -597997.6875, -598010.8125, -598035.8125, -598041.8125, -598049.125, -598053.0625, -598059.5625, -598074.5625, -598087.4375, -598106.4375, -598113.8125, -598119.8125, -598126.4375, -598130.375, -598135.0, -598135.375, -598135.8125, -598168.375, -598179.625, -598191.75, -598192.4375, -598206.25, -598216.625, -598220.4375, -598233.0625, -598241.6875, -598248.125, -598257.8125, -598258.25, -598268.75, -598282.125, -598286.25, -598303.5, -598311.0625, -598316.25, -598323.8125, -598326.3125, -598332.125, -598349.625, -598359.125, -598382.8125, -598389.0, -598395.6875, -598400.0, -598405.4375, -598406.0625, -598408.6875, -598436.375, -598447.1875, -598457.6875, -598460.0625, -598472.1875, -598481.8125, -598487.25, -598498.5, -598506.375, -598512.875, -598522.1875, -598524.375, -598533.1875, -598545.5625, -598552.375, -598568.375, -598575.375, -598581.375, -598588.375, -598591.8125, -598597.0625, -598613.125, -598628.25, -598649.625, -598656.25, -598663.1875, -598667.3125, -598673.625, -598674.0625, -598678.5625, -598702.25, -598702.6875, -598734.0625, -598739.4375, -598744.5625, -598748.125, -598750.5, -598772.1875, -598792.125, -598821.5625, -598827.0625, -598835.125, -598836.625, -598844.375, -598859.625, -598863.9375, -598883.375, -598890.125, -598895.1875, -598902.0625, -598903.4375, -598909.25, -598927.5, -598934.25, -598957.5, -598963.625, -598969.0625, -598974.0625, -598976.9375, -598979.5625, -599001.625, -599012.5, -599041.3125, -599046.9375, -599052.4375, -599055.4375, -599059.875, -599077.9375, -599090.875, -599114.875, -599120.875, -599127.5625, -599131.0, -599137.1875, -599151.1875, -599161.6875, -599180.1875, -599186.4375, -599192.25, -599197.8125, -599201.5625, -599205.625, -599222.5, -599237.5, -599259.5, -599265.5625, -599271.8125, -599275.8125, -599280.0625, -599282.0625, -599282.625, -599309.6875, -599315.375, -599351.75, -599355.25, -599362.625, -599373.4375, -599382.4375, -599395.875, -599402.8125, -599408.25, -599414.875, -599418.8125, -599425.0, -599437.9375, -599448.625, -599464.625, -599471.3125, -599476.8125, -599482.8125, -599486.625, -599491.125, -599491.8125, -599492.625, -599520.25, -599530.75, -599541.1875, -599542.5, -599554.5625, -599563.5, -599567.375, -599578.25, -599585.6875, -599591.4375, -599600.125, -599601.25, -599610.4375, -599621.75, -599626.1875, -599640.8125, -599647.625, -599652.4375, -599659.3125, -599661.3125, -599667.1875, -599682.125, -599690.25, -599710.0625, -599716.0625, -599721.5, -599726.0625, -599730.1875, -599731.5625, -599733.625, -599757.875, -599761.0, -599792.75, -599797.0, -599802.4375, -599804.1875, -599807.6875, -599827.375, -599834.4375, -599861.125, -599866.0, -599872.0625, -599874.0, -599879.25, -599894.5, -599901.625, -599922.125, -599927.8125, -599933.0, -599937.5625, -599940.1875, -599943.0625, -599961.8125, -599974.0, -599998.8125, -600003.6875, -600009.5625, -600012.5, -600016.6875, -600032.0625, -600043.5625, -600064.625, -600069.8125, -600075.5625, -600078.875, -600084.25, -600096.8125, -600108.375, -600124.6875, -600130.375, -600135.4375, -600140.75, -600144.25, -600147.875, -600148.4375, -600169.1875, -600188.875, -600217.9375, -600221.3125, -600229.0625, -600237.1875, -600245.375, -600255.1875, -600261.375, -600267.1875, -600273.5625, -600277.125, -600283.8125, -600293.0, -600302.6875, -600315.0, -600320.5, -600326.5, -600332.0625, -600335.4375, -600340.75, -600341.5625, -600344.9375, -600364.9375, -600366.9375, -600392.9375, -600398.0, -600402.5, -600405.1875, -600407.6875, -600425.375, -600438.8125, -600462.8125, -600467.375, -600473.6875, -600475.1875, -600481.5625, -600494.0, -600499.8125, -600516.6875, -600522.375, -600526.3125, -600531.6875, -600533.9375, -600537.75, -600553.875, -600562.625, -600583.5, -600588.375, -600593.3125, -600597.125, -600600.1875, -600601.0, -600620.1875, -600638.6875, -600664.0625, -600668.125, -600674.125, -600675.6875, -600680.4375, -600695.0, -600700.625, -600719.6875, -600724.6875, -600729.6875, -600732.8125, -600737.0, -600737.375, -600738.5625, -600762.0, -600762.6875, -600793.875, -600797.125, -600802.25, -600803.125, -600806.3125, -600825.0, -600827.75, -600853.25, -600857.0625, -600862.375, -600864.125, -600867.75, -600883.75, -600889.5, -600909.4375, -600914.375, -600919.0, -600922.4375, -600925.3125, -600926.25, -600945.4375, -600958.25, -600982.625, -600986.8125, -600992.3125, -600994.1875, -600998.625, -601012.9375, -601018.6875, -601037.875, -601042.5, -601047.5, -601050.6875, -601054.5625, -601054.6875, -601055.5625, -601078.875, -601087.625, -601095.9375, -601097.1875, -601107.3125, -601114.9375, -601118.0, -601127.0625, -601133.125, -601138.0625, -601145.8125, -601146.3125, -601154.625, -601164.0625, -601167.375, -601179.1875, -601184.9375, -601188.9375, -601195.0625, -601196.5, -601201.6875, -601213.9375, -601219.375, -601235.5625, -601240.625, -601245.0625, -601248.9375, -601252.5625, -601254.0, -601254.625, -601275.75, -601281.75, -601309.5, -601312.75, -601317.9375, -601318.4375, -601322.5, -601338.3125, -601341.5, -601362.5625, -601367.1875, -601371.25, -601374.0, -601377.3125, -601390.9375, -601400.75, -601418.625, -601423.125, -601427.9375, -601431.4375, -601435.0, -601435.5625, -601436.0625, -601457.9375, -601459.6875, -601489.8125, -601492.5, -601498.125, -601507.125, -601514.4375, -601525.625, -601530.875, -601535.1875, -601540.5, -601543.5625, -601547.25, -601547.375, -601549.5, -601569.4375, -601576.6875, -601584.5, -601586.375, -601595.0625, -601601.625, -601605.6875, -601613.75, -601619.625, -601624.3125, -601630.875, -601632.3125, -601639.0625, -601648.125, -601652.8125, -601663.875, -601669.125, -601673.25, -601678.75, -601680.75, -601685.1875, -601696.0, -601704.8125, -601719.1875, -601724.1875, -601729.0625, -601732.1875, -601736.25, -601737.6875, -601739.25, -601756.9375, -601761.5, -601784.875, -601788.5, -601793.0, -601794.875, -601797.5, -601811.625, -601818.9375, -601837.9375, -601842.125, -601847.1875, -601849.3125, -601853.3125, -601864.9375, -601872.9375, -601888.0, -601892.3125, -601896.6875, -601900.5625, -601903.25, -601905.125, -601905.625, -601925.0, -601933.1875, -601959.875, -601962.5, -601968.125, -601975.6875, -601982.3125, -601991.75, -601996.8125, -602001.1875, -602006.125, -602009.0, -602013.5625, -602022.5625, -602031.3125, -602043.0625, -602047.6875, -602051.6875, -602056.375, -602059.25, -602062.5625, -602063.25, -602064.25, -602084.1875, -602085.25, -602111.25, -602114.5, -602118.8125, -602119.625, -602122.25, -602137.875, -602140.6875, -602161.5625, -602165.0625, -602169.5, -602171.0, -602174.5, -602187.75, -602193.625, -602210.5625, -602214.5625, -602218.875, -602221.6875, -602224.3125, -602225.3125, -602241.5]}],\n",
       "                        {\"title\": {\"text\": \"Energy Minimization\"}, \"xaxis\": {\"title\": {\"text\": \"Energy Minimization Step\"}}, \"yaxis\": {\"title\": {\"text\": \"Potential Energy KJ/mol-1\"}}},\n",
       "                        {\"showLink\": false, \"linkText\": \"Export to plot.ly\", \"plotlyServerURL\": \"https://plot.ly\", \"responsive\": true}\n",
       "                    ).then(function(){\n",
       "                            \n",
       "var gd = document.getElementById('ad609f6e-83e5-4eb3-bccf-8446009b8f33');\n",
       "var x = new MutationObserver(function (mutations, observer) {{\n",
       "        var display = window.getComputedStyle(gd).display;\n",
       "        if (!display || display === 'none') {{\n",
       "            console.log([gd, 'removed!']);\n",
       "            Plotly.purge(gd);\n",
       "            observer.disconnect();\n",
       "        }}\n",
       "}});\n",
       "\n",
       "// Listen for the removal of the full notebook cells\n",
       "var notebookContainer = gd.closest('#notebook-container');\n",
       "if (notebookContainer) {{\n",
       "    x.observe(notebookContainer, {childList: true});\n",
       "}}\n",
       "\n",
       "// Listen for the clearing of the current output cell\n",
       "var outputEl = gd.closest('.output');\n",
       "if (outputEl) {{\n",
       "    x.observe(outputEl, {childList: true});\n",
       "}}\n",
       "\n",
       "                        })\n",
       "                };\n",
       "                });\n",
       "            </script>\n",
       "        </div>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import plotly\n",
    "import plotly.graph_objs as go\n",
    "\n",
    "#Read data from file and filter energy values higher than 1000 Kj/mol^-1\n",
    "with open(output_min_ene_xvg,'r') as energy_file:\n",
    "    x,y = map(\n",
    "        list,\n",
    "        zip(*[\n",
    "            (float(line.split()[0]),float(line.split()[1]))\n",
    "            for line in energy_file \n",
    "            if not line.startswith((\"#\",\"@\")) \n",
    "            if float(line.split()[1]) < 1000 \n",
    "        ])\n",
    "    )\n",
    "\n",
    "plotly.offline.init_notebook_mode(connected=True)\n",
    "\n",
    "fig = {\n",
    "    \"data\": [go.Scatter(x=x, y=y)],\n",
    "    \"layout\": go.Layout(title=\"Energy Minimization\",\n",
    "                        xaxis=dict(title = \"Energy Minimization Step\"),\n",
    "                        yaxis=dict(title = \"Potential Energy KJ/mol-1\")\n",
    "                       )\n",
    "}\n",
    "\n",
    "plotly.offline.iplot(fig)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
