{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# The BioBB REST API\n",
    "\n",
    "The [BioBB REST API](https://mmb.irbbarcelona.org/biobb-api) allows the execution of the [BioExcel Building Blocks](https://mmb.irbbarcelona.org/biobb/) in a remote server.\n",
    "\n",
    "## Documentation\n",
    "\n",
    "For an extense documentation section, please go to the [BioBB REST API website help](https://mmb.irbbarcelona.org/biobb-api/rest).\n",
    "\n",
    "## Settings\n",
    "\n",
    "### Auxiliar libraries used\n",
    "\n",
    "* [requests](https://pypi.org/project/requests/): Requests allows you to send *organic, grass-fed* HTTP/1.1 requests, without the need for manual labor.\n",
    "* [nb_conda_kernels](https://github.com/Anaconda-Platform/nb_conda_kernels): Enables a Jupyter Notebook or JupyterLab application in one conda environment to access kernels for Python, R, and other languages found in other environments.\n",
    "* [nglview](http://nglviewer.org/#nglview): Jupyter/IPython widget to interactively view molecular structures and trajectories in notebooks.\n",
    "* [ipywidgets](https://github.com/jupyter-widgets/ipywidgets): Interactive HTML widgets for Jupyter notebooks and the IPython kernel.\n",
    "* [plotly](https://plot.ly/python/offline/): Python interactive graphing library integrated in Jupyter notebooks.\n",
    "\n",
    "### Conda Installation and Launch\n",
    "\n",
    "```console\n",
    "git clone https://github.com/bioexcel/biobb_REST_API_documentation.git\n",
    "cd biobb_REST_API_documentation\n",
    "conda env create -f conda_env/environment.yml\n",
    "conda activate biobb_REST_API_documentation\n",
    "jupyter-nbextension enable --py --user widgetsnbextension\n",
    "jupyter-nbextension enable --py --user nglview\n",
    "jupyter-notebook biobb_REST_API_documentation/notebooks/biobb_REST_API_documentation.ipynb\n",
    "```\n",
    "\n",
    "***\n",
    "\n",
    "## Index\n",
    "\n",
    " * [Behaviour](#behaviour)\n",
    "     * [Tools information](#tools_info)\n",
    "         * [List of packages](#list_pckg)\n",
    "         * [List of tools](#list_tools)\n",
    "         * [Tool's properties](#tools_prop)\n",
    "     * [Launch tool](#launch_tool)\n",
    "     * [Retrieve status](#retrieve_status)\n",
    "     * [Retrieve data](#retrieve_data)         \n",
    " * [Examples](#examples)\n",
    "     * [Tools information](#tools_info_ex)\n",
    "         * [List of packages](#list_pckg_ex)\n",
    "         * [List of tools from a specific package](#list_tools_ex)\n",
    "         * [Tool's properties](#tools_prop_ex)\n",
    "     * [Launch tool](#launch_tool_ex)\n",
    "         * [Launch job with a YAML file config](#tool_yml_ex)\n",
    "         * [Launch job with a JSON file config](#tool_json_ex)\n",
    "         * [Launch job with a piython dictionary config](#tool_dict_ex)\n",
    "     * [Retrieve status](#retrieve_status_ex)\n",
    "     * [Retrieve data](#retrieve_data_ex)\n",
    " * [Practical cases](#practical_cases)\n",
    "     * [Example 1: download PDB file from RSCB database](#example1)\n",
    "     * [Example 2: extract heteroatom from a given structure](#example2)\n",
    "     * [Example 3: extract energy components from a given GROMACS energy file](#example3)\n",
    "\n",
    "***\n",
    "<img src=\"https://bioexcel.eu/wp-content/uploads/2019/04/Bioexcell_logo_1080px_transp.png\" alt=\"Bioexcel2 logo\"\n",
    "\ttitle=\"Bioexcel2 logo\" width=\"400\" />\n",
    "***\n",
    "\n",
    "<a id=\"behaviour\"></a>\n",
    "## Behaviour\n",
    "\n",
    "The BioBB REST API works as an asynchronous launcher of jobs, as these jobs can last from a few seconds to several minutes, there are some steps that must be performed for having the complete results of every tool.\n",
    "\n",
    "BioExcel Building Blocks are structured in [packages and tools](http://mmb.irbbarcelona.org/biobb/availability/source). Every call to the BioBB REST API executes one single tool and returns the output file(s) related to this specific tool.\n",
    "\n",
    "<a id=\"tools_info\"></a>\n",
    "### Tools information\n",
    "\n",
    "<a id=\"list_pckg\"></a>\n",
    "#### List of packages\n",
    "\n",
    "In order to get a complete list of available packages, we must do a GET request to the following endpoint:\n",
    "\n",
    "`http://mmb.irbbarcelona.org/biobb-api/rest/v1/launch`\n",
    "\n",
    "This endpoint returns a JSON HTTP response with status `200`. More information in the [BioBB REST API help section](http://mmb.irbbarcelona.org/biobb-api/rest/#/List%20of%20Services/getPckgList).\n",
    "\n",
    "<a id=\"list_tools\"></a>\n",
    "#### List of tools\n",
    "\n",
    "If there is need for a list of tools for a single package, we must do a GET request to the following endpoint:\n",
    "\n",
    "`http://mmb.irbbarcelona.org/biobb-api/rest/v1/launch/{package}`\n",
    "\n",
    "This endpoint returns a JSON HTTP response with status `200` or a `404` status if the package id is incorrect. More information in the [BioBB REST API help section](http://mmb.irbbarcelona.org/biobb-api/rest/#/List%20of%20Services/getToolsList).\n",
    "\n",
    "<a id=\"tools_prop\"></a>\n",
    "#### Tool's properties\n",
    "\n",
    "If there is only need for the information of a single tool, we must do a GET request to the following endpoint:\n",
    "\n",
    "`http://mmb.irbbarcelona.org/biobb-api/rest/v1/launch/{package}/{tool}`\n",
    "\n",
    "This endpoint returns a JSON HTTP response with status `200` or a `404` status if the package id and / or the tool id are incorrect. The reason for failure should be detailed in the JSON response. More information in the [BioBB REST API help section](http://mmb.irbbarcelona.org/biobb-api/rest/#/Launch%20Tool/getLaunchTool).\n",
    "\n",
    "<a id=\"launch_tool\"></a>\n",
    "### Launch tool\n",
    "\n",
    "For launching a tool, we must do a POST request to the following endpoint:\n",
    "\n",
    "`http://mmb.irbbarcelona.org/biobb-api/rest/v1/launch/{package}/{tool}`\n",
    "\n",
    "In the body of this POST request, we must add the file(s) needed as input (included the properties config file in JSON or YAML format) and the name for the output(s). The detailed list of inputs and outputs with its respectives properties can be found in the GET request of this same endpoint.\n",
    "\n",
    "This endpoint returns a JSON HTTP response with the following possible status:\n",
    "\n",
    "* `303`: The job has been successfully launched and the user must save the token provided and follow to the next endpoint (defined in the same JSON response)\n",
    "* `404`: There was some error launching the tool. The reason for failure should be detailed in the JSON response.\n",
    "* `500`: The job has been launched, but some internal server error has occurred during the execution.\n",
    "\n",
    "More information in the [BioBB REST API help section](http://mmb.irbbarcelona.org/biobb-api/rest/#/Launch%20Tool/postLaunchTool).\n",
    "\n",
    "<a id=\"retrieve_status\"></a>\n",
    "### Retrieve status\n",
    "\n",
    "If the previous endpoint returned a `303` status, we must do a GET request to the following endpoint providing the given token in the path:\n",
    "\n",
    "`http://mmb.irbbarcelona.org/biobb-api/rest/v1/retrieve/status/{token}`\n",
    "\n",
    "This endpoint checks the state of the job and returns a JSON HTTP response with the following possible status:\n",
    "\n",
    "* `200`: The job has finished successfully and in the JSON response we can found a list of output files generated by the job with its correspondent id for retrieving them on the next endpoint (defined in the same JSON message).\n",
    "* `202`: The job is still running.\n",
    "* `404`: Token incorrect, job unexisting or expired.\n",
    "* `500`: Some internal server error has occurred during the execution.\n",
    "\n",
    "More information in the [BioBB REST API help section](http://mmb.irbbarcelona.org/biobb-api/rest/#/Retrieve/getRetrieveStatus).\n",
    "\n",
    "<a id=\"retrieve_data\"></a>\n",
    "### Retrieve data\n",
    "\n",
    "Once the previous endpoint returns a `200` status, the output file(s) are ready for its retrieval, so we must do a GET request to the following endpoint providing the given file id in the path:\n",
    "\n",
    "`http://mmb.irbbarcelona.org/biobb-api/rest/v1/retrieve/data/{id}`\n",
    "\n",
    "This endpoint returns the requested file with a `200` status or a `404` status if the provided id is incorrect, the file doesn't exist or it has expired. More information in the [BioBB REST API help section](http://mmb.irbbarcelona.org/biobb-api/rest/#/Retrieve/getRetrieveData).\n",
    "\n",
    "Note that if we have executed a job that returns multiple output files, a call to this endpoint must be done for each of the output files generated by the job."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"examples\"></a>\n",
    "## Examples\n",
    "\n",
    "Below we will do calls to all the previously defined endpoints and define some functions for make easier the connection to the BioBB REST API through Jupyter Notebook.\n",
    "\n",
    "First off, we will import the Python requests and json library and set the root URI for the BioBB REST API."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import json\n",
    "\n",
    "apiURL  = \"http://mmb.irbbarcelona.org/biobb-api/rest/v1/\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"tools_info_ex\"></a>\n",
    "### Tools information"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Definition of simple GET / POST request functions and a class Response:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Class for returning response status and json content of a requested URL\n",
    "class Response:\n",
    "  def __init__(self, status, json):\n",
    "    self.status = status\n",
    "    self.json = json\n",
    "\n",
    "# Perform GET request\n",
    "def get_data(url):\n",
    "    r = requests.get(url)\n",
    "    return Response(r.status_code, json.loads(r.text))\n",
    "\n",
    "# Perform POST request\n",
    "def post_data(url, d, f):\n",
    "    r = requests.post(url, data = d, files = f)\n",
    "    return Response(r.status_code, json.loads(r.text))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"list_pckg_ex\"></a>\n",
    "#### List of packages"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"packages\": [\n",
      "    {\n",
      "      \"id\": \"biobb_analysis\",\n",
      "      \"tools\": [\n",
      "        {\n",
      "          \"id\": \"gmx_cluster\",\n",
      "          \"description\": \"Creates cluster structures from a given GROMACS compatible trajectory\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the gmx_cluster tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_cluster.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.tpr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.brk$\",\n",
      "                \".*\\\\.ent$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS trajectory file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.cpt$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_index_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the GROMACS index file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output cluster file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_cluster.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.cpt$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"gmx_rms\",\n",
      "          \"description\": \"Performs a Root Mean Square deviation (RMSd) analysis from a given GROMACS compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the gmx_rms tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_rms.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.tpr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.brk$\",\n",
      "                \".*\\\\.ent$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS trajectory file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.cpt$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_index_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the GROMACS index file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_xvg_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the XVG output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_rms.xvg\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xvg$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"gmx_rgyr\",\n",
      "          \"description\": \"Computes the radius of gyration (Rgyr) of a molecule about the x-, y- and z-axes, as a function of time, from a given GROMACS compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the gmx_rgyr tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_rgyr.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.tpr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.brk$\",\n",
      "                \".*\\\\.ent$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS trajectory file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.cpt$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_index_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the GROMACS index file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_xvg_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the XVG output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_rgyr.xvg\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xvg$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"gmx_energy\",\n",
      "          \"description\": \"Extracts energy components from a given GROMACS energy file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the gmx_energy tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_energy.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_energy_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input EDR file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/energy.edr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.edr$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_xvg_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the XVG output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_energy.xvg\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xvg$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"gmx_image\",\n",
      "          \"description\": \"Corrects periodicity (image) from a given GROMACS compatible trajectory file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the gmx_image tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_image.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS trajectory file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.cpt$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS input topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.tpr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.brk$\",\n",
      "                \".*\\\\.ent$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_index_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the GROMACS index file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_image.xtc\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"gmx_trjconv_str\",\n",
      "          \"description\": \"Converts between GROMACS compatible structure file formats and/or extracts a selection of atoms.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the gmx_trjconv_str tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_trjconv_str.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.cpt$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS input topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.tpr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.brk$\",\n",
      "                \".*\\\\.ent$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_index_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the GROMACS index file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_str_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_trjconv.str.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"gmx_trjconv_str_ens\",\n",
      "          \"description\": \"Extracts an ensemble of frames containing a selection of atoms from GROMACS compatible trajectory files.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the gmx_trjconv_str_ens tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_trjconv_str_ens.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS trajectory file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.cpt$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS input topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.tpr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.brk$\",\n",
      "                \".*\\\\.ent$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_index_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the GROMACS index file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_str_ens_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_trjconv.str.ens.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"gmx_trjconv_trj\",\n",
      "          \"description\": \"Converts between GROMACS compatible trajectory file formats and/or extracts a selection of atoms.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the gmx_trjconv_trj tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_trjconv_trj.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS trajectory file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.cpt$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_index_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the GROMACS index file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_trjconv.trj.xtc\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.g96$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tng$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_average\",\n",
      "          \"description\": \"Calculates a structure average of a given cpptraj compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_average tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_average.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed structure\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.average.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.rst7$\",\n",
      "                \".*\\\\.ncrst$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.sqm$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_bfactor\",\n",
      "          \"description\": \"Calculates the Bfactor fluctuations of a given cpptraj compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_bfactor tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_exp_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the experimental reference file (required if reference = experimental)\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/experimental.1e5t.pdb\",\n",
      "              \"formats\": null\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed analysis\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.bfactor.first.dat\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.dat$\",\n",
      "                \".*\\\\.agr$\",\n",
      "                \".*\\\\.xmgr$\",\n",
      "                \".*\\\\.gnu$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_rms\",\n",
      "          \"description\": \"Calculates the Root Mean Square deviation (RMSd) of a given cpptraj compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_rms tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_exp_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the experimental reference file (required if reference = experimental)\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/experimental.1e5t.pdb\",\n",
      "              \"formats\": null\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed analysis\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.rms.first.dat\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.dat$\",\n",
      "                \".*\\\\.agr$\",\n",
      "                \".*\\\\.xmgr$\",\n",
      "                \".*\\\\.gnu$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_rmsf\",\n",
      "          \"description\": \"Calculates the Root Mean Square fluctuations (RMSf) of a given cpptraj compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_rmsf tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_exp_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the experimental reference file (required if reference = experimental)\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/experimental.1e5t.pdb\",\n",
      "              \"formats\": null\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed analysis\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.rmsf.first.dat\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.dat$\",\n",
      "                \".*\\\\.agr$\",\n",
      "                \".*\\\\.xmgr$\",\n",
      "                \".*\\\\.gnu$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_rgyr\",\n",
      "          \"description\": \"Computes the radius of gyration (Rgyr) from a given cpptraj compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_rgyr tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_rgyr.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output analysis\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.rgyr.dat\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.dat$\",\n",
      "                \".*\\\\.agr$\",\n",
      "                \".*\\\\.xmgr$\",\n",
      "                \".*\\\\.gnu$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_dry\",\n",
      "          \"description\": \"Dehydrates a given cpptraj compatible trajectory stripping out solvent molecules and ions.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_dry tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_dry.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed trajectory\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.dry.netcdf\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.rst7$\",\n",
      "                \".*\\\\.ncrst$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.sqm$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_strip\",\n",
      "          \"description\": \"Strips a defined set of atoms (mask) from a given cpptraj compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_strip tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_strip.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed trajectory\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.strip.netcdf\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.rst7$\",\n",
      "                \".*\\\\.ncrst$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.sqm$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_snapshot\",\n",
      "          \"description\": \"Extracts a particular snapshot from a given cpptraj compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_snapshot tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_snapshot.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed structure\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.snapshot.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.rst7$\",\n",
      "                \".*\\\\.ncrst$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.sqm$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_slice\",\n",
      "          \"description\": \"Extracts a particular trajectory slice from a given cpptraj compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_slice tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_slice.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed trajectory\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.slice.netcdf\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.rst7$\",\n",
      "                \".*\\\\.ncrst$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.sqm$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_convert\",\n",
      "          \"description\": \"Converts between cpptraj compatible trajectory file formats and/or extracts a selection of atoms or frames.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_convert tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_convert.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed trajectory\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.convert.netcdf\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.rst7$\",\n",
      "                \".*\\\\.ncrst$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.sqm$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_mask\",\n",
      "          \"description\": \"Extracts a selection of atoms from a given cpptraj compatible trajectory.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_mask tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_mask.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed trajectory\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.mask.netcdf\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.rst7$\",\n",
      "                \".*\\\\.ncrst$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.sqm$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"cpptraj_image\",\n",
      "          \"description\": \"Corrects periodicity (image) from a given cpptraj trajectory file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cpptraj_image tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_image.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure or topology file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.prmtop$\",\n",
      "                \".*\\\\.parmtop$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_traj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input trajectory to be processed\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.cdf$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.restart$\",\n",
      "                \".*\\\\.ncrestart$\",\n",
      "                \".*\\\\.restartnc$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.charmm$\",\n",
      "                \".*\\\\.cor$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.sqm$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.conflib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpptraj_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output processed trajectory\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.image.netcdf\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.crd$\",\n",
      "                \".*\\\\.netcdf$\",\n",
      "                \".*\\\\.rst7$\",\n",
      "                \".*\\\\.ncrst$\",\n",
      "                \".*\\\\.dcd$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.binpos$\",\n",
      "                \".*\\\\.trr$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.sqm$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"biobb_chemistry\",\n",
      "      \"tools\": [\n",
      "        {\n",
      "          \"id\": \"acpype_params_ac\",\n",
      "          \"description\": \"Small molecule parameterization for AMBER MD package.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the acpype_params_ac tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_chemistry/master/biobb_chemistry/test/data/config/config_acpype_params_ac.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/acpype/acpype.params.mol2\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.mol2$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_frcmod\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the FRCMOD output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.ac.frcmod\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.frcmod$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_inpcrd\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the INPCRD output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.ac.inpcrd\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.inpcrd$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_lib\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the LIB output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.ac.lib\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.lib$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_prmtop\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the PRMTOP output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.ac.prmtop\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.prmtop$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"acpype_params_cns\",\n",
      "          \"description\": \"Small molecule parameterization for CNS/XPLOR MD package.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the acpype_params_cns tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_chemistry/master/biobb_chemistry/test/data/config/config_acpype_params_cns.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/acpype/acpype.params.mol2\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.mol2$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_par\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the PAR output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.cns.par\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.par$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_inp\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the INP output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.cns.inp\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.inp$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_top\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the TOP output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.cns.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"acpype_params_gmx\",\n",
      "          \"description\": \"Small molecule parameterization for GROMACS MD package.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the acpype_params_gmx tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_chemistry/master/biobb_chemistry/test/data/config/config_acpype_params_gmx.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/acpype/acpype.params.mol2\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.mol2$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_gro\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GRO output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.gmx.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_itp\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the ITP output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.gmx.itp\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.itp$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_top\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the TOP output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.gmx.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"acpype_params_gmx_opls\",\n",
      "          \"description\": \"Small molecule parameterization for OPLS/AA MD package.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the acpype_params_gmx_opls tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_chemistry/master/biobb_chemistry/test/data/config/config_acpype_params_gmx_opls.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/acpype/acpype.params.mol2\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.mol2$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_itp\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the ITP output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.gmx.opls.itp\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.itp$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path_top\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the TOP output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/acpype/ref_acpype.gmx.opls.top\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.top$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"babel_convert\",\n",
      "          \"description\": \"Small molecule format conversion.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the babel_convert tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_chemistry/master/biobb_chemistry/test/data/config/config_babel_convert.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/babel/babel.smi\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.abinit$\",\n",
      "                \".*\\\\.acesout$\",\n",
      "                \".*\\\\.acr$\",\n",
      "                \".*\\\\.adfout$\",\n",
      "                \".*\\\\.alc$\",\n",
      "                \".*\\\\.aoforce$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.axsf$\",\n",
      "                \".*\\\\.bgf$\",\n",
      "                \".*\\\\.box$\",\n",
      "                \".*\\\\.bs$\",\n",
      "                \".*\\\\.c09out$\",\n",
      "                \".*\\\\.c3d2$\",\n",
      "                \".*\\\\.caccrt$\",\n",
      "                \".*\\\\.can$\",\n",
      "                \".*\\\\.car$\",\n",
      "                \".*\\\\.castep$\",\n",
      "                \".*\\\\.ccc$\",\n",
      "                \".*\\\\.cdjson$\",\n",
      "                \".*\\\\.cdx$\",\n",
      "                \".*\\\\.cdxml$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.ck$\",\n",
      "                \".*\\\\.cml$\",\n",
      "                \".*\\\\.cmlr$\",\n",
      "                \".*\\\\.CONFIG$\",\n",
      "                \".*\\\\.CONTCAR$\",\n",
      "                \".*\\\\.CONTFF$\",\n",
      "                \".*\\\\.crk2d$\",\n",
      "                \".*\\\\.crk3d$\",\n",
      "                \".*\\\\.ct$\",\n",
      "                \".*\\\\.cub$\",\n",
      "                \".*\\\\.cube$\",\n",
      "                \".*\\\\.dallog$\",\n",
      "                \".*\\\\.dalmol$\",\n",
      "                \".*\\\\.dat$\",\n",
      "                \".*\\\\.dmol$\",\n",
      "                \".*\\\\.dx$\",\n",
      "                \".*\\\\.ent$\",\n",
      "                \".*\\\\.exyz$\",\n",
      "                \".*\\\\.fa$\",\n",
      "                \".*\\\\.fasta$\",\n",
      "                \".*\\\\.fch$\",\n",
      "                \".*\\\\.fchk$\",\n",
      "                \".*\\\\.fck$\",\n",
      "                \".*\\\\.feat$\",\n",
      "                \".*\\\\.fhiaims$\",\n",
      "                \".*\\\\.fract$\",\n",
      "                \".*\\\\.fs$\",\n",
      "                \".*\\\\.fsa$\",\n",
      "                \".*\\\\.g03$\",\n",
      "                \".*\\\\.g09$\",\n",
      "                \".*\\\\.g92$\",\n",
      "                \".*\\\\.g94$\",\n",
      "                \".*\\\\.g98$\",\n",
      "                \".*\\\\.gal$\",\n",
      "                \".*\\\\.gam$\",\n",
      "                \".*\\\\.gamess$\",\n",
      "                \".*\\\\.gamin$\",\n",
      "                \".*\\\\.gamout$\",\n",
      "                \".*\\\\.got$\",\n",
      "                \".*\\\\.gpr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.gukin$\",\n",
      "                \".*\\\\.gukout$\",\n",
      "                \".*\\\\.gzmat$\",\n",
      "                \".*\\\\.hin$\",\n",
      "                \".*\\\\.HISTORY$\",\n",
      "                \".*\\\\.inchi$\",\n",
      "                \".*\\\\.inp$\",\n",
      "                \".*\\\\.ins$\",\n",
      "                \".*\\\\.jin$\",\n",
      "                \".*\\\\.jout$\",\n",
      "                \".*\\\\.log$\",\n",
      "                \".*\\\\.lpmd$\",\n",
      "                \".*\\\\.mcdl$\",\n",
      "                \".*\\\\.mcif$\",\n",
      "                \".*\\\\.MDFF$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.ml2$\",\n",
      "                \".*\\\\.mmcif$\",\n",
      "                \".*\\\\.mmd$\",\n",
      "                \".*\\\\.mmod$\",\n",
      "                \".*\\\\.mol$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.mold$\",\n",
      "                \".*\\\\.molden$\",\n",
      "                \".*\\\\.molf$\",\n",
      "                \".*\\\\.moo$\",\n",
      "                \".*\\\\.mop$\",\n",
      "                \".*\\\\.mopcrt$\",\n",
      "                \".*\\\\.mopin$\",\n",
      "                \".*\\\\.mopout$\",\n",
      "                \".*\\\\.mpc$\",\n",
      "                \".*\\\\.mpo$\",\n",
      "                \".*\\\\.mpqc$\",\n",
      "                \".*\\\\.mrv$\",\n",
      "                \".*\\\\.msi$\",\n",
      "                \".*\\\\.nwo$\",\n",
      "                \".*\\\\.orca$\",\n",
      "                \".*\\\\.out$\",\n",
      "                \".*\\\\.outmol$\",\n",
      "                \".*\\\\.output$\",\n",
      "                \".*\\\\.pc$\",\n",
      "                \".*\\\\.pcjson$\",\n",
      "                \".*\\\\.pcm$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.pdbqt$\",\n",
      "                \".*\\\\.png$\",\n",
      "                \".*\\\\.pos$\",\n",
      "                \".*\\\\.POSCAR$\",\n",
      "                \".*\\\\.POSFF$\",\n",
      "                \".*\\\\.pqr$\",\n",
      "                \".*\\\\.pqs$\",\n",
      "                \".*\\\\.prep$\",\n",
      "                \".*\\\\.pwscf$\",\n",
      "                \".*\\\\.qcout$\",\n",
      "                \".*\\\\.res$\",\n",
      "                \".*\\\\.rsmi$\",\n",
      "                \".*\\\\.rxn$\",\n",
      "                \".*\\\\.sd$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.siesta$\",\n",
      "                \".*\\\\.smi$\",\n",
      "                \".*\\\\.smiles$\",\n",
      "                \".*\\\\.smy$\",\n",
      "                \".*\\\\.sy2$\",\n",
      "                \".*\\\\.t41$\",\n",
      "                \".*\\\\.tdd$\",\n",
      "                \".*\\\\.text$\",\n",
      "                \".*\\\\.therm$\",\n",
      "                \".*\\\\.tmol$\",\n",
      "                \".*\\\\.txt$\",\n",
      "                \".*\\\\.txyz$\",\n",
      "                \".*\\\\.unixyz$\",\n",
      "                \".*\\\\.VASP$\",\n",
      "                \".*\\\\.vmol$\",\n",
      "                \".*\\\\.xml$\",\n",
      "                \".*\\\\.xsf$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.xyz$\",\n",
      "                \".*\\\\.yob$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/babel/ref_babel.convert.mol2\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.acesin$\",\n",
      "                \".*\\\\.adf$\",\n",
      "                \".*\\\\.alc$\",\n",
      "                \".*\\\\.ascii$\",\n",
      "                \".*\\\\.bgf$\",\n",
      "                \".*\\\\.box$\",\n",
      "                \".*\\\\.bs$\",\n",
      "                \".*\\\\.c3d1$\",\n",
      "                \".*\\\\.c3d2$\",\n",
      "                \".*\\\\.cac$\",\n",
      "                \".*\\\\.caccrt$\",\n",
      "                \".*\\\\.cache$\",\n",
      "                \".*\\\\.cacint$\",\n",
      "                \".*\\\\.can$\",\n",
      "                \".*\\\\.cdjson$\",\n",
      "                \".*\\\\.cdxml$\",\n",
      "                \".*\\\\.cht$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.ck$\",\n",
      "                \".*\\\\.cml$\",\n",
      "                \".*\\\\.cmlr$\",\n",
      "                \".*\\\\.com$\",\n",
      "                \".*\\\\.confabreport$\",\n",
      "                \".*\\\\.CONFIG$\",\n",
      "                \".*\\\\.CONTCAR$\",\n",
      "                \".*\\\\.CONTFF$\",\n",
      "                \".*\\\\.copy$\",\n",
      "                \".*\\\\.crk2d$\",\n",
      "                \".*\\\\.crk3d$\",\n",
      "                \".*\\\\.csr$\",\n",
      "                \".*\\\\.cssr$\",\n",
      "                \".*\\\\.ct$\",\n",
      "                \".*\\\\.cub$\",\n",
      "                \".*\\\\.cube$\",\n",
      "                \".*\\\\.dalmol$\",\n",
      "                \".*\\\\.dmol$\",\n",
      "                \".*\\\\.dx$\",\n",
      "                \".*\\\\.ent$\",\n",
      "                \".*\\\\.exyz$\",\n",
      "                \".*\\\\.fa$\",\n",
      "                \".*\\\\.fasta$\",\n",
      "                \".*\\\\.feat$\",\n",
      "                \".*\\\\.fh$\",\n",
      "                \".*\\\\.fhiaims$\",\n",
      "                \".*\\\\.fix$\",\n",
      "                \".*\\\\.fps$\",\n",
      "                \".*\\\\.fpt$\",\n",
      "                \".*\\\\.fract$\",\n",
      "                \".*\\\\.fs$\",\n",
      "                \".*\\\\.fsa$\",\n",
      "                \".*\\\\.gamin$\",\n",
      "                \".*\\\\.gau$\",\n",
      "                \".*\\\\.gjc$\",\n",
      "                \".*\\\\.gjf$\",\n",
      "                \".*\\\\.gpr$\",\n",
      "                \".*\\\\.gr96$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.gukin$\",\n",
      "                \".*\\\\.gukout$\",\n",
      "                \".*\\\\.gzmat$\",\n",
      "                \".*\\\\.hin$\",\n",
      "                \".*\\\\.inchi$\",\n",
      "                \".*\\\\.inchikey$\",\n",
      "                \".*\\\\.inp$\",\n",
      "                \".*\\\\.jin$\",\n",
      "                \".*\\\\.k$\",\n",
      "                \".*\\\\.lmpdat$\",\n",
      "                \".*\\\\.lpmd$\",\n",
      "                \".*\\\\.mcdl$\",\n",
      "                \".*\\\\.mcif$\",\n",
      "                \".*\\\\.MDFF$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.ml2$\",\n",
      "                \".*\\\\.mmcif$\",\n",
      "                \".*\\\\.mmd$\",\n",
      "                \".*\\\\.mmod$\",\n",
      "                \".*\\\\.mna$\",\n",
      "                \".*\\\\.mol$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.mold$\",\n",
      "                \".*\\\\.molden$\",\n",
      "                \".*\\\\.molf$\",\n",
      "                \".*\\\\.molreport$\",\n",
      "                \".*\\\\.mop$\",\n",
      "                \".*\\\\.mopcrt$\",\n",
      "                \".*\\\\.mopin$\",\n",
      "                \".*\\\\.mp$\",\n",
      "                \".*\\\\.mpc$\",\n",
      "                \".*\\\\.mpd$\",\n",
      "                \".*\\\\.mpqcin$\",\n",
      "                \".*\\\\.mrv$\",\n",
      "                \".*\\\\.msms$\",\n",
      "                \".*\\\\.nul$\",\n",
      "                \".*\\\\.nw$\",\n",
      "                \".*\\\\.orcainp$\",\n",
      "                \".*\\\\.outmol$\",\n",
      "                \".*\\\\.paint$\",\n",
      "                \".*\\\\.pcjson$\",\n",
      "                \".*\\\\.pcm$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.pdbqt$\",\n",
      "                \".*\\\\.png$\",\n",
      "                \".*\\\\.pointcloud$\",\n",
      "                \".*\\\\.POSCAR$\",\n",
      "                \".*\\\\.POSFF$\",\n",
      "                \".*\\\\.pov$\",\n",
      "                \".*\\\\.pqr$\",\n",
      "                \".*\\\\.pqs$\",\n",
      "                \".*\\\\.qcin$\",\n",
      "                \".*\\\\.report$\",\n",
      "                \".*\\\\.rsmi$\",\n",
      "                \".*\\\\.rxn$\",\n",
      "                \".*\\\\.sd$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.smi$\",\n",
      "                \".*\\\\.smiles$\",\n",
      "                \".*\\\\.stl$\",\n",
      "                \".*\\\\.svg$\",\n",
      "                \".*\\\\.sy2$\",\n",
      "                \".*\\\\.tdd$\",\n",
      "                \".*\\\\.text$\",\n",
      "                \".*\\\\.therm$\",\n",
      "                \".*\\\\.tmol$\",\n",
      "                \".*\\\\.txt$\",\n",
      "                \".*\\\\.txyz$\",\n",
      "                \".*\\\\.unixyz$\",\n",
      "                \".*\\\\.VASP$\",\n",
      "                \".*\\\\.vmol$\",\n",
      "                \".*\\\\.xed$\",\n",
      "                \".*\\\\.xyz$\",\n",
      "                \".*\\\\.yob$\",\n",
      "                \".*\\\\.zin$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"babel_add_hydrogens\",\n",
      "          \"description\": \"Adds hydrogen atoms to small molecules.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the babel_add_hydrogens tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_chemistry/master/biobb_chemistry/test/data/config/config_babel_add_hydrogens.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/babel/babel.no.H.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.abinit$\",\n",
      "                \".*\\\\.acesout$\",\n",
      "                \".*\\\\.acr$\",\n",
      "                \".*\\\\.adfout$\",\n",
      "                \".*\\\\.alc$\",\n",
      "                \".*\\\\.aoforce$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.axsf$\",\n",
      "                \".*\\\\.bgf$\",\n",
      "                \".*\\\\.box$\",\n",
      "                \".*\\\\.bs$\",\n",
      "                \".*\\\\.c09out$\",\n",
      "                \".*\\\\.c3d2$\",\n",
      "                \".*\\\\.caccrt$\",\n",
      "                \".*\\\\.can$\",\n",
      "                \".*\\\\.car$\",\n",
      "                \".*\\\\.castep$\",\n",
      "                \".*\\\\.ccc$\",\n",
      "                \".*\\\\.cdjson$\",\n",
      "                \".*\\\\.cdx$\",\n",
      "                \".*\\\\.cdxml$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.ck$\",\n",
      "                \".*\\\\.cml$\",\n",
      "                \".*\\\\.cmlr$\",\n",
      "                \".*\\\\.CONFIG$\",\n",
      "                \".*\\\\.CONTCAR$\",\n",
      "                \".*\\\\.CONTFF$\",\n",
      "                \".*\\\\.crk2d$\",\n",
      "                \".*\\\\.crk3d$\",\n",
      "                \".*\\\\.ct$\",\n",
      "                \".*\\\\.cub$\",\n",
      "                \".*\\\\.cube$\",\n",
      "                \".*\\\\.dallog$\",\n",
      "                \".*\\\\.dalmol$\",\n",
      "                \".*\\\\.dat$\",\n",
      "                \".*\\\\.dmol$\",\n",
      "                \".*\\\\.dx$\",\n",
      "                \".*\\\\.ent$\",\n",
      "                \".*\\\\.exyz$\",\n",
      "                \".*\\\\.fa$\",\n",
      "                \".*\\\\.fasta$\",\n",
      "                \".*\\\\.fch$\",\n",
      "                \".*\\\\.fchk$\",\n",
      "                \".*\\\\.fck$\",\n",
      "                \".*\\\\.feat$\",\n",
      "                \".*\\\\.fhiaims$\",\n",
      "                \".*\\\\.fract$\",\n",
      "                \".*\\\\.fs$\",\n",
      "                \".*\\\\.fsa$\",\n",
      "                \".*\\\\.g03$\",\n",
      "                \".*\\\\.g09$\",\n",
      "                \".*\\\\.g92$\",\n",
      "                \".*\\\\.g94$\",\n",
      "                \".*\\\\.g98$\",\n",
      "                \".*\\\\.gal$\",\n",
      "                \".*\\\\.gam$\",\n",
      "                \".*\\\\.gamess$\",\n",
      "                \".*\\\\.gamin$\",\n",
      "                \".*\\\\.gamout$\",\n",
      "                \".*\\\\.got$\",\n",
      "                \".*\\\\.gpr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.gukin$\",\n",
      "                \".*\\\\.gukout$\",\n",
      "                \".*\\\\.gzmat$\",\n",
      "                \".*\\\\.hin$\",\n",
      "                \".*\\\\.HISTORY$\",\n",
      "                \".*\\\\.inchi$\",\n",
      "                \".*\\\\.inp$\",\n",
      "                \".*\\\\.ins$\",\n",
      "                \".*\\\\.jin$\",\n",
      "                \".*\\\\.jout$\",\n",
      "                \".*\\\\.log$\",\n",
      "                \".*\\\\.lpmd$\",\n",
      "                \".*\\\\.mcdl$\",\n",
      "                \".*\\\\.mcif$\",\n",
      "                \".*\\\\.MDFF$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.ml2$\",\n",
      "                \".*\\\\.mmcif$\",\n",
      "                \".*\\\\.mmd$\",\n",
      "                \".*\\\\.mmod$\",\n",
      "                \".*\\\\.mol$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.mold$\",\n",
      "                \".*\\\\.molden$\",\n",
      "                \".*\\\\.molf$\",\n",
      "                \".*\\\\.moo$\",\n",
      "                \".*\\\\.mop$\",\n",
      "                \".*\\\\.mopcrt$\",\n",
      "                \".*\\\\.mopin$\",\n",
      "                \".*\\\\.mopout$\",\n",
      "                \".*\\\\.mpc$\",\n",
      "                \".*\\\\.mpo$\",\n",
      "                \".*\\\\.mpqc$\",\n",
      "                \".*\\\\.mrv$\",\n",
      "                \".*\\\\.msi$\",\n",
      "                \".*\\\\.nwo$\",\n",
      "                \".*\\\\.orca$\",\n",
      "                \".*\\\\.out$\",\n",
      "                \".*\\\\.outmol$\",\n",
      "                \".*\\\\.output$\",\n",
      "                \".*\\\\.pc$\",\n",
      "                \".*\\\\.pcjson$\",\n",
      "                \".*\\\\.pcm$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.pdbqt$\",\n",
      "                \".*\\\\.png$\",\n",
      "                \".*\\\\.pos$\",\n",
      "                \".*\\\\.POSCAR$\",\n",
      "                \".*\\\\.POSFF$\",\n",
      "                \".*\\\\.pqr$\",\n",
      "                \".*\\\\.pqs$\",\n",
      "                \".*\\\\.prep$\",\n",
      "                \".*\\\\.pwscf$\",\n",
      "                \".*\\\\.qcout$\",\n",
      "                \".*\\\\.res$\",\n",
      "                \".*\\\\.rsmi$\",\n",
      "                \".*\\\\.rxn$\",\n",
      "                \".*\\\\.sd$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.siesta$\",\n",
      "                \".*\\\\.smi$\",\n",
      "                \".*\\\\.smiles$\",\n",
      "                \".*\\\\.smy$\",\n",
      "                \".*\\\\.sy2$\",\n",
      "                \".*\\\\.t41$\",\n",
      "                \".*\\\\.tdd$\",\n",
      "                \".*\\\\.text$\",\n",
      "                \".*\\\\.therm$\",\n",
      "                \".*\\\\.tmol$\",\n",
      "                \".*\\\\.txt$\",\n",
      "                \".*\\\\.txyz$\",\n",
      "                \".*\\\\.unixyz$\",\n",
      "                \".*\\\\.VASP$\",\n",
      "                \".*\\\\.vmol$\",\n",
      "                \".*\\\\.xml$\",\n",
      "                \".*\\\\.xsf$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.xyz$\",\n",
      "                \".*\\\\.yob$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/babel/ref_babel.hydrogens.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.acesin$\",\n",
      "                \".*\\\\.adf$\",\n",
      "                \".*\\\\.alc$\",\n",
      "                \".*\\\\.ascii$\",\n",
      "                \".*\\\\.bgf$\",\n",
      "                \".*\\\\.box$\",\n",
      "                \".*\\\\.bs$\",\n",
      "                \".*\\\\.c3d1$\",\n",
      "                \".*\\\\.c3d2$\",\n",
      "                \".*\\\\.cac$\",\n",
      "                \".*\\\\.caccrt$\",\n",
      "                \".*\\\\.cache$\",\n",
      "                \".*\\\\.cacint$\",\n",
      "                \".*\\\\.can$\",\n",
      "                \".*\\\\.cdjson$\",\n",
      "                \".*\\\\.cdxml$\",\n",
      "                \".*\\\\.cht$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.ck$\",\n",
      "                \".*\\\\.cml$\",\n",
      "                \".*\\\\.cmlr$\",\n",
      "                \".*\\\\.com$\",\n",
      "                \".*\\\\.confabreport$\",\n",
      "                \".*\\\\.CONFIG$\",\n",
      "                \".*\\\\.CONTCAR$\",\n",
      "                \".*\\\\.CONTFF$\",\n",
      "                \".*\\\\.copy$\",\n",
      "                \".*\\\\.crk2d$\",\n",
      "                \".*\\\\.crk3d$\",\n",
      "                \".*\\\\.csr$\",\n",
      "                \".*\\\\.cssr$\",\n",
      "                \".*\\\\.ct$\",\n",
      "                \".*\\\\.cub$\",\n",
      "                \".*\\\\.cube$\",\n",
      "                \".*\\\\.dalmol$\",\n",
      "                \".*\\\\.dmol$\",\n",
      "                \".*\\\\.dx$\",\n",
      "                \".*\\\\.ent$\",\n",
      "                \".*\\\\.exyz$\",\n",
      "                \".*\\\\.fa$\",\n",
      "                \".*\\\\.fasta$\",\n",
      "                \".*\\\\.feat$\",\n",
      "                \".*\\\\.fh$\",\n",
      "                \".*\\\\.fhiaims$\",\n",
      "                \".*\\\\.fix$\",\n",
      "                \".*\\\\.fps$\",\n",
      "                \".*\\\\.fpt$\",\n",
      "                \".*\\\\.fract$\",\n",
      "                \".*\\\\.fs$\",\n",
      "                \".*\\\\.fsa$\",\n",
      "                \".*\\\\.gamin$\",\n",
      "                \".*\\\\.gau$\",\n",
      "                \".*\\\\.gjc$\",\n",
      "                \".*\\\\.gjf$\",\n",
      "                \".*\\\\.gpr$\",\n",
      "                \".*\\\\.gr96$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.gukin$\",\n",
      "                \".*\\\\.gukout$\",\n",
      "                \".*\\\\.gzmat$\",\n",
      "                \".*\\\\.hin$\",\n",
      "                \".*\\\\.inchi$\",\n",
      "                \".*\\\\.inchikey$\",\n",
      "                \".*\\\\.inp$\",\n",
      "                \".*\\\\.jin$\",\n",
      "                \".*\\\\.k$\",\n",
      "                \".*\\\\.lmpdat$\",\n",
      "                \".*\\\\.lpmd$\",\n",
      "                \".*\\\\.mcdl$\",\n",
      "                \".*\\\\.mcif$\",\n",
      "                \".*\\\\.MDFF$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.ml2$\",\n",
      "                \".*\\\\.mmcif$\",\n",
      "                \".*\\\\.mmd$\",\n",
      "                \".*\\\\.mmod$\",\n",
      "                \".*\\\\.mna$\",\n",
      "                \".*\\\\.mol$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.mold$\",\n",
      "                \".*\\\\.molden$\",\n",
      "                \".*\\\\.molf$\",\n",
      "                \".*\\\\.molreport$\",\n",
      "                \".*\\\\.mop$\",\n",
      "                \".*\\\\.mopcrt$\",\n",
      "                \".*\\\\.mopin$\",\n",
      "                \".*\\\\.mp$\",\n",
      "                \".*\\\\.mpc$\",\n",
      "                \".*\\\\.mpd$\",\n",
      "                \".*\\\\.mpqcin$\",\n",
      "                \".*\\\\.mrv$\",\n",
      "                \".*\\\\.msms$\",\n",
      "                \".*\\\\.nul$\",\n",
      "                \".*\\\\.nw$\",\n",
      "                \".*\\\\.orcainp$\",\n",
      "                \".*\\\\.outmol$\",\n",
      "                \".*\\\\.paint$\",\n",
      "                \".*\\\\.pcjson$\",\n",
      "                \".*\\\\.pcm$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.pdbqt$\",\n",
      "                \".*\\\\.png$\",\n",
      "                \".*\\\\.pointcloud$\",\n",
      "                \".*\\\\.POSCAR$\",\n",
      "                \".*\\\\.POSFF$\",\n",
      "                \".*\\\\.pov$\",\n",
      "                \".*\\\\.pqr$\",\n",
      "                \".*\\\\.pqs$\",\n",
      "                \".*\\\\.qcin$\",\n",
      "                \".*\\\\.report$\",\n",
      "                \".*\\\\.rsmi$\",\n",
      "                \".*\\\\.rxn$\",\n",
      "                \".*\\\\.sd$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.smi$\",\n",
      "                \".*\\\\.smiles$\",\n",
      "                \".*\\\\.stl$\",\n",
      "                \".*\\\\.svg$\",\n",
      "                \".*\\\\.sy2$\",\n",
      "                \".*\\\\.tdd$\",\n",
      "                \".*\\\\.text$\",\n",
      "                \".*\\\\.therm$\",\n",
      "                \".*\\\\.tmol$\",\n",
      "                \".*\\\\.txt$\",\n",
      "                \".*\\\\.txyz$\",\n",
      "                \".*\\\\.unixyz$\",\n",
      "                \".*\\\\.VASP$\",\n",
      "                \".*\\\\.vmol$\",\n",
      "                \".*\\\\.xed$\",\n",
      "                \".*\\\\.xyz$\",\n",
      "                \".*\\\\.yob$\",\n",
      "                \".*\\\\.zin$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"babel_remove_hydrogens\",\n",
      "          \"description\": \"Removes hydrogen atoms to small molecules.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the babel_remove_hydrogens tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_chemistry/master/biobb_chemistry/test/data/config/config_babel_remove_hydrogens.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/babel/babel.H.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.abinit$\",\n",
      "                \".*\\\\.acesout$\",\n",
      "                \".*\\\\.acr$\",\n",
      "                \".*\\\\.adfout$\",\n",
      "                \".*\\\\.alc$\",\n",
      "                \".*\\\\.aoforce$\",\n",
      "                \".*\\\\.arc$\",\n",
      "                \".*\\\\.axsf$\",\n",
      "                \".*\\\\.bgf$\",\n",
      "                \".*\\\\.box$\",\n",
      "                \".*\\\\.bs$\",\n",
      "                \".*\\\\.c09out$\",\n",
      "                \".*\\\\.c3d2$\",\n",
      "                \".*\\\\.caccrt$\",\n",
      "                \".*\\\\.can$\",\n",
      "                \".*\\\\.car$\",\n",
      "                \".*\\\\.castep$\",\n",
      "                \".*\\\\.ccc$\",\n",
      "                \".*\\\\.cdjson$\",\n",
      "                \".*\\\\.cdx$\",\n",
      "                \".*\\\\.cdxml$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.ck$\",\n",
      "                \".*\\\\.cml$\",\n",
      "                \".*\\\\.cmlr$\",\n",
      "                \".*\\\\.CONFIG$\",\n",
      "                \".*\\\\.CONTCAR$\",\n",
      "                \".*\\\\.CONTFF$\",\n",
      "                \".*\\\\.crk2d$\",\n",
      "                \".*\\\\.crk3d$\",\n",
      "                \".*\\\\.ct$\",\n",
      "                \".*\\\\.cub$\",\n",
      "                \".*\\\\.cube$\",\n",
      "                \".*\\\\.dallog$\",\n",
      "                \".*\\\\.dalmol$\",\n",
      "                \".*\\\\.dat$\",\n",
      "                \".*\\\\.dmol$\",\n",
      "                \".*\\\\.dx$\",\n",
      "                \".*\\\\.ent$\",\n",
      "                \".*\\\\.exyz$\",\n",
      "                \".*\\\\.fa$\",\n",
      "                \".*\\\\.fasta$\",\n",
      "                \".*\\\\.fch$\",\n",
      "                \".*\\\\.fchk$\",\n",
      "                \".*\\\\.fck$\",\n",
      "                \".*\\\\.feat$\",\n",
      "                \".*\\\\.fhiaims$\",\n",
      "                \".*\\\\.fract$\",\n",
      "                \".*\\\\.fs$\",\n",
      "                \".*\\\\.fsa$\",\n",
      "                \".*\\\\.g03$\",\n",
      "                \".*\\\\.g09$\",\n",
      "                \".*\\\\.g92$\",\n",
      "                \".*\\\\.g94$\",\n",
      "                \".*\\\\.g98$\",\n",
      "                \".*\\\\.gal$\",\n",
      "                \".*\\\\.gam$\",\n",
      "                \".*\\\\.gamess$\",\n",
      "                \".*\\\\.gamin$\",\n",
      "                \".*\\\\.gamout$\",\n",
      "                \".*\\\\.got$\",\n",
      "                \".*\\\\.gpr$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.gukin$\",\n",
      "                \".*\\\\.gukout$\",\n",
      "                \".*\\\\.gzmat$\",\n",
      "                \".*\\\\.hin$\",\n",
      "                \".*\\\\.HISTORY$\",\n",
      "                \".*\\\\.inchi$\",\n",
      "                \".*\\\\.inp$\",\n",
      "                \".*\\\\.ins$\",\n",
      "                \".*\\\\.jin$\",\n",
      "                \".*\\\\.jout$\",\n",
      "                \".*\\\\.log$\",\n",
      "                \".*\\\\.lpmd$\",\n",
      "                \".*\\\\.mcdl$\",\n",
      "                \".*\\\\.mcif$\",\n",
      "                \".*\\\\.MDFF$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.ml2$\",\n",
      "                \".*\\\\.mmcif$\",\n",
      "                \".*\\\\.mmd$\",\n",
      "                \".*\\\\.mmod$\",\n",
      "                \".*\\\\.mol$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.mold$\",\n",
      "                \".*\\\\.molden$\",\n",
      "                \".*\\\\.molf$\",\n",
      "                \".*\\\\.moo$\",\n",
      "                \".*\\\\.mop$\",\n",
      "                \".*\\\\.mopcrt$\",\n",
      "                \".*\\\\.mopin$\",\n",
      "                \".*\\\\.mopout$\",\n",
      "                \".*\\\\.mpc$\",\n",
      "                \".*\\\\.mpo$\",\n",
      "                \".*\\\\.mpqc$\",\n",
      "                \".*\\\\.mrv$\",\n",
      "                \".*\\\\.msi$\",\n",
      "                \".*\\\\.nwo$\",\n",
      "                \".*\\\\.orca$\",\n",
      "                \".*\\\\.out$\",\n",
      "                \".*\\\\.outmol$\",\n",
      "                \".*\\\\.output$\",\n",
      "                \".*\\\\.pc$\",\n",
      "                \".*\\\\.pcjson$\",\n",
      "                \".*\\\\.pcm$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.pdbqt$\",\n",
      "                \".*\\\\.png$\",\n",
      "                \".*\\\\.pos$\",\n",
      "                \".*\\\\.POSCAR$\",\n",
      "                \".*\\\\.POSFF$\",\n",
      "                \".*\\\\.pqr$\",\n",
      "                \".*\\\\.pqs$\",\n",
      "                \".*\\\\.prep$\",\n",
      "                \".*\\\\.pwscf$\",\n",
      "                \".*\\\\.qcout$\",\n",
      "                \".*\\\\.res$\",\n",
      "                \".*\\\\.rsmi$\",\n",
      "                \".*\\\\.rxn$\",\n",
      "                \".*\\\\.sd$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.siesta$\",\n",
      "                \".*\\\\.smi$\",\n",
      "                \".*\\\\.smiles$\",\n",
      "                \".*\\\\.smy$\",\n",
      "                \".*\\\\.sy2$\",\n",
      "                \".*\\\\.t41$\",\n",
      "                \".*\\\\.tdd$\",\n",
      "                \".*\\\\.text$\",\n",
      "                \".*\\\\.therm$\",\n",
      "                \".*\\\\.tmol$\",\n",
      "                \".*\\\\.txt$\",\n",
      "                \".*\\\\.txyz$\",\n",
      "                \".*\\\\.unixyz$\",\n",
      "                \".*\\\\.VASP$\",\n",
      "                \".*\\\\.vmol$\",\n",
      "                \".*\\\\.xml$\",\n",
      "                \".*\\\\.xsf$\",\n",
      "                \".*\\\\.xtc$\",\n",
      "                \".*\\\\.xyz$\",\n",
      "                \".*\\\\.yob$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/babel/ref_babel.nohydrogens.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.acesin$\",\n",
      "                \".*\\\\.adf$\",\n",
      "                \".*\\\\.alc$\",\n",
      "                \".*\\\\.ascii$\",\n",
      "                \".*\\\\.bgf$\",\n",
      "                \".*\\\\.box$\",\n",
      "                \".*\\\\.bs$\",\n",
      "                \".*\\\\.c3d1$\",\n",
      "                \".*\\\\.c3d2$\",\n",
      "                \".*\\\\.cac$\",\n",
      "                \".*\\\\.caccrt$\",\n",
      "                \".*\\\\.cache$\",\n",
      "                \".*\\\\.cacint$\",\n",
      "                \".*\\\\.can$\",\n",
      "                \".*\\\\.cdjson$\",\n",
      "                \".*\\\\.cdxml$\",\n",
      "                \".*\\\\.cht$\",\n",
      "                \".*\\\\.cif$\",\n",
      "                \".*\\\\.ck$\",\n",
      "                \".*\\\\.cml$\",\n",
      "                \".*\\\\.cmlr$\",\n",
      "                \".*\\\\.com$\",\n",
      "                \".*\\\\.confabreport$\",\n",
      "                \".*\\\\.CONFIG$\",\n",
      "                \".*\\\\.CONTCAR$\",\n",
      "                \".*\\\\.CONTFF$\",\n",
      "                \".*\\\\.copy$\",\n",
      "                \".*\\\\.crk2d$\",\n",
      "                \".*\\\\.crk3d$\",\n",
      "                \".*\\\\.csr$\",\n",
      "                \".*\\\\.cssr$\",\n",
      "                \".*\\\\.ct$\",\n",
      "                \".*\\\\.cub$\",\n",
      "                \".*\\\\.cube$\",\n",
      "                \".*\\\\.dalmol$\",\n",
      "                \".*\\\\.dmol$\",\n",
      "                \".*\\\\.dx$\",\n",
      "                \".*\\\\.ent$\",\n",
      "                \".*\\\\.exyz$\",\n",
      "                \".*\\\\.fa$\",\n",
      "                \".*\\\\.fasta$\",\n",
      "                \".*\\\\.feat$\",\n",
      "                \".*\\\\.fh$\",\n",
      "                \".*\\\\.fhiaims$\",\n",
      "                \".*\\\\.fix$\",\n",
      "                \".*\\\\.fps$\",\n",
      "                \".*\\\\.fpt$\",\n",
      "                \".*\\\\.fract$\",\n",
      "                \".*\\\\.fs$\",\n",
      "                \".*\\\\.fsa$\",\n",
      "                \".*\\\\.gamin$\",\n",
      "                \".*\\\\.gau$\",\n",
      "                \".*\\\\.gjc$\",\n",
      "                \".*\\\\.gjf$\",\n",
      "                \".*\\\\.gpr$\",\n",
      "                \".*\\\\.gr96$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.gukin$\",\n",
      "                \".*\\\\.gukout$\",\n",
      "                \".*\\\\.gzmat$\",\n",
      "                \".*\\\\.hin$\",\n",
      "                \".*\\\\.inchi$\",\n",
      "                \".*\\\\.inchikey$\",\n",
      "                \".*\\\\.inp$\",\n",
      "                \".*\\\\.jin$\",\n",
      "                \".*\\\\.k$\",\n",
      "                \".*\\\\.lmpdat$\",\n",
      "                \".*\\\\.lpmd$\",\n",
      "                \".*\\\\.mcdl$\",\n",
      "                \".*\\\\.mcif$\",\n",
      "                \".*\\\\.MDFF$\",\n",
      "                \".*\\\\.mdl$\",\n",
      "                \".*\\\\.ml2$\",\n",
      "                \".*\\\\.mmcif$\",\n",
      "                \".*\\\\.mmd$\",\n",
      "                \".*\\\\.mmod$\",\n",
      "                \".*\\\\.mna$\",\n",
      "                \".*\\\\.mol$\",\n",
      "                \".*\\\\.mol2$\",\n",
      "                \".*\\\\.mold$\",\n",
      "                \".*\\\\.molden$\",\n",
      "                \".*\\\\.molf$\",\n",
      "                \".*\\\\.molreport$\",\n",
      "                \".*\\\\.mop$\",\n",
      "                \".*\\\\.mopcrt$\",\n",
      "                \".*\\\\.mopin$\",\n",
      "                \".*\\\\.mp$\",\n",
      "                \".*\\\\.mpc$\",\n",
      "                \".*\\\\.mpd$\",\n",
      "                \".*\\\\.mpqcin$\",\n",
      "                \".*\\\\.mrv$\",\n",
      "                \".*\\\\.msms$\",\n",
      "                \".*\\\\.nul$\",\n",
      "                \".*\\\\.nw$\",\n",
      "                \".*\\\\.orcainp$\",\n",
      "                \".*\\\\.outmol$\",\n",
      "                \".*\\\\.paint$\",\n",
      "                \".*\\\\.pcjson$\",\n",
      "                \".*\\\\.pcm$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.pdbqt$\",\n",
      "                \".*\\\\.png$\",\n",
      "                \".*\\\\.pointcloud$\",\n",
      "                \".*\\\\.POSCAR$\",\n",
      "                \".*\\\\.POSFF$\",\n",
      "                \".*\\\\.pov$\",\n",
      "                \".*\\\\.pqr$\",\n",
      "                \".*\\\\.pqs$\",\n",
      "                \".*\\\\.qcin$\",\n",
      "                \".*\\\\.report$\",\n",
      "                \".*\\\\.rsmi$\",\n",
      "                \".*\\\\.rxn$\",\n",
      "                \".*\\\\.sd$\",\n",
      "                \".*\\\\.sdf$\",\n",
      "                \".*\\\\.smi$\",\n",
      "                \".*\\\\.smiles$\",\n",
      "                \".*\\\\.stl$\",\n",
      "                \".*\\\\.svg$\",\n",
      "                \".*\\\\.sy2$\",\n",
      "                \".*\\\\.tdd$\",\n",
      "                \".*\\\\.text$\",\n",
      "                \".*\\\\.therm$\",\n",
      "                \".*\\\\.tmol$\",\n",
      "                \".*\\\\.txt$\",\n",
      "                \".*\\\\.txyz$\",\n",
      "                \".*\\\\.unixyz$\",\n",
      "                \".*\\\\.VASP$\",\n",
      "                \".*\\\\.vmol$\",\n",
      "                \".*\\\\.xed$\",\n",
      "                \".*\\\\.xyz$\",\n",
      "                \".*\\\\.yob$\",\n",
      "                \".*\\\\.zin$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"babel_minimize\",\n",
      "          \"description\": \"Energetically minimize small molecules.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the babel_minimize tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_chemistry/master/biobb_chemistry/test/data/config/config_babel_minimize.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/babel/babel.minimize.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/babel/ref_babel.minimize.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.mol2$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"reduce_add_hydrogens\",\n",
      "          \"description\": \"Adds hydrogen atoms to small molecules.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the reduce_add_hydrogens tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_chemistry/master/biobb_chemistry/test/data/config/config_reduce_add_hydrogens.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/ambertools/reduce.no.H.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/ambertools/ref_reduce.add.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"reduce_remove_hydrogens\",\n",
      "          \"description\": \"Removes hydrogen atoms to small molecules.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the reduce_remove_hydrogens tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/data/ambertools/reduce.H.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_chemistry/raw/master/biobb_chemistry/test/reference/ambertools/ref_reduce.remove.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"biobb_io\",\n",
      "      \"tools\": [\n",
      "        {\n",
      "          \"id\": \"ligand\",\n",
      "          \"description\": \"Downloads a ligand file from the MMB REST API.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the ligand tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_io/master/biobb_io/test/data/config/config_ligand.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output PDB ligand file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_io/raw/master/biobb_io/test/reference/api/ligand_12d.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"pdb\",\n",
      "          \"description\": \"Downloads a PDB file from the RCSB or MMB REST APIs.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the pdb tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output PDB file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_io/raw/master/biobb_io/test/reference/api/pdb_1ubq.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"pdb_variants\",\n",
      "          \"description\": \"Creates a text file containing a list of all the variants mapped to a RSCB PDB code from the corresponding UNIPROT entries.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the pdb_variants tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_mutations_list_txt\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the TXT file containing an ASCII comma separated values of the mutations\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.txt$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"pdb_cluster_zip\",\n",
      "          \"description\": \"Creates a zip file containing all the PDB files in the given sequence similarity cluster percentage of the given RSCB PDB code.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the pdb_cluster_zip tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_pdb_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the ZIP or PDB file containing the output PDB files\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_io/raw/master/biobb_io/test/reference/api/reference_output_pdb_zip_path.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"biobb_md\",\n",
      "      \"tools\": [\n",
      "        {\n",
      "          \"id\": \"pdb2gmx\",\n",
      "          \"description\": \"Creates a compressed (ZIP) GROMACS topology (TOP and ITP files) from a given PDB file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the pdb2gmx tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_md/master/biobb_md/test/data/config/config_pdb2gmx.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input PDB file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/egfr.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output GRO file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_pdb2gmx.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the output TOP topology in zip format\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_pdb2gmx.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"editconf\",\n",
      "          \"description\": \"Creates a GROMACS structure file (GRO) adding the information of the solvent box to the input structure file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the editconf tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input GRO file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/editconf.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output GRO file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_editconf.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"genion\",\n",
      "          \"description\": \"Creates a new compressed GROMACS topology adding ions until reaching the desired concentration to the input compressed GROMACS topology. \",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the genion tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_tpr_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input portable run input TPR file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/genion.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.tpr$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure GRO file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_genion.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the input TOP topology in zip format\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/genion.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the output topology TOP and ITP files zipball\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_genion.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"genrestr\",\n",
      "          \"description\": \"Creates a new GROMACS compressed topology applying the indicated force restrains to the given input compressed topology.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the genrestr tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure PDB, GRO or TPR format\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/genrestr.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.tpr$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_ndx_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input GROMACS index file, NDX format\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/genrestr.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_itp_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the output ITP topology file with restrains\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_genrestr.itp\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.itp$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"grompp\",\n",
      "          \"description\": \"Creates a GROMACS portable binary run input file (TPR) applying the desired properties from the input compressed GROMACS topology.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the grompp tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_md/master/biobb_md/test/data/config/config_grompp.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input GROMACS structure GRO file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/grompp.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the input GROMACS topology TOP and ITP files in zip format\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/grompp.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_tpr_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output portable binary run file TPR\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_grompp.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.tpr$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_cpt_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the input GROMACS checkpoint file CPT\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.cpt$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_ndx_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the input GROMACS index files NDX\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"mdrun\",\n",
      "          \"description\": \"Performs molecular dynamics simulations from an input GROMACS TPR file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the mdrun tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_tpr_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the portable binary run input file TPR\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/mdrun.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.tpr$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_trr_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the GROMACS uncompressed raw trajectory file TRR\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_mdrun.trr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.trr$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output GROMACS structure GRO file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_mdrun.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_edr_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output GROMACS portable energy file EDR\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_mdrun.edr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.edr$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_log_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output GROMACS trajectory log file LOG\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.log$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_xtc_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the GROMACS compressed trajectory file XTC\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xtc$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_cpt_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the output GROMACS checkpoint file CPT\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.cpt$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_dhdl_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the output dhdl\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.xvg$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"make_ndx\",\n",
      "          \"description\": \"Creates a GROMACS index file (NDX) from an input selection and an input GROMACS structure file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the make_ndx tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input GRO/PDB/TPR file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/make_ndx.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tpr$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_ndx_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output index NDX file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_make_ndx.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_ndx_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the input index NDX file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"select\",\n",
      "          \"description\": \"Creates a GROMACS index file (NDX) from an input selection and an input GROMACS structure file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the select tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_md/master/biobb_md/test/data/config/config_select.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input GRO/PDB/TPR file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/make_ndx.tpr\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\",\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.tpr$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_ndx_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output index NDX file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_select.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_ndx_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the input index NDX file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"solvate\",\n",
      "          \"description\": \"Creates a new compressed GROMACS topology file adding solvent molecules to a given input compressed GROMACS topology file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the solvate tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_solute_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input GRO file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/solvate.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output GRO file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_solvate.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the input TOP topology in zip format\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs/solvate.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the output topology in zip format\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs/ref_solvate.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"ndx2resttop\",\n",
      "          \"description\": \"Creates a new GROMACS compressed topology applying the force restrains to the input groups in the input index file to the given input compressed topology.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the ndx2resttop tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_md/master/biobb_md/test/data/config/config_ndx2resttop.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_ndx_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input NDX index file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs_extra/ndx2resttop.ndx\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.ndx$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the input TOP topology in zip format\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs_extra/ndx2resttop.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the output TOP topology in zip format\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs_extra/ref_ndx2resttop.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"append_ligand\",\n",
      "          \"description\": \"Takes a ligand ITP file and inserts it in a topology.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the append_ligand tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the input topology TOP and ITP files zipball\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs_extra/ndx2resttop.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_itp_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the ligand ITP file to be inserted in the topology\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/data/gromacs_extra/pep_ligand.itp\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.itp$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path/Name the output topology TOP and ITP files zipball\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_md/raw/master/biobb_md/test/reference/gromacs_extra/ref_appendligand.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"biobb_model\",\n",
      "      \"tools\": [\n",
      "        {\n",
      "          \"id\": \"fix_side_chain\",\n",
      "          \"description\": \"Reconstructs the missing side chains and heavy atoms of the given PDB file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the fix_side_chain tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input PDB file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_model/raw/master/biobb_model/test/data/model/2ki5.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output PDB file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_model/raw/master/biobb_model/test/reference/model/output_pdb_path.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"mutate\",\n",
      "          \"description\": \"Creates a new PDB file performing the mutations given in a list of amino acid mutations to the input PDB file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the mutate tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_model/master/biobb_model/test/data/config/config_mutate.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input PDB file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_model/raw/master/biobb_model/test/data/model/2ki5.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output PDB file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_model/raw/master/biobb_model/test/reference/model/output_mutated_pdb_path.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"biobb_pmx\",\n",
      "      \"tools\": [\n",
      "        {\n",
      "          \"id\": \"mutate\",\n",
      "          \"description\": \"pmx tool to insert mutated residues in structure files for free energy simulations\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the mutate tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_pmx/master/biobb_pmx/test/data/config/config_mutate.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the input structure file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_pmx/raw/master/biobb_pmx/test/data/pmx/frame99.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the output structure file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_pmx/raw/master/biobb_pmx/test/reference/pmx/ref_output_structure.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_b_structure_path\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Path to the mutated input structure file\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"gentop\",\n",
      "          \"description\": \"pmx tool to generate hybrid GROMACS topologies: adding a B state to an .itp or .top file for a hybrid residue\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the gentop tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_pmx/master/biobb_pmx/test/data/config/config_gentop.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the input GROMACS topology TOP and ITP files in zip format\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_pmx/raw/master/biobb_pmx/test/data/pmx/topology.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_top_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the output TOP topology in zip format\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_pmx/raw/master/biobb_pmx/test/reference/pmx/ref_output_topology.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"analyse\",\n",
      "          \"description\": \"pmx tool to calculate free energies from fast growth thermodynamic integration simulations.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the analyse tool\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_pmx/master/biobb_pmx/test/data/config/config_analyse.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_a_xvg_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the zip file containing the dgdl\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_pmx/raw/master/biobb_pmx/test/data/pmx/xvg_A.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_b_xvg_zip_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path the zip file containing the dgdl\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_pmx/raw/master/biobb_pmx/test/data/pmx/xvg_B.zip\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.zip$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_result_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the TXT results file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_pmx/raw/master/biobb_pmx/test/reference/pmx/ref_result.txt\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.txt$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_work_plot_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Path to the PNG plot results file\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.png$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"biobb_structure_utils\",\n",
      "      \"tools\": [\n",
      "        {\n",
      "          \"id\": \"cat_pdb\",\n",
      "          \"description\": \"Class to concat two PDB structures in a single PDB file.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the cat_pdb tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure1\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input structure 1 file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/cat_protein.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure2\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input structure 2 file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/cat_ligand.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output protein file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/ref_cat_pdb.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"extract_atoms\",\n",
      "          \"description\": \"Class to extract atoms from a 3D structure.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the extract_atoms tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_structure_utils/master/biobb_structure_utils/test/data/config/config_extract_atoms.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input structure file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/2vgb.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output structure file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/OE2_atoms.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"extract_chain\",\n",
      "          \"description\": \"Class to extract a chain from a 3D structure.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the extract_chain tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_structure_utils/master/biobb_structure_utils/test/data/config/config_extract_chain.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input structure file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/extract_chain.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output structure file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/ref_extract_chain.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"extract_heteroatoms\",\n",
      "          \"description\": \"Class to extract hetero-atoms from a 3D structure.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the extract_heteroatoms tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_structure_utils/master/biobb_structure_utils/test/data/config/config_extract_heteroatoms.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input structure file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/extract_heteroatom.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_heteroatom_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output heteroatom file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/ref_extract_heteroatom.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"extract_model\",\n",
      "          \"description\": \"Class to extract a model from a 3D structure.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the extract_model tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_structure_utils/master/biobb_structure_utils/test/data/config/config_extract_model.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input structure file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/extract_model.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output structure file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/ref_extract_model.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"extract_protein\",\n",
      "          \"description\": \"Class to extract a protein from a 3D structure.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the extract_protein tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input structure file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/extract_protein.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_protein_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output protein file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/ref_extract_protein.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"remove_ligand\",\n",
      "          \"description\": \"Class to remove the selected ligand atoms from a 3D structure.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the remove_ligand tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_structure_utils/master/biobb_structure_utils/test/data/config/config_remove_ligand.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input structure file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/WT_aq4_md_1.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output structure file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/WT_apo_md_1.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"remove_pdb_water\",\n",
      "          \"description\": \"Class to remove water molecules from PDB 3D structures.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the remove_pdb_water tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input PDB file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/WT_aq4_md_WAT.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_pdb_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output PDB file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/WT_apo_no_wat.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"renumber_structure\",\n",
      "          \"description\": \"Class to renumber atomic indexes from a 3D structure.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the renumber_structure tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": null,\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input structure file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/cl3.noH.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_structure_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output structure file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/renum_cl3_noH.pdb\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.pdb$\",\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_mapping_json_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output mapping json file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/cl3_output_mapping_json_path.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"sort_gro_residues\",\n",
      "          \"description\": \"Class to sort the selected residues from a GRO 3D structure.\",\n",
      "          \"arguments\": [\n",
      "            {\n",
      "              \"id\": \"config\",\n",
      "              \"required\": false,\n",
      "              \"description\": \"Configuration file for the sort_gro_residues tool.\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_structure_utils/master/biobb_structure_utils/test/data/config/config_sort_gro_residues.json\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.json$\",\n",
      "                \".*\\\\.yml$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"input_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Input GRO file path\",\n",
      "              \"filetype\": \"input\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/data/utils/WT_aq4_md_1.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            },\n",
      "            {\n",
      "              \"id\": \"output_gro_path\",\n",
      "              \"required\": true,\n",
      "              \"description\": \"Output sorted GRO file path\",\n",
      "              \"filetype\": \"output\",\n",
      "              \"sample\": \"https://github.com/bioexcel/biobb_structure_utils/raw/master/biobb_structure_utils/test/reference/utils/WT_aq4_md_sorted.gro\",\n",
      "              \"formats\": [\n",
      "                \".*\\\\.gro$\"\n",
      "              ]\n",
      "            }\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    }\n",
      "  ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "url = apiURL + 'launch'\n",
    "response = get_data(url)\n",
    "\n",
    "print(json.dumps(response.json, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"list_tools_ex\"></a>\n",
    "#### List of tools from a specific package"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"id\": \"biobb_analysis\",\n",
      "  \"tools\": [\n",
      "    {\n",
      "      \"id\": \"gmx_cluster\",\n",
      "      \"description\": \"Creates cluster structures from a given GROMACS compatible trajectory\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the gmx_cluster tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_cluster.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_structure_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.tpr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.brk$\",\n",
      "            \".*\\\\.ent$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the GROMACS trajectory file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.cpt$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_index_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the GROMACS index file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.ndx$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_pdb_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output cluster file\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_cluster.pdb\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.cpt$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"gmx_rms\",\n",
      "      \"description\": \"Performs a Root Mean Square deviation (RMSd) analysis from a given GROMACS compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the gmx_rms tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_rms.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_structure_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.tpr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.brk$\",\n",
      "            \".*\\\\.ent$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the GROMACS trajectory file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.cpt$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_index_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the GROMACS index file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.ndx$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_xvg_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the XVG output file\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_rms.xvg\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xvg$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"gmx_rgyr\",\n",
      "      \"description\": \"Computes the radius of gyration (Rgyr) of a molecule about the x-, y- and z-axes, as a function of time, from a given GROMACS compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the gmx_rgyr tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_rgyr.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_structure_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.tpr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.brk$\",\n",
      "            \".*\\\\.ent$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the GROMACS trajectory file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.cpt$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_index_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the GROMACS index file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.ndx$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_xvg_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the XVG output file\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_rgyr.xvg\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xvg$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"gmx_energy\",\n",
      "      \"description\": \"Extracts energy components from a given GROMACS energy file.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the gmx_energy tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_energy.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_energy_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input EDR file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/energy.edr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.edr$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_xvg_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the XVG output file\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_energy.xvg\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xvg$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"gmx_image\",\n",
      "      \"description\": \"Corrects periodicity (image) from a given GROMACS compatible trajectory file.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the gmx_image tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_image.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the GROMACS trajectory file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.cpt$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the GROMACS input topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.tpr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.brk$\",\n",
      "            \".*\\\\.ent$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_index_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the GROMACS index file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.ndx$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output file\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_image.xtc\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"gmx_trjconv_str\",\n",
      "      \"description\": \"Converts between GROMACS compatible structure file formats and/or extracts a selection of atoms.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the gmx_trjconv_str tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_trjconv_str.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_structure_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.cpt$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the GROMACS input topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.tpr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.brk$\",\n",
      "            \".*\\\\.ent$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_index_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the GROMACS index file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.ndx$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_str_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output file\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_trjconv.str.pdb\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"gmx_trjconv_str_ens\",\n",
      "      \"description\": \"Extracts an ensemble of frames containing a selection of atoms from GROMACS compatible trajectory files.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the gmx_trjconv_str_ens tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_trjconv_str_ens.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the GROMACS trajectory file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.cpt$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the GROMACS input topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/topology.tpr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.tpr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.brk$\",\n",
      "            \".*\\\\.ent$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_index_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the GROMACS index file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.ndx$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_str_ens_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output file\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_trjconv.str.ens.zip\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"gmx_trjconv_trj\",\n",
      "      \"description\": \"Converts between GROMACS compatible trajectory file formats and/or extracts a selection of atoms.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the gmx_trjconv_trj tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_gmx_trjconv_trj.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the GROMACS trajectory file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/trajectory.trr\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.cpt$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_index_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the GROMACS index file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/gromacs/index.ndx\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.ndx$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output file\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/gromacs/ref_trjconv.trj.xtc\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.g96$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.tng$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_average\",\n",
      "      \"description\": \"Calculates a structure average of a given cpptraj compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_average tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_average.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed structure\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.average.pdb\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.rst7$\",\n",
      "            \".*\\\\.ncrst$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.sqm$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_bfactor\",\n",
      "      \"description\": \"Calculates the Bfactor fluctuations of a given cpptraj compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_bfactor tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": null,\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_exp_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the experimental reference file (required if reference = experimental)\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/experimental.1e5t.pdb\",\n",
      "          \"formats\": null\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed analysis\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.bfactor.first.dat\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.dat$\",\n",
      "            \".*\\\\.agr$\",\n",
      "            \".*\\\\.xmgr$\",\n",
      "            \".*\\\\.gnu$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_rms\",\n",
      "      \"description\": \"Calculates the Root Mean Square deviation (RMSd) of a given cpptraj compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_rms tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": null,\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_exp_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the experimental reference file (required if reference = experimental)\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/experimental.1e5t.pdb\",\n",
      "          \"formats\": null\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed analysis\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.rms.first.dat\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.dat$\",\n",
      "            \".*\\\\.agr$\",\n",
      "            \".*\\\\.xmgr$\",\n",
      "            \".*\\\\.gnu$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_rmsf\",\n",
      "      \"description\": \"Calculates the Root Mean Square fluctuations (RMSf) of a given cpptraj compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_rmsf tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": null,\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_exp_path\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Path to the experimental reference file (required if reference = experimental)\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/experimental.1e5t.pdb\",\n",
      "          \"formats\": null\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed analysis\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.rmsf.first.dat\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.dat$\",\n",
      "            \".*\\\\.agr$\",\n",
      "            \".*\\\\.xmgr$\",\n",
      "            \".*\\\\.gnu$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_rgyr\",\n",
      "      \"description\": \"Computes the radius of gyration (Rgyr) from a given cpptraj compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_rgyr tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_rgyr.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output analysis\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.rgyr.dat\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.dat$\",\n",
      "            \".*\\\\.agr$\",\n",
      "            \".*\\\\.xmgr$\",\n",
      "            \".*\\\\.gnu$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_dry\",\n",
      "      \"description\": \"Dehydrates a given cpptraj compatible trajectory stripping out solvent molecules and ions.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_dry tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_dry.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed trajectory\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.dry.netcdf\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.rst7$\",\n",
      "            \".*\\\\.ncrst$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.sqm$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_strip\",\n",
      "      \"description\": \"Strips a defined set of atoms (mask) from a given cpptraj compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_strip tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_strip.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed trajectory\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.strip.netcdf\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.rst7$\",\n",
      "            \".*\\\\.ncrst$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.sqm$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_snapshot\",\n",
      "      \"description\": \"Extracts a particular snapshot from a given cpptraj compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_snapshot tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_snapshot.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed structure\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.snapshot.pdb\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.rst7$\",\n",
      "            \".*\\\\.ncrst$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.sqm$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_slice\",\n",
      "      \"description\": \"Extracts a particular trajectory slice from a given cpptraj compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_slice tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_slice.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed trajectory\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.slice.netcdf\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.rst7$\",\n",
      "            \".*\\\\.ncrst$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.sqm$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_convert\",\n",
      "      \"description\": \"Converts between cpptraj compatible trajectory file formats and/or extracts a selection of atoms or frames.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_convert tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_convert.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed trajectory\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.convert.netcdf\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.rst7$\",\n",
      "            \".*\\\\.ncrst$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.sqm$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_mask\",\n",
      "      \"description\": \"Extracts a selection of atoms from a given cpptraj compatible trajectory.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_mask tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_mask.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed trajectory\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.mask.netcdf\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.rst7$\",\n",
      "            \".*\\\\.ncrst$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.sqm$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"cpptraj_image\",\n",
      "      \"description\": \"Corrects periodicity (image) from a given cpptraj trajectory file.\",\n",
      "      \"arguments\": [\n",
      "        {\n",
      "          \"id\": \"config\",\n",
      "          \"required\": false,\n",
      "          \"description\": \"Configuration file for the cpptraj_image tool\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_image.json\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.json$\",\n",
      "            \".*\\\\.yml$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_top_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input structure or topology file\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.top$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.prmtop$\",\n",
      "            \".*\\\\.parmtop$\",\n",
      "            \".*\\\\.zip$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"input_traj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the input trajectory to be processed\",\n",
      "          \"filetype\": \"input\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.cdf$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.restart$\",\n",
      "            \".*\\\\.ncrestart$\",\n",
      "            \".*\\\\.restartnc$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.charmm$\",\n",
      "            \".*\\\\.cor$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.gro$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.cif$\",\n",
      "            \".*\\\\.arc$\",\n",
      "            \".*\\\\.sqm$\",\n",
      "            \".*\\\\.sdf$\",\n",
      "            \".*\\\\.conflib$\"\n",
      "          ]\n",
      "        },\n",
      "        {\n",
      "          \"id\": \"output_cpptraj_path\",\n",
      "          \"required\": true,\n",
      "          \"description\": \"Path to the output processed trajectory\",\n",
      "          \"filetype\": \"output\",\n",
      "          \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.image.netcdf\",\n",
      "          \"formats\": [\n",
      "            \".*\\\\.crd$\",\n",
      "            \".*\\\\.netcdf$\",\n",
      "            \".*\\\\.rst7$\",\n",
      "            \".*\\\\.ncrst$\",\n",
      "            \".*\\\\.dcd$\",\n",
      "            \".*\\\\.pdb$\",\n",
      "            \".*\\\\.mol2$\",\n",
      "            \".*\\\\.binpos$\",\n",
      "            \".*\\\\.trr$\",\n",
      "            \".*\\\\.xtc$\",\n",
      "            \".*\\\\.sqm$\"\n",
      "          ]\n",
      "        }\n",
      "      ]\n",
      "    }\n",
      "  ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "package = 'biobb_analysis'\n",
    "url = apiURL + 'launch/' + package\n",
    "response = get_data(url)\n",
    "\n",
    "print(json.dumps(response.json, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"tools_prop_ex\"></a>\n",
    "#### Tool's properties"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"id\": \"cpptraj_average\",\n",
      "  \"description\": \"Calculates a structure average of a given cpptraj compatible trajectory.\",\n",
      "  \"arguments\": [\n",
      "    {\n",
      "      \"id\": \"config\",\n",
      "      \"required\": false,\n",
      "      \"description\": \"Configuration file for the cpptraj_average tool\",\n",
      "      \"filetype\": \"input\",\n",
      "      \"sample\": \"https://raw.githubusercontent.com/bioexcel/biobb_analysis/master/biobb_analysis/test/data/config/config_cpptraj_average.json\",\n",
      "      \"formats\": [\n",
      "        \".*\\\\.json$\",\n",
      "        \".*\\\\.yml$\"\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"input_top_path\",\n",
      "      \"required\": true,\n",
      "      \"description\": \"Path to the input structure or topology file\",\n",
      "      \"filetype\": \"input\",\n",
      "      \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.parm.top\",\n",
      "      \"formats\": [\n",
      "        \".*\\\\.top$\",\n",
      "        \".*\\\\.pdb$\",\n",
      "        \".*\\\\.prmtop$\",\n",
      "        \".*\\\\.parmtop$\",\n",
      "        \".*\\\\.zip$\"\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"input_traj_path\",\n",
      "      \"required\": true,\n",
      "      \"description\": \"Path to the input trajectory to be processed\",\n",
      "      \"filetype\": \"input\",\n",
      "      \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/data/ambertools/cpptraj.traj.dcd\",\n",
      "      \"formats\": [\n",
      "        \".*\\\\.crd$\",\n",
      "        \".*\\\\.cdf$\",\n",
      "        \".*\\\\.netcdf$\",\n",
      "        \".*\\\\.restart$\",\n",
      "        \".*\\\\.ncrestart$\",\n",
      "        \".*\\\\.restartnc$\",\n",
      "        \".*\\\\.dcd$\",\n",
      "        \".*\\\\.charmm$\",\n",
      "        \".*\\\\.cor$\",\n",
      "        \".*\\\\.pdb$\",\n",
      "        \".*\\\\.mol2$\",\n",
      "        \".*\\\\.trr$\",\n",
      "        \".*\\\\.gro$\",\n",
      "        \".*\\\\.binpos$\",\n",
      "        \".*\\\\.xtc$\",\n",
      "        \".*\\\\.cif$\",\n",
      "        \".*\\\\.arc$\",\n",
      "        \".*\\\\.sqm$\",\n",
      "        \".*\\\\.sdf$\",\n",
      "        \".*\\\\.conflib$\"\n",
      "      ]\n",
      "    },\n",
      "    {\n",
      "      \"id\": \"output_cpptraj_path\",\n",
      "      \"required\": true,\n",
      "      \"description\": \"Path to the output processed structure\",\n",
      "      \"filetype\": \"output\",\n",
      "      \"sample\": \"https://github.com/bioexcel/biobb_analysis/raw/master/biobb_analysis/test/reference/ambertools/ref_cpptraj.average.pdb\",\n",
      "      \"formats\": [\n",
      "        \".*\\\\.crd$\",\n",
      "        \".*\\\\.netcdf$\",\n",
      "        \".*\\\\.rst7$\",\n",
      "        \".*\\\\.ncrst$\",\n",
      "        \".*\\\\.dcd$\",\n",
      "        \".*\\\\.pdb$\",\n",
      "        \".*\\\\.mol2$\",\n",
      "        \".*\\\\.binpos$\",\n",
      "        \".*\\\\.trr$\",\n",
      "        \".*\\\\.xtc$\",\n",
      "        \".*\\\\.sqm$\"\n",
      "      ]\n",
      "    }\n",
      "  ]\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "package = 'biobb_analysis'\n",
    "tool = 'cpptraj_average'\n",
    "url = apiURL + 'launch/' + package + '/' + tool\n",
    "response = get_data(url)\n",
    "\n",
    "print(json.dumps(response.json, indent=2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"launch_tool_ex\"></a>\n",
    "### Launch tool\n",
    "\n",
    "Definition of functions needed for launch a job:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from io import BytesIO\n",
    "from pathlib import Path\n",
    "\n",
    "# Function used for encode python dictionary to JSON file\n",
    "def encode_config(data):\n",
    "    jsonData = json.dumps(data)\n",
    "    binaryData = jsonData.encode()\n",
    "    return BytesIO(binaryData)\n",
    "\n",
    "# Launch job\n",
    "def launch_job(url, **kwargs):\n",
    "    data = {}\n",
    "    files = {}\n",
    "    # Fill data (output paths) and files (input files) objects\n",
    "    for key, value in kwargs.items():\n",
    "        # Inputs / Outputs\n",
    "        if type(value) is str:\n",
    "            if key.startswith('input'):\n",
    "                files[key] = (value,  open(value, 'rb'))\n",
    "            elif key.startswith('output'):\n",
    "                data[key] = value\n",
    "            elif Path(value).is_file():\n",
    "                files[key] = (value,  open(value, 'rb'))\n",
    "        # Properties (in case properties are provided as a dictionary instead of a file)\n",
    "        if type(value) is dict:\n",
    "            files['config'] = ('prop.json', encode_config(value))\n",
    "    # Request URL with data and files\n",
    "    response = post_data(url, data, files)\n",
    "    # Print REST API response\n",
    "    print(json.dumps(response.json, indent=2))\n",
    "    # Save token if status == 303\n",
    "    if response.status == 303:\n",
    "        token = response.json['token']\n",
    "        return token"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Hereafter we will launch a job on *biobb_analysis.cpptraj_average* tool with the provided *files/* in the files folder of this same repository. The response is a JSON with the status code, the state of the job, a message and a token for checking the job status.\n",
    "\n",
    "<a id=\"tool_yml_ex\"></a>\n",
    "#### Launch job with a YAML file config\n",
    "\n",
    "File config:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```yaml \n",
    "properties:\n",
    "    in_parameters:\n",
    "      start: 1\n",
    "      end: -1\n",
    "      step: 1\n",
    "      mask: c-alpha\n",
    "    out_parameters:\n",
    "      format: pdb\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function below sends POST data and files to the *{package}/{tool}* endpoint. The config properties are sent as a YAML file.\n",
    "\n",
    "The response is a JSON with the status code, the state of the job, a message and a token that will be used for checking the job status in the next step. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"code\": 303,\n",
      "  \"state\": \"RUNNING\",\n",
      "  \"message\": \"The requested job has has been successfully launched, please go to /retrieve/status/{token} for checking job status.\",\n",
      "  \"token\": \"187a022fd3e2ecab8a5d41f406e0e0695496998bc4ac187e2592f7e2e1d50e236f7f0cfeb4f82e4e6a41af996606e20ef72f8619ef2d0c1e8b72e56ec3334dc2\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Launch BioBB on REST API with YAML config file\n",
    "\n",
    "token = launch_job(url = apiURL + 'launch/biobb_analysis/cpptraj_average', \n",
    "                   config = 'files/config.yml',\n",
    "                   input_top_path = 'files/cpptraj.parm.top',\n",
    "                   input_traj_path = 'files/cpptraj.traj.dcd',\n",
    "                   output_cpptraj_path = 'output.cpptraj.average.pdb')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"tool_json_ex\"></a>\n",
    "#### Launch job with a JSON file config\n",
    "\n",
    "File config:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```json\n",
    "{\n",
    "\t\"in_parameters\": {\n",
    "\t\t\"start\": 1,\n",
    "\t\t\"end\": -1,\n",
    "\t\t\"step\": 1,\n",
    "\t\t\"mask\": \"c-alpha\"\n",
    "\t},\n",
    "\t\"out_parameters\": {\n",
    "\t\t\"format\": \"pdb\"\n",
    "\t}\n",
    "}\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function below sends POST data and files to the *{package}/{tool}* endpoint. The config properties are sent as a JSON file.\n",
    "\n",
    "The response is a JSON with the status code, the state of the job, a message and a token that will be used for checking the job status in the next step. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"code\": 303,\n",
      "  \"state\": \"RUNNING\",\n",
      "  \"message\": \"The requested job has has been successfully launched, please go to /retrieve/status/{token} for checking job status.\",\n",
      "  \"token\": \"74cefaa8814fbdb741f9b57e9697dcb1c7fb7bc9e96050460400b3bebf7d1db8ae9597cc581a41c4a8ea84b8a7e8b7dcdd04455362edd8d072e87fdfebacb4fa\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Launch BioBB on REST API with JSON config file\n",
    "\n",
    "token = launch_job(url = apiURL + 'launch/biobb_analysis/cpptraj_average', \n",
    "                   config = 'files/config.json',\n",
    "                   input_top_path = 'files/cpptraj.parm.top',\n",
    "                   input_traj_path = 'files/cpptraj.traj.dcd',\n",
    "                   output_cpptraj_path = 'output.cpptraj.average.pdb')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"tool_dict_ex\"></a>\n",
    "#### Launch job with a python dictionary config"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function below sends POST data and files to the *{package}/{tool}* endpoint. The config properties are sent as a python dictionary embedded in the code.\n",
    "\n",
    "The response is a JSON with the status code, the state of the job, a message and a token that will be used for checking the job status in the next step. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"code\": 303,\n",
      "  \"state\": \"RUNNING\",\n",
      "  \"message\": \"The requested job has has been successfully launched, please go to /retrieve/status/{token} for checking job status.\",\n",
      "  \"token\": \"b51bb21eddd4728574e61bc0e0cfaf696cc2849e88fd965e4bad20f74bf9458ed85809583c9926991278d6ef0fc6c08aa1fff66137dbbb358f59814f6433c4f3\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Launch BioBB on REST API with JSON config file\n",
    "\n",
    "prop = {\n",
    "    \"in_parameters\" : {\n",
    "      \"start\": 1,\n",
    "      \"end\": -1,\n",
    "      \"step\": 1,\n",
    "      \"mask\": \"c-alpha\"\n",
    "    },\n",
    "    \"out_parameters\" : {\n",
    "      \"format\": \"pdb\"\n",
    "    }\n",
    "}\n",
    "\n",
    "token = launch_job(url = apiURL + 'launch/biobb_analysis/cpptraj_average', \n",
    "                   config = prop,\n",
    "                   input_top_path = 'files/cpptraj.parm.top',\n",
    "                   input_traj_path = 'files/cpptraj.traj.dcd',\n",
    "                   output_cpptraj_path = 'output.cpptraj.average.pdb')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"retrieve_status_ex\"></a>\n",
    "### Retrieve status\n",
    "\n",
    "Definition of functions needed for retrieve the status of a job:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import datetime\n",
    "from time import sleep\n",
    "\n",
    "# Checks status until a provided \"ok\" status is returned by the response\n",
    "def check_status(url, ok, error):\n",
    "    counter = 0\n",
    "    while True:\n",
    "        if counter < 10: slp = 1\n",
    "        if counter >= 10 and counter < 60: slp = 10\n",
    "        if counter >= 60: slp = 60\n",
    "        counter = counter + slp\n",
    "        sleep(slp)\n",
    "        r = requests.get(url)\n",
    "        if r.status_code == ok or r.status_code == error:\n",
    "            return counter\n",
    "            break\n",
    "\n",
    "# Function that checks the status and parses the reponse JSON for saving th output files in a list\n",
    "def check_job(token):\n",
    "    # define retrieve status URL\n",
    "    url = apiURL + 'retrieve/status/' + token\n",
    "    # check status until job has finished\n",
    "    counter = check_status(url, 200, 500)\n",
    "    # Get content when status = 200\n",
    "    response = get_data(url)\n",
    "    # Save id for the generated output_files\n",
    "    if response.status == 200:\n",
    "        out_files = []\n",
    "        for outf in response.json['output_files']:\n",
    "            item = { 'id': outf['id'], 'name': outf['name'] }\n",
    "            out_files.append(item)\n",
    "\n",
    "    # Print REST API response\n",
    "    print(\"Total elapsed time: %s\" % str(datetime.timedelta(seconds=counter)))\n",
    "    print(\"REST API JSON response:\")\n",
    "    print(json.dumps(response.json, indent=4))\n",
    "    \n",
    "    if response.status == 200: \n",
    "        return out_files\n",
    "    else: return None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function below checks the status of a job and awaits until the response status is `200`. The response is a JSON with the status code, the state of the job, a message, a list with all the generated output files and the date of the expiration of these files. Additionally, the function also provides the elapsed time since the job has been launched until it has finished. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total elapsed time: 0:00:10\n",
      "REST API JSON response:\n",
      "{\n",
      "    \"code\": 200,\n",
      "    \"state\": \"FINISHED\",\n",
      "    \"message\": \"The requested job has finished successfully, please go to /retrieve/data/{id} for each output_files.\",\n",
      "    \"output_files\": [\n",
      "        {\n",
      "            \"id\": \"5e4181b8e9e9d3.65557139\",\n",
      "            \"name\": \"output.cpptraj.average.pdb\",\n",
      "            \"size\": 77397,\n",
      "            \"mimetype\": \"text/plain\"\n",
      "        }\n",
      "    ],\n",
      "    \"expiration\": \"February 12, 2020 00:00 GMT+0000\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Check job status\n",
    "out_files = check_job(token)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"retrieve_data_ex\"></a>\n",
    "### Retrieve data\n",
    "\n",
    "Definition of functions needed for retrieve the output file(s) generated by a job:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Downloads to disk a file from a given URL\n",
    "def get_file(url, filename):\n",
    "    r = requests.get(url, allow_redirects=True)\n",
    "    file = open(filename,'wb') \n",
    "    file.write(r.content) \n",
    "    file.close()\n",
    "\n",
    "# Retrieves all the files provided in the out_files list\n",
    "def retrieve_data(out_files):\n",
    "    if not out_files:\n",
    "        return \"No files provided\"\n",
    "    for outf in out_files:\n",
    "        get_file(apiURL + 'retrieve/data/' + outf['id'], outf['name'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The function below makes a single call to the *retrieve/data* endpoint for each output file got in the *retrieve/status* endpoint and save the generated file(s) to disk."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save generated file(s) to disk\n",
    "\n",
    "retrieve_data(out_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"practical_cases\"></a>\n",
    "## Practical cases"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we will execute some Bioexcel Building Blocks through the BioBB REST API and with the results we will do some interactions with other python libraries such as [plotly](https://plot.ly/python/offline/) or [nglview](http://nglviewer.org/#nglview)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"example1\"></a>\n",
    "### Example 1: download PDB file from RSCB database"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Launch the *biobb_io.pdb* job that downloads a PDB file from the RSCB database:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"code\": 303,\n",
      "  \"state\": \"RUNNING\",\n",
      "  \"message\": \"The requested job has has been successfully launched, please go to /retrieve/status/{token} for checking job status.\",\n",
      "  \"token\": \"f44c86485d496ecf67b7d405a32b6db52ede4aca53495b3eaf1fb3080b9d2960251d3c87316399a9bd3c2fc477c9f637fb5daef0aa0e4e4a7e170b1241b8cb6e\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Downloading desired PDB file\n",
    "\n",
    "# Create properties dict and inputs/outputs\n",
    "downloaded_pdb = '3EBP.pdb'\n",
    "prop = {\n",
    "    'pdb_code': '3EBP',\n",
    "    'filter': False\n",
    "}\n",
    "\n",
    "# Launch bb on REST API\n",
    "token = launch_job(url = apiURL + 'launch/biobb_io/pdb', \n",
    "                   config = prop,\n",
    "                   output_pdb_path = downloaded_pdb)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total elapsed time: 0:00:06\n",
      "REST API JSON response:\n",
      "{\n",
      "    \"code\": 200,\n",
      "    \"state\": \"FINISHED\",\n",
      "    \"message\": \"The requested job has finished successfully, please go to /retrieve/data/{id} for each output_files.\",\n",
      "    \"output_files\": [\n",
      "        {\n",
      "            \"id\": \"5e4181c8eb5347.47687951\",\n",
      "            \"name\": \"3EBP.pdb\",\n",
      "            \"size\": 609120,\n",
      "            \"mimetype\": \"text/plain\"\n",
      "        }\n",
      "    ],\n",
      "    \"expiration\": \"February 12, 2020 00:00 GMT+0000\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Check job status\n",
    "out_files = check_job(token)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save generated file to disk\n",
    "retrieve_data(out_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Visualize downloaded PDB in NGLView:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import nglview\n",
    "\n",
    "# Show protein\n",
    "view = nglview.show_structure_file(downloaded_pdb)\n",
    "view.add_representation(repr_type='ball+stick', selection='het')\n",
    "view._remote_call('setSize', target='Widget', args=['','600px'])\n",
    "view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src='ngl1.png'></img>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"example2\"></a>\n",
    "### Example 2: extract heteroatom from a given structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Launch the *biobb_structure_utils.extract_heteroatoms* job that extracts a heteroatom from a PDB file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"code\": 303,\n",
      "  \"state\": \"RUNNING\",\n",
      "  \"message\": \"The requested job has has been successfully launched, please go to /retrieve/status/{token} for checking job status.\",\n",
      "  \"token\": \"20973e6d36e8c72eb886e83e70d592d6b1470bf5420aafd900f1cd59dba83d2828713afd1efcbfd68bc5367ae130acd1cec3c277dc0f3235ad4966db82bbf99d\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Extracting heteroatom from a given structure\n",
    "\n",
    "# Create properties dict and inputs/outputs\n",
    "heteroatom = 'CPB.pdb'\n",
    "prop = {\n",
    "    'heteroatoms': [{\n",
    "        'name': 'CPB'\n",
    "    }]\n",
    "}\n",
    "\n",
    "# Launch bb on REST API\n",
    "token = launch_job(url = apiURL + 'launch/biobb_structure_utils/extract_heteroatoms', \n",
    "                   config = prop,\n",
    "                   input_structure_path = downloaded_pdb,\n",
    "                   output_heteroatom_path = heteroatom)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total elapsed time: 0:00:05\n",
      "REST API JSON response:\n",
      "{\n",
      "    \"code\": 200,\n",
      "    \"state\": \"FINISHED\",\n",
      "    \"message\": \"The requested job has finished successfully, please go to /retrieve/data/{id} for each output_files.\",\n",
      "    \"output_files\": [\n",
      "        {\n",
      "            \"id\": \"5e4181d75115b2.19311021\",\n",
      "            \"name\": \"CPB.pdb\",\n",
      "            \"size\": 2268,\n",
      "            \"mimetype\": \"text/plain\"\n",
      "        }\n",
      "    ],\n",
      "    \"expiration\": \"February 12, 2020 00:00 GMT+0000\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Check job status\n",
    "out_files = check_job(token)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save generated file to disk\n",
    "retrieve_data(out_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Visualize generated extracted heteroatom in NGLView:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Show protein\n",
    "view = nglview.show_structure_file(heteroatom)\n",
    "view.add_representation(repr_type='ball+stick', selection='het')\n",
    "view._remote_call('setSize', target='Widget', args=['','600px'])\n",
    "view"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src='ngl2.png'></img>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<a id=\"example3\"></a>\n",
    "### Example 3: extract energy components from a given GROMACS energy file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"code\": 303,\n",
      "  \"state\": \"RUNNING\",\n",
      "  \"message\": \"The requested job has has been successfully launched, please go to /retrieve/status/{token} for checking job status.\",\n",
      "  \"token\": \"fea0cc96a9e4e35f1f4417f9f87c8f93f1443718664fef6c18efb68445211acb76b2a751fad19d3a2e82bb3ede6fc304afe5fa4c97db3ef6748a5f49767c4cff\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# GMXEnergy: Getting system energy by time \n",
    "\n",
    "# Create prop dict and inputs/outputs\n",
    "output_min_ene_xvg ='file_min_ene.xvg'\n",
    "output_min_edr = 'files/1AKI_min.edr'\n",
    "prop = {\n",
    "    'terms':  [\"Potential\"]\n",
    "}\n",
    "\n",
    "# Launch bb on REST API\n",
    "token = launch_job(url = apiURL + 'launch/biobb_analysis/gmx_energy',\n",
    "                   config = prop,\n",
    "                   input_energy_path = output_min_edr,\n",
    "                   output_xvg_path = output_min_ene_xvg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total elapsed time: 0:00:05\n",
      "REST API JSON response:\n",
      "{\n",
      "    \"code\": 200,\n",
      "    \"state\": \"FINISHED\",\n",
      "    \"message\": \"The requested job has finished successfully, please go to /retrieve/data/{id} for each output_files.\",\n",
      "    \"output_files\": [\n",
      "        {\n",
      "            \"id\": \"5e4181e5c4cb89.22227444\",\n",
      "            \"name\": \"file_min_ene.xvg\",\n",
      "            \"size\": 54143,\n",
      "            \"mimetype\": \"text/plain\"\n",
      "        }\n",
      "    ],\n",
      "    \"expiration\": \"February 12, 2020 00:00 GMT+0000\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "# Check job status\n",
    "out_files = check_job(token)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save generated file to disk\n",
    "retrieve_data(out_files)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Visualize generated energy file in plotly:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import plotly\n",
    "import plotly.graph_objs as go\n",
    "\n",
    "#Read data from file and filter energy values higher than 1000 Kj/mol^-1\n",
    "with open(output_min_ene_xvg,'r') as energy_file:\n",
    "    x,y = map(\n",
    "        list,\n",
    "        zip(*[\n",
    "            (float(line.split()[0]),float(line.split()[1]))\n",
    "            for line in energy_file \n",
    "            if not line.startswith((\"#\",\"@\")) \n",
    "            if float(line.split()[1]) < 1000 \n",
    "        ])\n",
    "    )\n",
    "\n",
    "plotly.offline.init_notebook_mode(connected=True)\n",
    "\n",
    "fig = {\n",
    "    \"data\": [go.Scatter(x=x, y=y)],\n",
    "    \"layout\": go.Layout(title=\"Energy Minimization\",\n",
    "                        xaxis=dict(title = \"Energy Minimization Step\"),\n",
    "                        yaxis=dict(title = \"Potential Energy KJ/mol-1\")\n",
    "                       )\n",
    "}\n",
    "\n",
    "plotly.offline.iplot(fig)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src='plot.png'></img>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
